<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>TreeProfiler Tutorial &#8212; treeprofiler [1.2.2-beta] documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=fbc843e6"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Welcome to treeprofiler’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="treeprofiler-tutorial">
<h1><a class="toc-backref" href="#id4" role="doc-backlink">TreeProfiler Tutorial</a><a class="headerlink" href="#treeprofiler-tutorial" title="Link to this heading">¶</a></h1>
</section>
<section id="introduction">
<h1><a class="toc-backref" href="#id5" role="doc-backlink">Introduction</a><a class="headerlink" href="#introduction" title="Link to this heading">¶</a></h1>
<p>TreeProfiler is command-line tool for profiling metadata table into phylogenetic tree with descriptive analysis and output visualization</p>
</section>
<section id="table-of-contents">
<h1><a class="toc-backref" href="#id6" role="doc-backlink">Table of Contents</a><a class="headerlink" href="#table-of-contents" title="Link to this heading">¶</a></h1>
<nav class="contents" id="contents">
<p class="topic-title">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#treeprofiler-tutorial" id="id4">TreeProfiler Tutorial</a></p></li>
<li><p><a class="reference internal" href="#introduction" id="id5">Introduction</a></p></li>
<li><p><a class="reference internal" href="#table-of-contents" id="id6">Table of Contents</a></p></li>
<li><p><a class="reference internal" href="#installation" id="id7">Installation</a></p>
<ul>
<li><p><a class="reference internal" href="#dependencies" id="id8">Dependencies</a></p></li>
<li><p><a class="reference internal" href="#quick-install-via-pip" id="id9">Quick install via pip</a></p></li>
<li><p><a class="reference internal" href="#quick-start-with-examples-dataset" id="id10">Quick Start with examples dataset</a></p></li>
<li><p><a class="reference internal" href="#manual-installation" id="id11">Manual installation</a></p>
<ul>
<li><p><a class="reference internal" href="#install-ete-v4" id="id12">Install ETE v4</a></p></li>
<li><p><a class="reference internal" href="#install-treeprofiler" id="id13">Install TreeProfiler</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#input-files" id="id14">Input files</a></p></li>
<li><p><a class="reference internal" href="#basic-usage" id="id15">Basic Usage</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#using-treeprofiler" id="id16">Using TreeProfiler</a></p>
<ul>
<li><p><a class="reference internal" href="#reading-input-tree" id="id17">Reading input tree</a></p>
<ul>
<li><p><a class="reference internal" href="#tree-format" id="id18">Tree format</a></p></li>
<li><p><a class="reference internal" href="#tree-parser" id="id19">Tree parser</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#treeprofiler-annotate-computing-phylogenetic-profiles-and-annotation" id="id20"><code class="docutils literal notranslate"><span class="pre">treeprofiler-annotate</span></code> computing phylogenetic profiles and annotation</a></p>
<ul>
<li><p><a class="reference internal" href="#annotate-metadata-into-tree" id="id21">Annotate metadata into tree</a></p>
<ul>
<li><p><a class="reference internal" href="#basic-metadata-in-tsv-csv-format" id="id22">Basic metadata in TSV/CSV format</a></p></li>
<li><p><a class="reference internal" href="#metadata-tsv-csv-as-a-numerical-data-matrix" id="id23">Metadata TSV/CSV as a numerical data matrix</a></p></li>
<li><p><a class="reference internal" href="#metadata-without-column-names" id="id24">Metadata without column names</a></p></li>
<li><p><a class="reference internal" href="#missing-value-detection" id="id25">Missing value detection</a></p></li>
<li><p><a class="reference internal" href="#unmapped-tree-leaf-property-detection" id="id26">Unmapped tree leaf property detection</a></p></li>
<li><p><a class="reference internal" href="#handling-duplicated-leaf-names-in-metadata" id="id27">Handling duplicated leaf names in metadata</a></p></li>
<li><p><a class="reference internal" href="#specifying-data-types-for-metadata-columns" id="id28">Specifying data types for metadata columns</a></p></li>
<li><p><a class="reference internal" href="#annotate-metadata-to-internal-nodes" id="id29">Annotate metadata to internal nodes</a></p>
<ul>
<li><p><a class="reference internal" href="#customizing-summary-methods-for-different-columns" id="id30">Customizing summary methods for different columns</a></p></li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference internal" href="#taxonomic-annotation" id="id31">Taxonomic annotation</a></p>
<ul>
<li><p><a class="reference internal" href="#using-different-taxonomic-databases-from-gtdb-ncbi" id="id32">Using different taxonomic databases from GTDB/NCBI</a></p></li>
<li><p><a class="reference internal" href="#identifying-taxon-names-in-different-metadata-fields-columns" id="id33">Identifying taxon names in different metadata fields/columns</a></p></li>
<li><p><a class="reference internal" href="#ignoring-unclassified-taxonomic-annotations" id="id34">Ignoring unclassified taxonomic annotations</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#functional-annotation" id="id35">Functional annotation</a></p>
<ul>
<li><p><a class="reference internal" href="#annotating-multiple-sequence-alignments-msa" id="id36">Annotating Multiple Sequence Alignments(MSA)</a></p></li>
<li><p><a class="reference internal" href="#annotating-eggnog-mapper-output-and-domain-information" id="id37">Annotating EggNOG-Mapper output and Domain information</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#analytical-methods-in-treeprofiler" id="id38">Analytical methods in treeProfiler</a></p>
<ul>
<li><p><a class="reference internal" href="#ancestral-character-reconstruction" id="id39">Ancestral character reconstruction</a></p></li>
<li><p><a class="reference internal" href="#phylogenetic-signal-delta-statistic" id="id40">Phylogenetic signal delta statistic</a></p></li>
<li><p><a class="reference internal" href="#acr-for-contiunous-data" id="id41">ACR for contiunous data</a></p></li>
<li><p><a class="reference internal" href="#lineage-specificity-analysis" id="id42">Lineage specificity analysis</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#output-formats-for-annotated-trees" id="id43">Output formats for annotated trees</a></p>
<ul>
<li><p><a class="reference internal" href="#use-standard-input-and-output-formats" id="id44">Use standard input and output formats</a></p></li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference internal" href="#treeprofiler-plot-visualizing-phylogenetic-profiles-and-annotation" id="id45"><code class="docutils literal notranslate"><span class="pre">treeprofiler-plot</span></code> visualizing phylogenetic profiles and annotation</a></p>
<ul>
<li><p><a class="reference internal" href="#interactive-visualization-interface" id="id46">Interactive visualization interface</a></p></li>
<li><p><a class="reference internal" href="#basic-options-of-visualizing-layouts" id="id47">Basic options of visualizing layouts</a></p></li>
<li><p><a class="reference internal" href="#layouts-for-categorical-data" id="id48">Layouts for categorical data</a></p>
<ul>
<li><p><a class="reference internal" href="#lable-layout" id="id49">Lable Layout</a></p></li>
<li><p><a class="reference internal" href="#rectangle-layout" id="id50">Rectangle Layout</a></p></li>
<li><p><a class="reference internal" href="#colorbranch-layout" id="id51">ColorBranch Layout</a></p></li>
<li><p><a class="reference internal" href="#bubble-layout" id="id52">Bubble Layout</a></p></li>
<li><p><a class="reference internal" href="#background-layout" id="id53">Background Layout</a></p></li>
<li><p><a class="reference internal" href="#piechart-layout" id="id54">Piechart Layout</a></p></li>
<li><p><a class="reference internal" href="#profiling-layout" id="id55">Profiling Layout</a></p></li>
<li><p><a class="reference internal" href="#categorical-matrix-layout" id="id56">Categorical Matrix Layout</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#layouts-for-boolean-data" id="id57">Layouts for boolean data</a></p>
<ul>
<li><p><a class="reference internal" href="#binary-layout" id="id58">Binary Layout</a></p></li>
<li><p><a class="reference internal" href="#binary-aggregate-layout" id="id59">Binary Aggregate Layout</a></p></li>
<li><p><a class="reference internal" href="#binary-unicolor-layout" id="id60">Binary Unicolor Layout</a></p></li>
<li><p><a class="reference internal" href="#binary-unicolor-aggregate-layout" id="id61">Binary Unicolor Aggregate Layout</a></p></li>
<li><p><a class="reference internal" href="#binary-matrix-layout" id="id62">Binary Matrix Layout</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#layouts-for-numerical-data" id="id63">Layouts for Numerical data</a></p>
<ul>
<li><p><a class="reference internal" href="#id1" id="id64">ColorBranch Layout</a></p></li>
<li><p><a class="reference internal" href="#id2" id="id65">Bubble Layout</a></p></li>
<li><p><a class="reference internal" href="#heatmap-layout" id="id66">Heatmap Layout</a></p></li>
<li><p><a class="reference internal" href="#barplot-layout" id="id67">Barplot Layout</a></p></li>
<li><p><a class="reference internal" href="#numerical-matrix-layout" id="id68">Numerical Matrix Layout</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#layouts-for-analytical-data" id="id69">Layouts for Analytical Data</a></p>
<ul>
<li><p><a class="reference internal" href="#ancestral-character-reconstruction-and-phylogenetic-signal-layout" id="id70">Ancestral Character Reconstruction and Phylogenetic Signal Layout</a></p></li>
<li><p><a class="reference internal" href="#lineage-specificity-layout" id="id71">Lineage Specificity Layout</a></p></li>
<li><p><a class="reference internal" href="#ancestral-character-reconstruction-for-continuous-traits" id="id72">Ancestral Character Reconstruction for continuous traits</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#layouts-for-multiple-sequence-alignment" id="id73">Layouts for Multiple Sequence Alignment</a></p></li>
<li><p><a class="reference internal" href="#layouts-for-eggnog-mapper-pfam-smart-annotations" id="id74">Layouts for eggnog-mapper pfam/smart annotations</a></p></li>
<li><p><a class="reference internal" href="#layouts-for-eggnog-mapper-functional-annotations" id="id75">Layouts for eggnog-mapper functional annotations</a></p></li>
<li><p><a class="reference internal" href="#layouts-for-taxonomy-annotation" id="id76">Layouts for Taxonomy annotation</a></p>
<ul>
<li><p><a class="reference internal" href="#taxonclade-layout" id="id77">TaxonClade Layout</a></p></li>
<li><p><a class="reference internal" href="#taxonrectangle-layout" id="id78">TaxonRectangle Layout</a></p></li>
<li><p><a class="reference internal" href="#taxoncollapse-layout" id="id79">TaxonCollapse Layout</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#customize-color-in-layout-with-color-config" id="id80">Customize color in layout with color config</a></p>
<ul>
<li><p><a class="reference internal" href="#basic-setting" id="id81">Basic setting</a></p>
<ul>
<li><p><a class="reference internal" href="#customize-color-for-categorical-data" id="id82">Customize color for categorical data</a></p></li>
<li><p><a class="reference internal" href="#customize-color-for-boolean-data" id="id83">Customize color for boolean data</a></p></li>
<li><p><a class="reference internal" href="#customize-color-for-numerical-data" id="id84">Customize color for numerical data</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#advanced-setting" id="id85">Advanced Setting</a></p>
<ul>
<li><p><a class="reference internal" href="#setting-color-range-for-numerical-data" id="id86">Setting color range for numerical data</a></p></li>
<li><p><a class="reference internal" href="#combining-settings-of-value-and-color-range-for-numerical-data" id="id87">Combining settings of value and color range for numerical data</a></p></li>
<li><p><a class="reference internal" href="#wildcard-settings-of-color-range-for-numerical-matrix" id="id88">Wildcard settings of color range for numerical matrix</a></p></li>
<li><p><a class="reference internal" href="#customize-color-settings-with-complex-conditions" id="id89">Customize color settings with complex conditions</a></p></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference internal" href="#conditional-query-in-annotated-tree-can-be-used-in-both-annotate-and-plot" id="id90">Conditional query in annotated tree (can be used in both annotate and plot)</a></p>
<ul>
<li><p><a class="reference internal" href="#query-syntax" id="id91">Query Syntax</a></p>
<ul>
<li><p><a class="reference internal" href="#basic-query-syntax" id="id92">Basic Query Syntax</a></p>
<ul>
<li><p><a class="reference internal" href="#conditional-pruning" id="id93">Conditional pruning</a></p></li>
<li><p><a class="reference internal" href="#conditional-highlight" id="id94">Conditional highlight</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#query-in-internal-nodes" id="id95">Query in internal nodes</a></p></li>
<li><p><a class="reference internal" href="#and-and-or-conditions" id="id96">AND and OR conditions</a></p></li>
<li><p><a class="reference internal" href="#conditional-query-with-config-file" id="id97">Conditional query with config file</a></p>
<ul>
<li><p><a class="reference internal" href="#highlight-with-custom-config" id="id98">Highlight with custom config</a></p></li>
<li><p><a class="reference internal" href="#collapse-with-custom-config" id="id99">Collapse with custom config</a></p></li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference internal" href="#conditional-limit-based-on-taxonomic-level" id="id100">Conditional limit based on taxonomic level</a></p></li>
<li><p><a class="reference internal" href="#choose-the-metric-for-internal-nodes-in-numerical-layout" id="id101">Choose the metric for internal nodes in numerical layout</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#demo1-explore-gtdb-taxonomic-tree-with-metadata-and-habitat-information-of-progenome3" id="id102">Demo1 Explore GTDB taxonomic tree with metadata and habitat information of progenome3</a></p></li>
<li><p><a class="reference internal" href="#demo2-explore-large-nifh-gene-tree-with-functional-and-taxonomic-information" id="id103">Demo2 Explore large NifH gene tree with functional and taxonomic information</a></p></li>
</ul>
</li>
</ul>
</nav>
</section>
<section id="installation">
<h1><a class="toc-backref" href="#id7" role="doc-backlink">Installation</a><a class="headerlink" href="#installation" title="Link to this heading">¶</a></h1>
<section id="dependencies">
<h2><a class="toc-backref" href="#id8" role="doc-backlink">Dependencies</a><a class="headerlink" href="#dependencies" title="Link to this heading">¶</a></h2>
<dl class="simple">
<dt>TreeProfiler requires</dt><dd><ul class="simple">
<li><p>Python version &gt;= 3.10</p></li>
<li><p>ETE Toolkit v4</p></li>
<li><p>biopython &gt;= 1.8</p></li>
<li><p>selenium &gt;= 4.24</p></li>
<li><p>scipy &gt;= 1.8.0</p></li>
<li><p>matplotlib &gt;= 3.4</p></li>
<li><p>pymc &gt;= 5.0.0</p></li>
<li><p>aesara</p></li>
<li><p>pastml (custom)</p></li>
</ul>
</dd>
</dl>
</section>
<section id="quick-install-via-pip">
<h2><a class="toc-backref" href="#id9" role="doc-backlink">Quick install via pip</a><a class="headerlink" href="#quick-install-via-pip" title="Link to this heading">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Install ETE Toolkit v4</span>
<span class="n">pip</span> <span class="n">install</span> <span class="o">--</span><span class="n">force</span><span class="o">-</span><span class="n">reinstall</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">etetoolkit</span><span class="o">/</span><span class="n">ete</span><span class="o">/</span><span class="n">archive</span><span class="o">/</span><span class="n">ete4</span><span class="o">.</span><span class="n">zip</span>


<span class="c1"># Install TreeProfiler dependencies</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">biopython</span> <span class="n">selenium</span> <span class="n">scipy</span> <span class="n">matplotlib</span> <span class="n">pymc</span> <span class="n">aesara</span>

<span class="c1"># Install custom pastml package for ete4</span>
<span class="n">pip</span> <span class="n">install</span> <span class="s2">&quot;git+https://github.com/dengzq1234/pastml.git@pastml2ete4&quot;</span>

<span class="c1"># Install TreeProfiler tool via pip</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">TreeProfiler</span>

<span class="c1"># Or install directly from github</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">compgenomicslab</span><span class="o">/</span><span class="n">TreeProfiler</span><span class="o">/</span><span class="n">archive</span><span class="o">/</span><span class="n">main</span><span class="o">.</span><span class="n">zip</span>
<span class="c1"># or development mode for latestest update</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">git</span><span class="o">+</span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">compgenomicslab</span><span class="o">/</span><span class="n">TreeProfiler</span><span class="nd">@dev</span><span class="o">-</span><span class="n">repo</span>
</pre></div>
</div>
</section>
<section id="quick-start-with-examples-dataset">
<h2><a class="toc-backref" href="#id10" role="doc-backlink">Quick Start with examples dataset</a><a class="headerlink" href="#quick-start-with-examples-dataset" title="Link to this heading">¶</a></h2>
<p>TreeProfiler provide various example dataset for testing in <code class="docutils literal notranslate"><span class="pre">examples/</span></code> or <a class="reference external" href="https://github.com/compgenomicslab/TreeProfiler/tree/main/examples">https://github.com/compgenomicslab/TreeProfiler/tree/main/examples</a>, each directory consists a demo script <code class="docutils literal notranslate"><span class="pre">*_demo.sh</span></code> for quick starting different functions in TreeProfiler which already has annotate-plot pipeline of example data. User can fast explore different example tree with different visualizations. Here is the demonstration:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># execute demo script of example1</span>
<span class="n">cd</span> <span class="n">examples</span><span class="o">/</span><span class="n">basic_example1</span><span class="o">/</span>
<span class="n">sh</span> <span class="o">./</span><span class="n">example1_demo</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p>This script will annotate example tree with two metadata tables and visualize the tree with different properties. The output will be shown in the terminal and a local server will be launched for interactive visualization.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>  start parsing...
  Time for parse_csv to run:  0.001968860626220703
  Time for load_metadata_to_tree to run:  0.0003094673156738281
  Time for merge annotations to run:  0.05160331726074219
  Time for annotate_taxa to run:  4.76837158203125e-07
  Visualize properties categorical data random_type in rectangle_layout, numerical data sample1, sample2 in heatmap_layout and barplot_layout.
  Current trees in memory: 0
  Added tree example with id 0.
  * Serving Flask app &#39;ete4.smartview.gui.server&#39; (lazy loading)
  * Environment: production
    WARNING: This is a development server. Do not use it in a production deployment.
    Use a production WSGI server instead.
  * Debug mode: on
  * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)


As the session starts in local server http://127.0.0.1:5000, annotated tree and selected properties are visualized at the interactive session. ![treeprofiler interface](https://github.com/dengzq1234/treeprofiler_gallery/blob/main/figure1_all.png?raw=true) Here is detailed introduction of interactive session of visualization([here](#interactive-visualization-interface))
</pre></div>
</div>
<img alt="treeprofiler interface" class="align-center" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/figure1_all.png?raw=true" />
<p>Here is detailed introduction of interactive session of visualization([here](#interactive-visualization-interface))</p>
<p>Check other tutorial scripts</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># display demo script of each example</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">basic_example1</span><span class="o">/</span><span class="n">example1_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">automatic_query</span><span class="o">/</span><span class="n">highlight_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">automatic_query</span><span class="o">/</span><span class="n">collapse_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">automatic_query</span><span class="o">/</span><span class="n">prune_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">basic_example2</span><span class="o">/</span><span class="n">example2_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">taxonomy_example</span><span class="o">/</span><span class="n">ncbi</span><span class="o">/</span><span class="n">ncbi_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">taxonomy_example</span><span class="o">/</span><span class="n">gtdb</span><span class="o">/</span><span class="n">gtdb_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">pratical_example</span><span class="o">/</span><span class="n">progenome3</span><span class="o">/</span><span class="n">progenome_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">pratical_example</span><span class="o">/</span><span class="n">gtdb_r202</span><span class="o">/</span><span class="n">gtdbv202full_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">pratical_example</span><span class="o">/</span><span class="n">gtdb_r202</span><span class="o">/</span><span class="n">gtdbv202lite_demo</span><span class="o">.</span><span class="n">sh</span>
<span class="o">./</span><span class="n">examples</span><span class="o">/</span><span class="n">pratical_example</span><span class="o">/</span><span class="n">emapper</span><span class="o">/</span><span class="n">emapper_demo</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
</section>
<section id="manual-installation">
<h2><a class="toc-backref" href="#id11" role="doc-backlink">Manual installation</a><a class="headerlink" href="#manual-installation" title="Link to this heading">¶</a></h2>
<section id="install-ete-v4">
<h3><a class="toc-backref" href="#id12" role="doc-backlink">Install ETE v4</a><a class="headerlink" href="#install-ete-v4" title="Link to this heading">¶</a></h3>
<p>Quick way</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">etetoolkit</span><span class="o">/</span><span class="n">ete</span><span class="o">/</span><span class="n">archive</span><span class="o">/</span><span class="n">ete4</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
<p>For local development
To install ETE in a local directory to help with the development, you can:</p>
<ul>
<li><p>Clone this repository (git clone <a class="reference external" href="https://github.com/etetoolkit/ete.git">https://github.com/etetoolkit/ete.git</a>)</p></li>
<li><p>Install dependecies
- If you are using conda:</p>
<p><code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">conda-forge</span> <span class="pre">cython</span> <span class="pre">bottle</span> <span class="pre">brotli</span> <span class="pre">pyqt</span> <span class="pre">numpy&lt;2.0</span></code></p>
<ul class="simple">
<li><p>Otherwise, you can install them with</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">&lt;dependencies&gt;</span></code></p>
<ul class="simple">
<li><p>Build and install ete4 from the repository’s root directory:</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">-e</span> <span class="pre">.</span></code></p>
</li>
</ul>
<p>(In Linux there may be some cases where the gcc library must be installed, which can be done with <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">conda-forge</span> <span class="pre">gcc_linux-64</span></code>)</p>
</section>
<section id="install-treeprofiler">
<h3><a class="toc-backref" href="#id13" role="doc-backlink">Install TreeProfiler</a><a class="headerlink" href="#install-treeprofiler" title="Link to this heading">¶</a></h3>
<p>Install dependencies</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># install BioPython, selenium, scipy via conda</span>
<span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">biopython</span> <span class="n">selenium</span> <span class="n">scipy</span> <span class="n">matplotlib</span> <span class="n">pymc</span>

<span class="c1"># or pip</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">biopython</span> <span class="n">selenium</span> <span class="n">scipy</span> <span class="n">matplotlib</span>
</pre></div>
</div>
<p>Install TreeProfiler</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># install TreeProfiler</span>
<span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">compgenomicslab</span><span class="o">/</span><span class="n">TreeProfiler</span>
<span class="n">cd</span> <span class="n">TreeProfiler</span><span class="o">/</span>
<span class="n">python</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">install</span>
</pre></div>
</div>
<p>or install inrectly from github</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># install directly</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">compgenomicslab</span><span class="o">/</span><span class="n">TreeProfiler</span><span class="o">/</span><span class="n">archive</span><span class="o">/</span><span class="n">main</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
</section>
</section>
<section id="input-files">
<h2><a class="toc-backref" href="#id14" role="doc-backlink">Input files</a><a class="headerlink" href="#input-files" title="Link to this heading">¶</a></h2>
<p>TreeProfiler takes following file types as input</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Input</p></th>
<th class="head"><p>Filetype</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Tree</p></td>
<td><p>newick, ete</p></td>
</tr>
<tr class="row-odd"><td><p>Metadata</p></td>
<td><p>tar.gz, tsv</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p>ete format is a novel format developed to solve the situation we encounter in the previous step, annotated tree can be recover easily with all the annotated data without changing the data type. Besides, the ete format optimized the tree file size after mapped with its associated data. Hence it’s very handy for programers in their own script. At this moment we can only view the ete format in treeprofiler, but we will make the ete format more universal to other phylogenetic software.</p></li>
<li><p>Metadata input could be single or multiple files, either tar.gz compressed file(s) which contains multiple .tsv or plain .tsv file(s).</p></li>
</ul>
</section>
<section id="basic-usage">
<h2><a class="toc-backref" href="#id15" role="doc-backlink">Basic Usage</a><a class="headerlink" href="#basic-usage" title="Link to this heading">¶</a></h2>
<dl class="simple">
<dt>TreeProfiler has two main subcommand:</dt><dd><ul class="simple">
<li><p>annotate</p></li>
<li><p>plot</p></li>
</ul>
</dd>
</dl>
<p>The first one <code class="docutils literal notranslate"><span class="pre">annotate</span></code> is used to annotate your input tree and corresponding metadata, TreeProfiler will map all the metadata into corresponding tree node. In this step, annotated tree will be generated in newick and ete format</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">annotate</span> <span class="o">--</span><span class="n">tree</span> <span class="n">tree</span><span class="o">.</span><span class="n">nw</span> <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">newick</span> <span class="o">--</span><span class="n">metadata</span> <span class="n">metadata</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">outdir</span> <span class="o">./</span>
</pre></div>
</div>
<p>The second subcommand <code class="docutils literal notranslate"><span class="pre">plot</span></code> is used to visualize tree with associated metadata. By default, treeprofiler will launch an interactive session at localhost for user to explore input tree.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot tree with newick format</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> <span class="o">--</span><span class="n">tree</span> <span class="n">tree_annotated</span><span class="o">.</span><span class="n">nw</span> <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">newick</span>

<span class="c1"># plot tree with ete format</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> <span class="o">--</span><span class="n">tree</span> <span class="n">tree_annotated</span><span class="o">.</span><span class="n">ete</span> <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span>
</pre></div>
</div>
</section>
</section>
<section id="using-treeprofiler">
<h1><a class="toc-backref" href="#id16" role="doc-backlink">Using TreeProfiler</a><a class="headerlink" href="#using-treeprofiler" title="Link to this heading">¶</a></h1>
<p>In this Tutorial we will use TreeProfiler and demostrate basic usage with data in <code class="docutils literal notranslate"><span class="pre">examples/</span></code></p>
<section id="reading-input-tree">
<h2><a class="toc-backref" href="#id17" role="doc-backlink">Reading input tree</a><a class="headerlink" href="#reading-input-tree" title="Link to this heading">¶</a></h2>
<section id="tree-format">
<h3><a class="toc-backref" href="#id18" role="doc-backlink">Tree format</a><a class="headerlink" href="#tree-format" title="Link to this heading">¶</a></h3>
<p>TreeProfiler accpept input tree in <code class="docutils literal notranslate"><span class="pre">.nw</span></code> or <code class="docutils literal notranslate"><span class="pre">.ete</span></code> by putting <code class="docutils literal notranslate"><span class="pre">--input-type</span> <span class="pre">{newick,ete}</span></code> flag to identify. By default, TreeProfiler will automatically detech the format of tree. The difference between <code class="docutils literal notranslate"><span class="pre">.nw</span></code> and <code class="docutils literal notranslate"><span class="pre">.ete</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">newick</span></code> format is more universal and be able to used in different other phylogenetic software although associated data of tree nodes will be considered as plain text.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ete</span></code> format is a novel format developed to solve the situation we encounter in the previous step, annotated tree can be <strong>recover easily with all the annotated data without changing the data type</strong>. Besides, the ete format optimized the tree file size after mapped with its associated data. Hence it’s very handy for programers in their own script. At this moment we can only view the ete format in treeprofiler, but we will make the ete format more universal to other phylogenetic software. <strong>Hence using ete format in ``plot`` subcommand is highly reccomended</strong></p></li>
</ul>
</div></blockquote>
</section>
<section id="tree-parser">
<h3><a class="toc-backref" href="#id19" role="doc-backlink">Tree parser</a><a class="headerlink" href="#tree-parser" title="Link to this heading">¶</a></h3>
<p>TreeProfiler provides argument <code class="docutils literal notranslate"><span class="pre">--internal</span> <span class="pre">{name,support}</span></code> to specify <code class="docutils literal notranslate"><span class="pre">newick</span></code> tree when it include values in internal node. <code class="docutils literal notranslate"><span class="pre">[default:</span> <span class="pre">name]</span></code></p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>newick</p></th>
<th class="head"><p>leaves</p></th>
<th class="head"><p>internal_node value</p></th>
<th class="head"><p>internal_parser</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>(A:0.5, B:0.5)Internal_C:0.5;</p></td>
<td><p>A, B</p></td>
<td><p>Internal_C</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">name</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>(A:0.5, B:0.5)0.99:0.5;</p></td>
<td><p>A, B</p></td>
<td><p>0.99</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">support</span></code></p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="treeprofiler-annotate-computing-phylogenetic-profiles-and-annotation">
<h2><a class="toc-backref" href="#id20" role="doc-backlink"><code class="docutils literal notranslate"><span class="pre">treeprofiler-annotate</span></code> computing phylogenetic profiles and annotation</a><a class="headerlink" href="#treeprofiler-annotate-computing-phylogenetic-profiles-and-annotation" title="Link to this heading">¶</a></h2>
<p>TreeProfiler <code class="docutils literal notranslate"><span class="pre">annotate</span></code> subcommand is the step that annotate input metadata to target tree. As a result, itwill generate the following output file:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated.nw</em>, newick format with annotated tree</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated.ete</em>, ete format with annotated tree</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated_prop2type.txt</em>, config file where store the datatype of each annotated properties</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated.tsv</em>,  metadata in tab-sarated values format with annotated and summarized internal nodes information.</p></li>
</ol>
<section id="annotate-metadata-into-tree">
<h3><a class="toc-backref" href="#id21" role="doc-backlink">Annotate metadata into tree</a><a class="headerlink" href="#annotate-metadata-into-tree" title="Link to this heading">¶</a></h3>
<p>In the following sub session we will describe the usage of following arguments in <code class="docutils literal notranslate"><span class="pre">annotate</span></code> step for metadata:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-m,</span> <span class="pre">--metadata</span> <span class="pre">METADATA</span> <span class="pre">[METADATA</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;metadata.csv&gt; .csv, .tsv filename</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-s,</span> <span class="pre">--metadata-sep</span> <span class="pre">METADATA_SEP</span></code></p></td>
<td><p>Column separator of metadata table <code class="docutils literal notranslate"><span class="pre">[default:</span> <span class="pre">\t]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--data-matrix</span> <span class="pre">DATA_MATRIX</span> <span class="pre">[DATA_MATRIX</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;datamatrix.csv&gt; .csv, .tsv. Numerical matrix data metadata table as array to tree, please do not provide column headers in this file, filename will become the property name in the tree.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--no-headers</span></code></p></td>
<td><p>Metadata table doesn’t contain columns name, namespace <code class="docutils literal notranslate"><span class="pre">col``+``index</span></code> will be assigned as the key of property such as <code class="docutils literal notranslate"><span class="pre">col1</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--duplicate</span></code></p></td>
<td><p>Treeprofiler will aggregate duplicated metadata to a list as a property if metadata contains duplicated row.</p></td>
</tr>
</tbody>
</table>
<section id="basic-metadata-in-tsv-csv-format">
<h4><a class="toc-backref" href="#id22" role="doc-backlink">Basic metadata in TSV/CSV format</a><a class="headerlink" href="#basic-metadata-in-tsv-csv-format" title="Link to this heading">¶</a></h4>
<p>TreeProfiler allows users to input metadata in tsv/csv file by setting <code class="docutils literal notranslate"><span class="pre">--metadata</span> <span class="pre">&lt;filename.tsv|.csv&gt;</span></code>  and <code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">&lt;seperator&gt;</span></code>. By default, the first column of metadata should be names of target tree leaves and metadata should contain column names for each column of metadata.</p>
<p>For annotating more than one metadata inputs to tree such as <code class="docutils literal notranslate"><span class="pre">--metadata</span> <span class="pre">table1.tsv</span> <span class="pre">table2.tsv</span></code>.</p>
<p>Check metadata</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cd examples/basic_example0/
tree ./
./
├── boolean.tsv
├── categorical_duplicated.tsv
├── categorical.tsv
├── data.array
├── demo1.tree
├── numerical.tsv
└── show_tree_props.py

# check metadata structure
head categorical.tsv
name,categorical1
Taxa_0,A
Taxa_1,B
Taxa_2,B
Taxa_3,C
</pre></div>
</div>
<p>Run <code class="docutils literal notranslate"><span class="pre">annotate</span></code> subcommand</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## annotate tree with more than one metadata tsv, seperated by ``,``</span>
<span class="c1"># set the correct filename and seperator</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">.</span>
</pre></div>
</div>
<p>After annotation, treeprofiler will generate annotated tree</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ls</span> <span class="n">demo1</span><span class="o">*</span>
<span class="n">demo1_annotated</span><span class="o">.</span><span class="n">ete</span>  <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>  <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">tsv</span>  <span class="n">demo1_prop2type</span><span class="o">.</span><span class="n">txt</span>  <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span>
</pre></div>
</div>
<p>Now we can check annotated tree</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># show tree&#39;s properties</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>

<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;categorical1_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;A--1||B--2||C--2&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span> <span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span> <span class="s1">&#39;categorical1&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">}</span>
</pre></div>
</div>
</section>
<section id="metadata-tsv-csv-as-a-numerical-data-matrix">
<h4><a class="toc-backref" href="#id23" role="doc-backlink">Metadata TSV/CSV as a numerical data matrix</a><a class="headerlink" href="#metadata-tsv-csv-as-a-numerical-data-matrix" title="Link to this heading">¶</a></h4>
<p>treeprofiler can handle the whole tsv/csv file as one property and annotate it to related leaves, by using <code class="docutils literal notranslate"><span class="pre">--data-matrix</span> <span class="pre">&lt;filename.tsv|.csv&gt;</span></code> It must be numerical data matrix and without headers. Once annotated the property of data-matrix will be named by the filename (see example below)</p>
<p>The difference between <code class="docutils literal notranslate"><span class="pre">--data-matrix</span></code> and <code class="docutils literal notranslate"><span class="pre">--metadata</span></code> is that the former sees the whole metadata file as a node property and stores the rows as an array in leaf nodes, and the latter sees each column from metadata as each single property of leaf nodes.</p>
<p>Using data array file <code class="docutils literal notranslate"><span class="pre">data.array</span></code> from the previous example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># annotated data.array file to tree</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">matrix</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">.</span>

<span class="c1"># data.array is stored as one property in tree node and value is stored as array</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;data.array_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;1.0244|-0.667|-1.7740000000000002|-0.8620000000000001|-0.6552&#39;</span><span class="p">,</span>
<span class="s1">&#39;data.array_max&#39;</span><span class="p">:</span> <span class="s1">&#39;3.671|1.937|4.362|1.585|2.746&#39;</span><span class="p">,</span>
<span class="s1">&#39;data.array_min&#39;</span><span class="p">:</span> <span class="s1">&#39;-2.591|-2.356|-4.825|-3.326|-2.479&#39;</span><span class="p">,</span>
<span class="s1">&#39;data.array_std&#39;</span><span class="p">:</span> <span class="s1">&#39;2.3121192529798287|1.5156064132880938|3.524138873540599|1.9937640783202009|1.906460531980665&#39;</span><span class="p">,</span>
<span class="s1">&#39;data.array_sum&#39;</span><span class="p">:</span> <span class="s1">&#39;5.122|-3.335|-8.870000000000001|-4.3100000000000005|-3.276&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span>
<span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;data.array&#39;</span><span class="p">:</span> <span class="s1">&#39;-2.591|1.937|-3.898|0.447|-1.349&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="metadata-without-column-names">
<h4><a class="toc-backref" href="#id24" role="doc-backlink">Metadata without column names</a><a class="headerlink" href="#metadata-without-column-names" title="Link to this heading">¶</a></h4>
<p>If metadata does not have headers, by setting <code class="docutils literal notranslate"><span class="pre">--no-headers</span></code> to set the metadata properly, therefore treeprofiler will name each column by <code class="docutils literal notranslate"><span class="pre">col</span></code> + <code class="docutils literal notranslate"><span class="pre">&lt;column</span> <span class="pre">number&gt;</span></code> as the property key in each leaf node, such as <code class="docutils literal notranslate"><span class="pre">col1</span></code>, <code class="docutils literal notranslate"><span class="pre">col2</span></code>, etc.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># data.array doesn&#39;t have headers for each column</span>
<span class="n">head</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span>
<span class="n">Taxa_0</span><span class="p">,</span><span class="o">-</span><span class="mf">2.591</span><span class="p">,</span><span class="mf">1.937</span><span class="p">,</span><span class="o">-</span><span class="mf">3.898</span><span class="p">,</span><span class="mf">0.447</span><span class="p">,</span><span class="o">-</span><span class="mf">1.349</span>
<span class="n">Taxa_1</span><span class="p">,</span><span class="mf">3.366</span><span class="p">,</span><span class="o">-</span><span class="mf">1.871</span><span class="p">,</span><span class="mf">4.362</span><span class="p">,</span><span class="mf">1.585</span><span class="p">,</span><span class="o">-</span><span class="mf">2.479</span>
<span class="n">Taxa_2</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mf">0.098</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mf">3.326</span><span class="p">,</span><span class="mf">2.746</span>
<span class="n">Taxa_3</span><span class="p">,</span><span class="mf">3.671</span><span class="p">,</span><span class="o">-</span><span class="mf">0.947</span><span class="p">,</span><span class="o">-</span><span class="mf">4.509</span><span class="p">,</span><span class="o">-</span><span class="mf">3.131</span><span class="p">,</span><span class="o">-</span><span class="mf">2.194</span>

<span class="c1"># need to add --no-headers flag to tell treeprofiler</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">headers</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">.</span>

<span class="c1"># check properties</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;col1_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;1.92825&#39;</span><span class="p">,</span>
<span class="s1">&#39;col1_max&#39;</span><span class="p">:</span> <span class="s1">&#39;3.671&#39;</span><span class="p">,</span>
<span class="s1">&#39;col1_min&#39;</span><span class="p">:</span> <span class="s1">&#39;0.0&#39;</span><span class="p">,</span>
<span class="s1">&#39;col1_std&#39;</span><span class="p">:</span> <span class="s1">&#39;3.463526916666666&#39;</span><span class="p">,</span>
<span class="s1">&#39;col1_sum&#39;</span><span class="p">:</span> <span class="s1">&#39;7.713&#39;</span><span class="p">,</span>
<span class="s1">&#39;col2_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;-1.318&#39;</span><span class="p">,</span>
<span class="o">...</span><span class="p">}</span>

<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;col1&#39;</span><span class="p">:</span> <span class="s1">&#39;-2.591&#39;</span><span class="p">,</span>
<span class="s1">&#39;col2&#39;</span><span class="p">:</span> <span class="s1">&#39;1.937&#39;</span><span class="p">,</span>
<span class="s1">&#39;col3&#39;</span><span class="p">:</span> <span class="s1">&#39;-3.898&#39;</span><span class="p">,</span>
<span class="s1">&#39;col4&#39;</span><span class="p">:</span> <span class="s1">&#39;0.447&#39;</span><span class="p">,</span>
<span class="s1">&#39;col5&#39;</span><span class="p">:</span> <span class="s1">&#39;-1.349&#39;</span><span class="p">}</span>
</pre></div>
</div>
</section>
<section id="missing-value-detection">
<h4><a class="toc-backref" href="#id25" role="doc-backlink">Missing value detection</a><a class="headerlink" href="#missing-value-detection" title="Link to this heading">¶</a></h4>
<p>Metadata column which fullfills one of the following criterias will be consider as missing value:</p>
<ul class="simple">
<li><p>Entirely symbolic characters. Such as <code class="docutils literal notranslate"><span class="pre">+</span></code>, <code class="docutils literal notranslate"><span class="pre">-</span></code>, <code class="docutils literal notranslate"><span class="pre">~</span></code>, <code class="docutils literal notranslate"><span class="pre">.</span></code>, etc.</p></li>
<li><p>The exact strings <code class="docutils literal notranslate"><span class="pre">none</span></code>, <code class="docutils literal notranslate"><span class="pre">None</span></code>, <code class="docutils literal notranslate"><span class="pre">null</span></code>, <code class="docutils literal notranslate"><span class="pre">Null</span></code>, or <code class="docutils literal notranslate"><span class="pre">NaN</span></code>.</p></li>
<li><p>An empty string (zero characters).</p></li>
</ul>
<p>Missing value will replaced by string ‘NaN’ in the corresponding property.</p>
</section>
<section id="unmapped-tree-leaf-property-detection">
<h4><a class="toc-backref" href="#id26" role="doc-backlink">Unmapped tree leaf property detection</a><a class="headerlink" href="#unmapped-tree-leaf-property-detection" title="Link to this heading">¶</a></h4>
<p>If Metadata doesn’t cover input tree leaf, tree leaf will be unannotated.</p>
</section>
<section id="handling-duplicated-leaf-names-in-metadata">
<h4><a class="toc-backref" href="#id27" role="doc-backlink">Handling duplicated leaf names in metadata</a><a class="headerlink" href="#handling-duplicated-leaf-names-in-metadata" title="Link to this heading">¶</a></h4>
<p>In general, treeprofiler expects each row of metadata corresponding to one leaf, such as</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span>
<span class="c1">#name,categorical1</span>
<span class="n">Taxa_0</span><span class="p">,</span><span class="n">A</span>
<span class="n">Taxa_1</span><span class="p">,</span><span class="n">B</span>
<span class="n">Taxa_2</span><span class="p">,</span><span class="n">B</span>
<span class="n">Taxa_3</span><span class="p">,</span><span class="n">C</span>
<span class="n">Taxa_4</span><span class="p">,</span><span class="n">C</span>
</pre></div>
</div>
<p>Although treeprofiler can handle metadata with rows with duplicated leafnames such as</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">categorical_duplicated</span><span class="o">.</span><span class="n">tsv</span>
<span class="c1">#name,categorical1</span>
<span class="n">Taxa_0</span><span class="p">,</span><span class="n">A</span>
<span class="n">Taxa_0</span><span class="p">,</span><span class="n">B</span>
<span class="n">Taxa_2</span><span class="p">,</span><span class="n">B</span>
<span class="n">Taxa_2</span><span class="p">,</span><span class="n">C</span>
<span class="n">Taxa_3</span><span class="p">,</span><span class="n">C</span>
<span class="n">Taxa_3</span><span class="p">,</span><span class="n">A</span>
<span class="n">Taxa_4</span><span class="p">,</span><span class="n">C</span>
</pre></div>
</div>
<p>In order to do so, users need to add <code class="docutils literal notranslate"><span class="pre">--duplicate</span></code> , by doing so, metadata from the same leaf will be aggregate into the same column. Such as the Taxa_0 from the above table, at the end value <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> will be both annotated to property <code class="docutils literal notranslate"><span class="pre">categorical1</span></code> (see above demo).
<strong>If not, treeprofielr will take one the first row of metadata that appear as the metadata for related leaf!</strong></p>
<p>example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">-</span><span class="n">m</span> <span class="n">categorical_duplicated</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">duplicate</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">.</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;categorical1_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;A--2||B--2||C--3&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span> <span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span> <span class="s1">&#39;categorical1&#39;</span><span class="p">:</span> <span class="s1">&#39;A|B&#39;</span><span class="p">}</span>
</pre></div>
</div>
</section>
<section id="specifying-data-types-for-metadata-columns">
<h4><a class="toc-backref" href="#id28" role="doc-backlink">Specifying data types for metadata columns</a><a class="headerlink" href="#specifying-data-types-for-metadata-columns" title="Link to this heading">¶</a></h4>
<p>Although TreeProfiler can automatically detect datatype of each column, users still can determine the datatype using the following arguments using:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--text-prop</span> <span class="pre">TEXT_PROP</span> <span class="pre">[TEXT_PROP</span> <span class="pre">...]</span></code></p></td>
<td><p>names of columns which need to be read as categorical data</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--multiple-text-prop</span> <span class="pre">MULTIPLE_TEXT_PROP</span> <span class="pre">[MULTIPLE_TEXT_PROP</span> <span class="pre">...]</span></code></p></td>
<td><p>names of columns which need to be read as categorical data containing more than one value and separated by <cite>,</cite> such as <a class="reference external" href="GO:0000003,GO:0000902,GO:0000904,GO:0003006">GO:0000003,GO:0000902,GO:0000904,GO:0003006</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--num-prop</span> <span class="pre">NUM_PROP</span> <span class="pre">[NUM_PROP</span> <span class="pre">...]</span></code></p></td>
<td><p>names of columns which need to be read as numerical data</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--bool-prop</span> <span class="pre">BOOL_PROP</span> <span class="pre">[BOOL_PROP</span> <span class="pre">...]</span></code></p></td>
<td><p>names of columns which need to be read as boolean data</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--text-prop-idx</span> <span class="pre">TEXT_PROP_IDX</span> <span class="pre">[TEXT_PROP_IDX</span> <span class="pre">...]</span></code></p></td>
<td><p>1 2 3 or [1-5] index of columns which need to be read as categorical data</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--num-prop-idx</span> <span class="pre">NUM_PROP_IDX</span> <span class="pre">[NUM_PROP_IDX</span> <span class="pre">...]</span></code></p></td>
<td><p>1 2 3 or [1-5] index columns which need to be read as numerical data</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--bool-prop-idx</span> <span class="pre">BOOL_PROP_IDX</span> <span class="pre">[BOOL_PROP_IDX</span> <span class="pre">...]</span></code></p></td>
<td><p>1 2 3 or [1-5] index columns which need to be read as boolean data</p></td>
</tr>
</tbody>
</table>
</section>
<section id="annotate-metadata-to-internal-nodes">
<h4><a class="toc-backref" href="#id29" role="doc-backlink">Annotate metadata to internal nodes</a><a class="headerlink" href="#annotate-metadata-to-internal-nodes" title="Link to this heading">¶</a></h4>
<p>At the above example, we only mapped metadata to leaf nodes, in this example, we will also profile <strong>internal nodes</strong> annotation and analysis of their children nodes. Argument that in related to summary methods are:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Applied datatype</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Summarized properties Internal node</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--num-stat</span> <span class="pre">{all,sum,avg,max,min,std,none}</span></code></p></td>
<td><p>numerical data matrix</p></td>
<td><p>Descriptive Statistic (average, sum, max, min, standard deviation)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_avg</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_sum</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_max</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_min</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_std</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--counter-stat</span> <span class="pre">{raw,relative,none}</span></code></p></td>
<td><p>str
boolean
list</p></td>
<td><p>Raw/Relative Counter</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_counter</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--num-stat</span> <span class="pre">{all,sum,avg,max,min,std,none}</span></code></p></td>
<td><p>float
int</p></td>
<td><p>Descriptive Statistic (average, sum, max, min, standard deviation)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_avg</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_sum</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_max</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_min</span></code>
<code class="docutils literal notranslate"><span class="pre">&lt;prop</span> <span class="pre">name&gt;_std</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--column-summary-method</span> <span class="pre">COLUMN_SUMMARY_METHOD</span> <span class="pre">[COLUMN_SUMMARY_METHOD</span> <span class="pre">...]</span></code></p></td>
<td><p>all</p></td>
<td><p>Specify summary method for individual columns in the format ColumnName=Method, such as <code class="docutils literal notranslate"><span class="pre">--column-summary-method</span> <span class="pre">sample1=none</span> <span class="pre">sample2=avg</span> <span class="pre">random_type=relative</span> <span class="pre">alignment=none</span></code></p></td>
<td></td>
</tr>
</tbody>
</table>
<p>TreeProfiler can infer automatically the datatype of each column in your metadata, including</p>
<ul class="simple">
<li><p><cite>list</cite> (seperate by <cite>,</cite> )</p></li>
<li><p><cite>string</cite> (categorcial data)</p></li>
<li><p><cite>numerical</cite> (numerical data, float or integer)</p></li>
<li><p><cite>booleans</cite></p></li>
</ul>
<p>Internal node will summurize children nodes information according to their datatypes.</p>
<p>demo tree</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>      ╭╴A
╴root╶┤
      │   ╭╴B
      ╰╴D╶┤
          ╰╴C
</pre></div>
</div>
<p>demo metadata</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>#name</p></th>
<th class="head"><p>text_property</p></th>
<th class="head"><p>multiple_text_property</p></th>
<th class="head"><p>numerical_property</p></th>
<th class="head"><p>bool_property</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><p>vowel</p></td>
<td><p>a,b,c</p></td>
<td><p>10</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>B</p></td>
<td><p>consonant</p></td>
<td><p>b,c,d</p></td>
<td><p>4</p></td>
<td><p>False</p></td>
</tr>
<tr class="row-even"><td><p>C</p></td>
<td><p>consonant</p></td>
<td><p>c,d,e</p></td>
<td><p>9</p></td>
<td><p>True</p></td>
</tr>
</tbody>
</table>
<p>TreeProfiler will infer the datatypes of above metadata and adopt different summary methods:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><ul class="simple">
<li></li>
</ul>
</th>
<th class="head"><p>text_property</p></th>
<th class="head"><p>multiple_text_property</p></th>
<th class="head"><p>numerical_property</p></th>
<th class="head"><p>bool_property</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>datatype</p></td>
<td><p>string</p></td>
<td><p>list</p></td>
<td><p>float</p></td>
<td><p>bool</p></td>
</tr>
<tr class="row-odd"><td><p>method</p></td>
<td><p>counter</p></td>
<td><p>counter</p></td>
<td><p>average, sum, max, min, standard deviation</p></td>
<td><p>counter</p></td>
</tr>
</tbody>
</table>
<ol class="arabic simple">
<li><p>Categorical</p></li>
</ol>
<p>boolean and text properties (categorical data) of leaf nodes will be summarized as counters in internal nodes, currently users can choose using <code class="docutils literal notranslate"><span class="pre">raw</span></code> (default), <code class="docutils literal notranslate"><span class="pre">relative</span></code> or <code class="docutils literal notranslate"><span class="pre">none</span></code> for counter. Users can use <code class="docutils literal notranslate"><span class="pre">--counter-stat</span> <span class="pre">{raw,relative,none}</span></code> to choose the counter, it will automatically apply to all categorical properties.</p>
<p>After annotation, internal nodes will be summarized. If property was summarize with <code class="docutils literal notranslate"><span class="pre">counter</span></code>, in internal node will be named as <code class="docutils literal notranslate"><span class="pre">&lt;property_name&gt;_counter</span></code></p>
<p>Users can choose either counter is raw or relative count by using <code class="docutils literal notranslate"><span class="pre">--counter-stat</span></code></p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>internal_node properties</p></th>
<th class="head"><p>statistic method</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>&lt;prop name&gt;_counter</p></td>
<td><p>raw(default), relative</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>internal_node</p></th>
<th class="head"><p>text_property_counter</p></th>
<th class="head"><p>multiple_text_property_counter</p></th>
<th class="head"><p>bool_property_counter</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>D</p></td>
<td><p>consonant–2</p></td>
<td><p>b–1||c–2||d–2||e–1</p></td>
<td><p>True–1||False–1</p></td>
</tr>
<tr class="row-odd"><td><p>root</p></td>
<td><p>vowel–1||consonant–2</p></td>
<td><p>a–2||b–2||c–3||d–2||e–1</p></td>
<td><p>True–2||False–1</p></td>
</tr>
</tbody>
</table>
<p>Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># raw counter (default)</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">counter</span><span class="o">-</span><span class="n">stat</span> <span class="n">raw</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;categorical1_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;A--1||B--2||C--2&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span>
<span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;categorical1&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span>
<span class="p">}</span>

<span class="c1">#relative counter to calculate the percentage</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">counter</span><span class="o">-</span><span class="n">stat</span> <span class="n">relative</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;categorical1_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;A--0.20||B--0.40||C--0.40&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span>
<span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;categorical1&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span>
<span class="p">}</span>

<span class="c1">#set to none</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">counter</span><span class="o">-</span><span class="n">stat</span> <span class="n">none</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span> <span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span> <span class="s1">&#39;categorical1&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">}</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>Numerical</p></li>
</ol>
<p>By default, numerical feature will be calculated all the descriptive statistic, but users can choose specific one to be calculated by using <code class="docutils literal notranslate"><span class="pre">--num-stat</span> <span class="pre">{all,</span> <span class="pre">sum,</span> <span class="pre">avg,</span> <span class="pre">max,</span> <span class="pre">min,</span> <span class="pre">std,</span> <span class="pre">none}</span></code>. <code class="docutils literal notranslate"><span class="pre">all</span></code> (default) means it will conduct all the statistic. <code class="docutils literal notranslate"><span class="pre">none</span></code> means annotation will only conduct in leaf nodes.</p>
<p>If property was numerical data, in internal node will be named as</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>internal_node properties</p></th>
<th class="head"><p>statistic method</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>&lt;prop name&gt;_avg</p></td>
<td><p>average</p></td>
</tr>
<tr class="row-odd"><td><p>&lt;prop name&gt;_sum</p></td>
<td><p>sum</p></td>
</tr>
<tr class="row-even"><td><p>&lt;prop name&gt;_max</p></td>
<td><p>maximum</p></td>
</tr>
<tr class="row-odd"><td><p>&lt;prop name&gt;_min</p></td>
<td><p>minimum</p></td>
</tr>
<tr class="row-even"><td><p>&lt;prop name&gt;_std</p></td>
<td><p>standard deviation</p></td>
</tr>
</tbody>
</table>
<p>Noticed that <code class="docutils literal notranslate"><span class="pre">--num-stat</span></code> will also work on <code class="docutils literal notranslate"><span class="pre">--data-matrix</span></code> data.</p>
<p>In our demo, it would be:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>internal_node</p></th>
<th class="head"><p>numerical_property_avg</p></th>
<th class="head"><p>numerical_property_sum</p></th>
<th class="head"><p>numerical_property_max</p></th>
<th class="head"><p>numerical_property_max</p></th>
<th class="head"><p>numerical_property_max</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>D</p></td>
<td><p>6.5</p></td>
<td><p>13</p></td>
<td><p>9</p></td>
<td><p>4</p></td>
<td><p>2.5</p></td>
</tr>
<tr class="row-odd"><td><p>root</p></td>
<td><p>7.67</p></td>
<td><p>23</p></td>
<td><p>10</p></td>
<td><p>4</p></td>
<td><p>2.32</p></td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># conduct all statistic (by default)</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">numerical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">stat</span> <span class="nb">all</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;0.5384554640742852&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_max&#39;</span><span class="p">:</span> <span class="s1">&#39;0.7817176831389784&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_min&#39;</span><span class="p">:</span> <span class="s1">&#39;0.3276816717486982&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_std&#39;</span><span class="p">:</span> <span class="s1">&#39;0.028430041000376213&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_sum&#39;</span><span class="p">:</span> <span class="s1">&#39;2.692277320371426&#39;</span><span class="p">,</span>
<span class="o">....</span>
<span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;random_column1&#39;</span><span class="p">:</span> <span class="s1">&#39;0.45303222603186877&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column2&#39;</span><span class="p">:</span> <span class="s1">&#39;1.9801547427961053&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column3&#39;</span><span class="p">:</span> <span class="s1">&#39;43.0&#39;</span><span class="p">}</span>

<span class="c1"># conduct only average</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">numerical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">stat</span> <span class="n">avg</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;0.5384554640742852&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column2_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;0.12655333321138568&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column3_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;52.2&#39;</span>
<span class="p">}</span>

<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;random_column1&#39;</span><span class="p">:</span> <span class="s1">&#39;0.45303222603186877&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column2&#39;</span><span class="p">:</span> <span class="s1">&#39;1.9801547427961053&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column3&#39;</span><span class="p">:</span> <span class="s1">&#39;43.0&#39;</span><span class="p">}</span>


<span class="c1"># conduct none statistic</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">numerical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">stat</span> <span class="n">none</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;random_column1&#39;</span><span class="p">:</span> <span class="s1">&#39;0.45303222603186877&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column2&#39;</span><span class="p">:</span> <span class="s1">&#39;1.9801547427961053&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column3&#39;</span><span class="p">:</span> <span class="s1">&#39;43.0&#39;</span>
<span class="p">}</span>

<span class="c1"># data matrix is also effected by --num-stat setting</span>

<span class="c1"># only average</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">matrix</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">stat</span> <span class="n">avg</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;data.array_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;1.0244|-0.667|-1.7740000000000002|-0.8620000000000001|-0.6552&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span>
<span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;data.array&#39;</span><span class="p">:</span> <span class="s1">&#39;-2.591|1.937|-3.898|0.447|-1.349&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
<section id="customizing-summary-methods-for-different-columns">
<h5><a class="toc-backref" href="#id30" role="doc-backlink">Customizing summary methods for different columns</a><a class="headerlink" href="#customizing-summary-methods-for-different-columns" title="Link to this heading">¶</a></h5>
<p>Using <code class="docutils literal notranslate"><span class="pre">--column-summary-method</span></code>  can specify the summary method of each properties, simply add <code class="docutils literal notranslate"><span class="pre">&lt;property</span> <span class="pre">name&gt;=&lt;summary</span> <span class="pre">method&gt;</span></code> . For categorical data, options are <code class="docutils literal notranslate"><span class="pre">{raw,relative,none}</span></code>; for numerical data, options are  <code class="docutils literal notranslate"><span class="pre">{all,</span> <span class="pre">sum,</span> <span class="pre">avg,</span> <span class="pre">max,</span> <span class="pre">min,</span> <span class="pre">std,</span> <span class="pre">none}</span></code> .</p>
<p>such as <code class="docutils literal notranslate"><span class="pre">--column-summary-method</span> <span class="pre">sample1=none</span> <span class="pre">sample2=avg</span> <span class="pre">random_type=relative</span> <span class="pre">alignment=none</span></code></p>
<p>Noted that <code class="docutils literal notranslate"><span class="pre">--data-matrix</span></code> can be effected by <code class="docutils literal notranslate"><span class="pre">--column-summary-method</span></code> setting, in this case filename of the data matrix is property name, such as <code class="docutils literal notranslate"><span class="pre">--data-matrix</span> <span class="pre">file.tsv</span> <span class="pre">--column-summary-method</span> <span class="pre">file.tsv=avg</span></code></p>
<p>example, here we use three different metadata: <code class="docutils literal notranslate"><span class="pre">categorical.tsv</span></code>, <code class="docutils literal notranslate"><span class="pre">numerical.tsv</span></code> and <code class="docutils literal notranslate"><span class="pre">data</span> <span class="pre">matrix</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># cusomtize different summary methods for different column/property</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span> <span class="n">numerical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">matrix</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">column</span><span class="o">-</span><span class="n">summary</span><span class="o">-</span><span class="n">method</span> \
<span class="n">categorical1</span><span class="o">=</span><span class="n">relative</span> \
<span class="n">random_column1</span><span class="o">=</span><span class="nb">all</span> \
<span class="n">random_column2</span><span class="o">=</span><span class="n">none</span> \
<span class="n">random_column3</span><span class="o">=</span><span class="nb">sum</span> \
<span class="n">data</span><span class="o">.</span><span class="n">array</span><span class="o">=</span><span class="n">avg</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;categorical1_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;A--0.20||B--0.40||C--0.40&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;0.5384554640742852&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_max&#39;</span><span class="p">:</span> <span class="s1">&#39;0.7817176831389784&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_min&#39;</span><span class="p">:</span> <span class="s1">&#39;0.3276816717486982&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_std&#39;</span><span class="p">:</span> <span class="s1">&#39;0.028430041000376213&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1_sum&#39;</span><span class="p">:</span> <span class="s1">&#39;2.692277320371426&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column3_sum&#39;</span><span class="p">:</span> <span class="s1">&#39;261.0&#39;</span><span class="p">,</span>
<span class="s1">&#39;data.array_avg&#39;</span><span class="p">:</span> <span class="s1">&#39;1.0244|-0.667|-1.7740000000000002|-0.8620000000000001|-0.6552&#39;</span>
<span class="p">}</span>
<span class="n">target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_0&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.190563</span><span class="p">,</span>
<span class="s1">&#39;categorical1&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column1&#39;</span><span class="p">:</span> <span class="s1">&#39;0.45303222603186877&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column2&#39;</span><span class="p">:</span> <span class="s1">&#39;1.9801547427961053&#39;</span><span class="p">,</span>
<span class="s1">&#39;random_column3&#39;</span><span class="p">:</span> <span class="s1">&#39;43.0&#39;</span><span class="p">,</span>
<span class="s1">&#39;data.array&#39;</span><span class="p">:</span> <span class="s1">&#39;-2.591|1.937|-3.898|0.447|-1.349&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="taxonomic-annotation">
<h3><a class="toc-backref" href="#id31" role="doc-backlink">Taxonomic annotation</a><a class="headerlink" href="#taxonomic-annotation" title="Link to this heading">¶</a></h3>
<p>Treeprofiler annotate tree node with target taxonomy, you can either use [GTDB](<a class="reference external" href="https://gtdb.ecogenomic.org/">https://gtdb.ecogenomic.org/</a>) or [NCBI](<a class="reference external" href="https://www.ncbi.nlm.nih.gov/">https://www.ncbi.nlm.nih.gov/</a>) taxonomic database, such as following commands</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">TAXON_COLUMN</span></code></p></td>
<td><p>Choose the column in metadata which represents taxon for activating the taxonomic annotation. Default is the first column, which should be the column of leaf_name.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--taxadb</span> <span class="pre">{NCBI,GTDB,customdb}</span></code></p></td>
<td><p>NCBI or GTDB, choose the Taxonomic Database for annotation.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--taxon-delimiter</span> <span class="pre">TAXON_DELIMITER</span></code></p></td>
<td><p>Delimiter of taxa columns. <code class="docutils literal notranslate"><span class="pre">[default:</span> <span class="pre">None]</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--taxa-field</span> <span class="pre">TAXA_FIELD</span></code></p></td>
<td><p>Field of taxa name after delimiter. <code class="docutils literal notranslate"><span class="pre">[default:</span> <span class="pre">0]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--taxa-dump</span> <span class="pre">TAXA_DUMP</span></code></p></td>
<td><p>Path to taxonomic database dump file for a specific version, such as GTDB taxadump (<a class="reference external" href="https://github.com/etetoolkit/ete-data/raw/main/gtdb_taxonomy/gtdblatest/gtdb_latest_dump.tar.gz">https://github.com/etetoolkit/ete-data/raw/main/gtdb_taxonomy/gtdblatest/gtdb_latest_dump.tar.gz</a>) or NCBI taxadump (<a class="reference external" href="https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz">https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz</a>).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--gtdb-version</span> <span class="pre">{95,202,207,214,220}</span></code></p></td>
<td><p>GTDB version for taxonomic annotation, such as 220. If it is not provided, the latest version will be used.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--ignore-unclassified</span></code></p></td>
<td><p>Ignore unclassified taxa in taxonomic annotation.</p></td>
</tr>
</tbody>
</table>
<p>In this part we will demostrate the usage of taxonomic annotation in examples of <code class="docutils literal notranslate"><span class="pre">examples/taxonomy_example</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">examples</span><span class="o">/</span><span class="n">taxonomy_example</span>
<span class="n">ls</span> <span class="o">./</span>
<span class="n">demo3</span><span class="o">.</span><span class="n">tree</span>  <span class="n">demo4</span><span class="o">.</span><span class="n">tree</span>  <span class="n">gtdb202dump</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>  <span class="n">missing_gtdb_v202</span><span class="o">.</span><span class="n">tree</span>  <span class="n">ncbi</span><span class="o">.</span><span class="n">tree</span>
<span class="n">demo3</span><span class="o">.</span><span class="n">tsv</span>   <span class="n">demo4</span><span class="o">.</span><span class="n">tsv</span>   <span class="n">gtdb_v202</span><span class="o">.</span><span class="n">tree</span>      <span class="n">missing_ncbi</span><span class="o">.</span><span class="n">tree</span>       <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<section id="using-different-taxonomic-databases-from-gtdb-ncbi">
<h4><a class="toc-backref" href="#id32" role="doc-backlink">Using different taxonomic databases from GTDB/NCBI</a><a class="headerlink" href="#using-different-taxonomic-databases-from-gtdb-ncbi" title="Link to this heading">¶</a></h4>
<p>To start taxonomic annotation, using <code class="docutils literal notranslate"><span class="pre">--taxon-column</span></code> and <code class="docutils literal notranslate"><span class="pre">--taxadb</span></code> to locate where is the taxon and which taxonomic databases to be used. If taxon is leaf name, then using <code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">name</span></code>. Otherwise <code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">&lt;prop_name&gt;</span></code> which refers to the column in the metadata.</p>
<p>Examples in NCBI taxonomic database</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check example tree</span>
<span class="n">cat</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">tree</span>
<span class="p">((</span><span class="mi">9606</span><span class="p">,</span> <span class="mi">9598</span><span class="p">),</span> <span class="mi">10090</span><span class="p">);</span>

<span class="c1"># run taxonomic annotation and locate taxon column in leaf name</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">ncbi</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># check annotation results</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">ncbi_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;evoltype&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;no rank-cellular organisms|superkingdom-Eukaryota|clade-Eumetazoa|phylum-Chordata|superclass-Sarcopterygii|kingdom-Metazoa|class-Mammalia|subphylum-Craniata|superorder-Euarchontoglires&#39;</span><span class="p">,</span>
<span class="s1">&#39;lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;1|131567|2759|33154|33208|6072|33213|33511|7711|89593|7742|7776|117570|117571|8287|1338369|32523|32524|40674|32525|9347|1437010|314146&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;root|Eukaryota|Eumetazoa|Chordata|Vertebrata|Gnathostomata|Sarcopterygii|Eutheria|Tetrapoda|Amniota|Theria|Opisthokonta|Metazoa|Bilateria|Deuterostomia|Mammalia|Craniata|Teleostomi|Euteleostomi|cellular organisms|Euarchontoglires|Dipnotetrapodomorpha|Boreoeutheria&#39;</span><span class="p">,</span> <span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;superorder&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;Euarchontoglires&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;10090|9606|9598&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;314146&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;9606&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span>
<span class="s1">&#39;Homo sapiens&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;no rank-cellular organisms|superkingdom-Eukaryota|clade-Eumetazoa|phylum-Chordata|superclass-Sarcopterygii|order-Primates|parvorder-Catarrhini|family-Hominidae|genus-Homo|species-Homo sapiens|kingdom-Metazoa|class-Mammalia|subphylum-Craniata|subfamily-Homininae|superorder-Euarchontoglires|infraorder-Simiiformes|superfamily-Hominoidea|suborder-Haplorrhini&#39;</span><span class="p">,</span>
<span class="s1">&#39;lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;1|131567|2759|33154|33208|6072|33213|33511|7711|89593|7742|7776|117570|117571|8287|1338369|32523|32524|40674|32525|9347|1437010|314146|9443|376913|314293|9526|314295|9604|207598|9605|9606&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;root|Eukaryota|Eumetazoa|Chordata|Vertebrata|Gnathostomata|Sarcopterygii|Eutheria|Primates|Catarrhini|Hominidae|Homo|Homo sapiens|Tetrapoda|Amniota|Theria|Opisthokonta|Metazoa|Bilateria|Deuterostomia|Mammalia|Craniata|Teleostomi|Euteleostomi|cellular organisms|Homininae|Euarchontoglires|Simiiformes|Hominoidea|Haplorrhini|Dipnotetrapodomorpha|Boreoeutheria&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;species&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;Homo sapiens&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;9606&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;9606&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Examples in GTDB taxonomic database</p>
<p>For gtdb taxa, users can choose <code class="docutils literal notranslate"><span class="pre">--gtdb-version</span> <span class="pre">{95,202,207,214,220}</span></code> to select certain version, if not, latest gtdb db will be used.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check example tree</span>
<span class="n">cat</span> <span class="n">gtdb_v202</span><span class="o">.</span><span class="n">tree</span>
<span class="p">(</span><span class="n">GB_GCA_011358815</span><span class="mf">.1</span><span class="p">:</span><span class="mi">1</span><span class="p">,(</span><span class="n">RS_GCF_000019605</span><span class="mf">.1</span><span class="p">:</span><span class="mi">1</span><span class="p">,(</span><span class="n">RS_GCF_003948265</span><span class="mf">.1</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">GB_GCA_003344655</span><span class="mf">.1</span><span class="p">:</span><span class="mi">1</span><span class="p">):</span><span class="mf">0.5</span><span class="p">):</span><span class="mf">0.5</span><span class="p">);</span>

<span class="c1"># default using latest version, in this case on tree from version 202, it should go empty</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">gtdb_v202</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">gtdb</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">gtdb_v202_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;evoltype&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;Unknown&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;None&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;RS_GCF_000019605.1|RS_GCF_003948265.1|GB_GCA_011358815.1|GB_GCA_003344655.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;None&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;GB_GCA_011358815.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;Unknown&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;GB_GCA_011358815.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;GB_GCA_011358815.1&#39;</span>
<span class="p">}</span>

<span class="c1">#annotate tree using the proper version of GTDB</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">gtdb_v202</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">gtdb</span> \
<span class="o">--</span><span class="n">gtdb</span><span class="o">-</span><span class="n">version</span> <span class="mi">202</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># now it&#39;s correctly annotated</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">gtdb_v202_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;evoltype&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;superkingdom-d__Archaea|phylum-p__Thermoproteota|class-c__Korarchaeia|order-o__Korarchaeales|family-f__Korarchaeaceae|genus-g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;1|2|79|2172|2173|2174|2175&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;root|d__Archaea|p__Thermoproteota|c__Korarchaeia|o__Korarchaeales|f__Korarchaeaceae|g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;RS_GCF_003948265.1|GB_GCA_011358815.1|RS_GCF_000019605.1|GB_GCA_003344655.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;g__Korarchaeum&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;GB_GCA_011358815.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;superkingdom-d__Archaea|phylum-p__Thermoproteota|class-c__Korarchaeia|order-o__Korarchaeales|family-f__Korarchaeaceae|genus-g__Korarchaeum|species-s__Korarchaeum cryptofilum|subspecies-s__Korarchaeum cryptofilum&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;root|d__Archaea|p__Thermoproteota|c__Korarchaeia|o__Korarchaeales|f__Korarchaeaceae|g__Korarchaeum|s__Korarchaeum cryptofilum|GB_GCA_011358815.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;subspecies&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;s__Korarchaeum cryptofilum&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;GB_GCA_011358815.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;GB_GCA_011358815.1&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="identifying-taxon-names-in-different-metadata-fields-columns">
<h4><a class="toc-backref" href="#id33" role="doc-backlink">Identifying taxon names in different metadata fields/columns</a><a class="headerlink" href="#identifying-taxon-names-in-different-metadata-fields-columns" title="Link to this heading">¶</a></h4>
<p>When Taxon properties are embeded in different column or field in metadata, treeprofiler provides <code class="docutils literal notranslate"><span class="pre">--taxon-column</span></code>, <code class="docutils literal notranslate"><span class="pre">--taxon-delimiter</span></code> and <code class="docutils literal notranslate"><span class="pre">--taxa-field</span></code> to identify taxon term in order to process taxonomic annotation sucessfully. Here is summary of different cases with corresponding setting.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>metadata (<cite>,</cite> as column seperator)</p></th>
<th class="head"><p>taxon to be identified</p></th>
<th class="head"><p>command line setting</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">#leafname,col1</span></code>
<code class="docutils literal notranslate"><span class="pre">9598,wt</span></code></p></td>
<td><p>9598</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">name</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">#leafname,col1</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">7739.XP_002609184.1,wt</span></code></p>
</td>
<td><p>7739</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">name</span> <span class="pre">--taxon-delimiter</span> <span class="pre">.</span> <span class="pre">--taxa-field</span> <span class="pre">0</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">#leafname,ncbi_id</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">leaf_A,7739</span></code></p>
</td>
<td><p>7739</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">ncbi_id</span> <span class="pre">--taxon-delimiter</span> <span class="pre">.</span> <span class="pre">--taxa-field</span> <span class="pre">0</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">#leafname,ncbi_id</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">leaf_A,7739.XP_002609184.1</span></code></p>
</td>
<td><p>7739</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">ncbi_id</span> <span class="pre">--taxon-delimiter</span> <span class="pre">.</span> <span class="pre">--taxa-field</span> <span class="pre">0</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">#leafname,col1</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">RS_GCF_001560035.1,wt</span></code></p>
</td>
<td><p>RS_GCF_001560035.1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">default</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">#leafname,gtdb_id</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">leaf_A,d__Archaea;p__Asgardarchaeota;c__Heimdallarchaeia;o__UBA460;f__Kariarchaeaceae;g__LC-2;s__LC-2</span> <span class="pre">sp001940725</span></code></p>
</td>
<td><p>s__LC-2 sp001940725</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">--taxon-column</span> <span class="pre">gtdb_id</span> <span class="pre">--taxon-delimiter</span> <span class="pre">;</span> <span class="pre">--taxa-field</span> <span class="pre">-1</span></code></p></td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check example tree and metadata</span>
<span class="n">cat</span> <span class="n">demo3</span><span class="o">.</span><span class="n">tree</span>
<span class="p">(</span><span class="n">Taxa_2</span><span class="p">:</span><span class="mf">0.471596</span><span class="p">,((</span><span class="n">Taxa_0</span><span class="p">:</span><span class="mf">0.767844</span><span class="p">,</span><span class="n">Taxa_1</span><span class="p">:</span><span class="mf">0.792161</span><span class="p">)</span><span class="mf">0.313833</span><span class="p">:</span><span class="mf">0.684109</span><span class="p">,</span><span class="n">Taxa_3</span><span class="p">:</span><span class="mf">0.805286</span><span class="p">):</span><span class="mf">0.188666</span><span class="p">);</span>

<span class="n">cat</span> <span class="n">demo3</span><span class="o">.</span><span class="n">tsv</span>
<span class="c1">#name gtdb_taxid</span>
<span class="n">Taxa_0</span>        <span class="n">GB_GCA_011358815</span><span class="mf">.1</span><span class="nd">@sample1</span>
<span class="n">Taxa_1</span>        <span class="n">RS_GCF_000019605</span><span class="mf">.1</span><span class="nd">@sample2</span>
<span class="n">Taxa_2</span>        <span class="n">RS_GCF_003948265</span><span class="mf">.1</span><span class="nd">@sample3</span>
<span class="n">Taxa_3</span>        <span class="n">GB_GCA_003344655</span><span class="mf">.1</span><span class="nd">@sample4</span>

<span class="c1"># therefore, locate taxa id correctly</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo3</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">-</span><span class="n">m</span> <span class="n">demo3</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">gtdb_taxid</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">gtdb</span> \
<span class="o">--</span><span class="n">gtdb</span><span class="o">-</span><span class="n">version</span> <span class="mi">202</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">delimiter</span> <span class="o">@</span> \
<span class="o">--</span><span class="n">taxa</span><span class="o">-</span><span class="n">field</span> <span class="mi">0</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo3_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;evoltype&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;superkingdom-d__Archaea|phylum-p__Thermoproteota|class-c__Korarchaeia|order-o__Korarchaeales|family-f__Korarchaeaceae|genus-g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;root|d__Archaea|p__Thermoproteota|c__Korarchaeia|o__Korarchaeales|f__Korarchaeaceae|g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;genus&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_3|Taxa_0|Taxa_1|Taxa_2&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;g__Korarchaeum&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_2&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.471596</span><span class="p">,</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;gtdb_taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;RS_GCF_003948265.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;superkingdom-d__Archaea|phylum-p__Thermoproteota|class-c__Korarchaeia|order-o__Korarchaeales|family-f__Korarchaeaceae|genus-g__Korarchaeum|species-s__Korarchaeum cryptofilum|subspecies-s__Korarchaeum cryptofilum&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;root|d__Archaea|p__Thermoproteota|c__Korarchaeia|o__Korarchaeales|f__Korarchaeaceae|g__Korarchaeum|s__Korarchaeum cryptofilum|RS_GCF_003948265.1&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;subspecies&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;s__Korarchaeum cryptofilum&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_2&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;RS_GCF_003948265.1&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="ignoring-unclassified-taxonomic-annotations">
<h4><a class="toc-backref" href="#id34" role="doc-backlink">Ignoring unclassified taxonomic annotations</a><a class="headerlink" href="#ignoring-unclassified-taxonomic-annotations" title="Link to this heading">¶</a></h4>
<p>Taxonomic annotation will annotate the internal nodes based on the taxa of leaf nodes, but if leaf node has unknown taxonomic information, the internal nodes will return unknown annotation. Using <code class="docutils literal notranslate"><span class="pre">--ignore-unclassified</span></code> to ignore the unknown annotation from leaves.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check tree with unknown taxa</span>
<span class="p">(</span><span class="n">Taxa_1</span><span class="p">:</span><span class="mi">1</span><span class="p">,(</span><span class="n">RS_GCF_000019605</span><span class="mf">.1</span><span class="p">:</span><span class="mi">1</span><span class="p">,(</span><span class="n">Taxa_2</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">GB_GCA_003344655</span><span class="mf">.1</span><span class="p">:</span><span class="mi">1</span><span class="p">):</span><span class="mf">0.5</span><span class="p">):</span><span class="mf">0.5</span><span class="p">);</span>

<span class="c1"># normal way to annotate tree will cause unknown annotation</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">missing_gtdb_v202</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">gtdb</span> \
<span class="o">--</span><span class="n">gtdb</span><span class="o">-</span><span class="n">version</span> <span class="mi">202</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">missing_gtdb_v202_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;evoltype&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;Unknown&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;None&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_2|GB_GCA_003344655.1|RS_GCF_000019605.1|Taxa_1&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;None&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_1&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;Unknown&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_1&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_1&#39;</span>
<span class="p">}</span>

<span class="c1"># now adding --ignore-unclassified</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">missing_gtdb_v202</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">gtdb</span> \
<span class="o">--</span><span class="n">gtdb</span><span class="o">-</span><span class="n">version</span> <span class="mi">202</span> \
<span class="o">--</span><span class="n">ignore</span><span class="o">-</span><span class="n">unclassified</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">missing_gtdb_v202_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;evoltype&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span>
<span class="s1">&#39;lca&#39;</span><span class="p">:</span> <span class="s1">&#39;superkingdom-d__Archaea|phylum-p__Thermoproteota|class-c__Korarchaeia|order-o__Korarchaeales|family-f__Korarchaeaceae|genus-g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;root|d__Archaea|p__Thermoproteota|c__Korarchaeia|o__Korarchaeales|f__Korarchaeaceae|g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;genus&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;g__Korarchaeum&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_1|RS_GCF_000019605.1|GB_GCA_003344655.1|Taxa_2&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;g__Korarchaeum&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_1&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="s1">&#39;common_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;named_lineage&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;rank&#39;</span><span class="p">:</span> <span class="s1">&#39;Unknown&#39;</span><span class="p">,</span>
<span class="s1">&#39;sci_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
<span class="s1">&#39;species&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_1&#39;</span><span class="p">,</span>
<span class="s1">&#39;taxid&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_1&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
</section>
<section id="functional-annotation">
<h3><a class="toc-backref" href="#id35" role="doc-backlink">Functional annotation</a><a class="headerlink" href="#functional-annotation" title="Link to this heading">¶</a></h3>
<p>Treeprofiler provides module to annotate the tree leaf with functional annotation and sequence alignment file, such as eggnog-mapper output and multiple sequence alignments</p>
<p>In the following session we use example in <code class="docutils literal notranslate"><span class="pre">examples/pratical_example/emapper</span></code></p>
<section id="annotating-multiple-sequence-alignments-msa">
<h4><a class="toc-backref" href="#id36" role="doc-backlink">Annotating Multiple Sequence Alignments(MSA)</a><a class="headerlink" href="#annotating-multiple-sequence-alignments-msa" title="Link to this heading">¶</a></h4>
<p>treeprofiler can anntotate msa to tree and automatically calculate the consesus sequence in the internal node (fixed threshold 0.7), alignment will stored in nodes with property name <cite>alignment</cite>. Using <cite>–column-summary-method alignment=none</cite> can switch off the function for calculating consensus sequence for internal nodes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># annotate alignment</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> <span class="o">--</span><span class="n">tree</span> <span class="n">nifH</span><span class="o">.</span><span class="n">nw</span> <span class="o">--</span><span class="n">alignment</span> <span class="n">nifH</span><span class="o">.</span><span class="n">faa</span><span class="o">.</span><span class="n">aln</span>

<span class="c1"># mute consensus sequence</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">nifH</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">alignment</span> <span class="n">nifH</span><span class="o">.</span><span class="n">faa</span><span class="o">.</span><span class="n">aln</span> \
<span class="o">--</span><span class="n">column</span><span class="o">-</span><span class="n">summary</span><span class="o">-</span><span class="n">method</span> <span class="n">alignment</span><span class="o">=</span><span class="n">none</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>
</pre></div>
</div>
</section>
<section id="annotating-eggnog-mapper-output-and-domain-information">
<h4><a class="toc-backref" href="#id37" role="doc-backlink">Annotating EggNOG-Mapper output and Domain information</a><a class="headerlink" href="#annotating-eggnog-mapper-output-and-domain-information" title="Link to this heading">¶</a></h4>
<p>[EggNOG-mapper](<a class="reference external" href="http://eggnog-mapper.embl.de/">http://eggnog-mapper.embl.de/</a>), is a tool for fast functional annotation of novel sequences. It uses precomputed orthologous groups and phylogenies from the eggNOG database (<a class="reference external" href="http://eggnog5.embl.de">http://eggnog5.embl.de</a>) to transfer functional information from fine-grained orthologs only.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--emapper-annotations</span> <span class="pre">EMAPPER_ANNOTATIONS</span></code></p></td>
<td><p>Attach eggNOG-mapper output out.emapper.annotations</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--emapper-pfam</span> <span class="pre">EMAPPER_PFAM</span></code></p></td>
<td><p>Attach eggNOG-mapper pfam output out.emapper.pfams</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--emapper-smart</span> <span class="pre">EMAPPER_SMART</span></code></p></td>
<td><p>Attach eggNOG-mapper smart output out.emapper.smart</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--alignment</span> <span class="pre">ALIGNMENT</span></code></p></td>
<td><p>Sequence alignment, .fasta format</p></td>
</tr>
</tbody>
</table>
<p>It generates three kind of ouput file,</p>
<ol class="arabic simple">
<li><p>Raw standard output, <code class="docutils literal notranslate"><span class="pre">*.out.emapper.annotations</span></code>, that contains functional annotations and prthology predictions, for example:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Mon Feb 27 09:05:50 2023</span>
<span class="c1">## emapper-2.1.9</span>
<span class="c1">## /data/shared/home/emapper/miniconda3/envs/eggnog-mapper-2.1/bin/emapper.py --cpu 20 --mp_start_method forkserver --data_dir /dev/shm/ -o out --output_dir /emapper_web_jobs/emapper_jobs/user_data/MM_knn6rw6j --temp_dir /emapper_web_jobs/emapper_jobs/user_data/MM_knn6rw6j --override -m diamond --dmnd_ignore_warnings --dmnd_algo ctg -i /emapper_web_jobs/emapper_jobs/user_data/MM_knn6rw6j/queries.fasta --evalue 0.001 --score 60 --pident 40 --query_cover 20 --subject_cover 20 --itype proteins --tax_scope auto --target_orthologs all --go_evidence non-electronic --pfam_realign denovo --num_servers 2 --report_orthologs --decorate_gff yes --excel</span>
<span class="c1">##</span>
<span class="c1">#query      seed_ortholog   evalue  score   eggNOG_OGs      max_annot_lvl   COG_category    Description     Preferred_name  GOs     EC      KEGG_ko KEGG_Pathway    KEGG_Module     KEGG_Reaction   KEGG_rclass     BRITE   KEGG_TC CAZy    BiGG_Reaction   PFAMs</span>
<span class="o">....</span>
<span class="c1">## 272 queries scanned</span>
<span class="c1">## Total time (seconds): 45.73449420928955</span>
<span class="c1">## Rate: 5.95 q/s</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>[Pfam](<a class="reference external" href="http://pfam.xfam.org/">http://pfam.xfam.org/</a>) domain annotations, <code class="docutils literal notranslate"><span class="pre">*.out.emapper.pfam</span></code>, for example:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## Mon Feb 27 09:05:52 2023</span>
<span class="c1">## emapper-2.1.9</span>
<span class="c1">## /data/shared/home/emapper/miniconda3/envs/eggnog-mapper-2.1/bin/emapper.py --cpu 20 --mp_start_method forkserver --data_dir /dev/shm/ -o out --output_dir /emapper_web_jobs/emapper_jobs/user_data/MM_knn6rw6j --temp_dir /emapper_web_jobs/emapper_jobs/user_data/MM_knn6rw6j --override -m diamond --dmnd_ignore_warnings --dmnd_algo ctg -i /emapper_web_jobs/emapper_jobs/user_data/MM_knn6rw6j/queries.fasta --evalue 0.001 --score 60 --pident 40 --query_cover 20 --subject_cover 20 --itype proteins --tax_scope auto --target_orthologs all --go_evidence non-electronic --pfam_realign denovo --num_servers 2 --report_orthologs --decorate_gff yes --excel</span>
<span class="c1">##</span>
<span class="c1"># query_name        hit     evalue  sum_score       query_length    hmmfrom hmmto   seqfrom seqto   query_coverage</span>
<span class="o">...</span>
<span class="c1">## 272 queries scanned</span>
<span class="c1">## Total time (seconds): 28.74908423423767</span>
<span class="c1">## Rate: 9.46 q/s</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p>[SMART](<a class="reference external" href="http://smart.embl-heidelberg.de/">http://smart.embl-heidelberg.de/</a>) domain annotation, <code class="docutils literal notranslate"><span class="pre">*.out.emapper.smart.out</span></code>, for example:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mf">10020.</span><span class="n">ENSDORP00000023664</span>    <span class="n">MAGE_N</span>  <span class="mi">10</span>      <span class="mi">63</span>      <span class="mf">220000.115599899</span>
<span class="mf">10020.</span><span class="n">ENSDORP00000023664</span>    <span class="n">PTN</span>     <span class="mi">44</span>      <span class="mi">128</span>     <span class="mf">683.160049964146</span>
<span class="mf">10020.</span><span class="n">ENSDORP00000023664</span>    <span class="n">Ephrin_rec_like</span> <span class="mi">73</span>      <span class="mi">117</span>     <span class="mf">248282.169266432</span>
<span class="mf">10020.</span><span class="n">ENSDORP00000023664</span>    <span class="n">PreSET</span>  <span class="mi">87</span>      <span class="mi">186</span>     <span class="mf">494.036044144428</span>
<span class="o">....</span>
</pre></div>
</div>
<dl class="simple">
<dt>TreeProfiler allows users annotate EggNOG-mapper  standard output to target tree with following arguments</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--emapper-annotations</span></code>, attach eggNOG-mapper output <code class="docutils literal notranslate"><span class="pre">out.emapper.annotations</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--emapper-pfam</span></code>, attach eggNOG-mapper pfam output <code class="docutils literal notranslate"><span class="pre">out.emapper.pfams</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--emapper-smart</span></code>, attach eggNOG-mapper smart output <code class="docutils literal notranslate"><span class="pre">out.emapper.smart</span></code>.</p></li>
</ul>
</dd>
</dl>
<table class="docutils align-default" id="id3">
<caption><span class="caption-text">emapper annotation output and the summary method</span><a class="headerlink" href="#id3" title="Link to this table">¶</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Field</p></th>
<th class="head"><p>datatype</p></th>
<th class="head"><p>summary method</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>seed_ortholog</p></td>
<td><p>str</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>evalue</p></td>
<td><p>float</p></td>
<td><p>descriptive stat</p></td>
</tr>
<tr class="row-even"><td><p>score</p></td>
<td><p>float</p></td>
<td><p>descriptive stat</p></td>
</tr>
<tr class="row-odd"><td><p>eggNOG_OGs</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>max_annot_lvl</p></td>
<td><p>str</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>COG_category</p></td>
<td><p>str</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>Description</p></td>
<td><p>str</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>Preferred_name</p></td>
<td><p>str</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>GOs</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>EC</p></td>
<td><p>str</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>KEGG_ko</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>KEGG_Pathway</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>KEGG_Module</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>KEGG_Reaction</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>KEGG_rclass</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>BRITE</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>KEGG_TC</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>CAZy</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-even"><td><p>BiGG_Reaction</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
<tr class="row-odd"><td><p>PFAMs</p></td>
<td><p>list</p></td>
<td><p>counter</p></td>
</tr>
</tbody>
</table>
<p>[check EggNOG-mapper annotation example](#demo2-explore-eggnog-mapper-annotations-data-with-taxonomic-annotation)</p>
</section>
</section>
<section id="analytical-methods-in-treeprofiler">
<h3><a class="toc-backref" href="#id38" role="doc-backlink">Analytical methods in treeProfiler</a><a class="headerlink" href="#analytical-methods-in-treeprofiler" title="Link to this heading">¶</a></h3>
<p>we use examples in <code class="docutils literal notranslate"><span class="pre">examples/analytic_example</span></code></p>
<section id="ancestral-character-reconstruction">
<h4><a class="toc-backref" href="#id39" role="doc-backlink">Ancestral character reconstruction</a><a class="headerlink" href="#ancestral-character-reconstruction" title="Link to this heading">¶</a></h4>
<p>treeprofiler has integrated [pastml](<a class="reference external" href="https://github.com/evolbioinfo/pastml">https://github.com/evolbioinfo/pastml</a>), a flexible platform for ancestral reconstruction with tree with –reconstruct flag in annotate step such as:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--acr-discrete-columns</span> <span class="pre">ACR_DISCRETE_COLUMNS</span> <span class="pre">[ACR_DISCRETE_COLUMNS</span> <span class="pre">...]</span></code></p></td>
<td><p>names of columns to perform acr analysis for discrete traits</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--acr-continuous-columns</span> <span class="pre">ACR_CONTINUOUS_COLUMNS</span> <span class="pre">[ACR_CONTINUOUS_COLUMNS</span> <span class="pre">...]</span></code></p></td>
<td><p>names of columns to perform acr analysis for continuous traits</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--prediction-method</span> <span class="pre">{MPPA,MAP,JOINT,DOWNPASS,ACCTRAN,DELTRAN,COPY,ALL,MP,ML,BAYESIAN}</span></code></p></td>
<td><p>Prediction method for ACR analysis.
For <strong>Discrete</strong> traits: <code class="docutils literal notranslate"><span class="pre">MPPA</span></code>, <code class="docutils literal notranslate"><span class="pre">MAP</span></code>, <code class="docutils literal notranslate"><span class="pre">JOINT</span></code>, <code class="docutils literal notranslate"><span class="pre">DOWNPASS</span></code>, <code class="docutils literal notranslate"><span class="pre">ACCTRAN</span></code>, <code class="docutils literal notranslate"><span class="pre">DELTRAN</span></code>, <code class="docutils literal notranslate"><span class="pre">COPY</span></code>, <code class="docutils literal notranslate"><span class="pre">ALL</span></code>, <code class="docutils literal notranslate"><span class="pre">ML</span></code>, <code class="docutils literal notranslate"><span class="pre">MP</span></code>.
For <strong>Continuous</strong> traits: <code class="docutils literal notranslate"><span class="pre">ML</span></code>, <code class="docutils literal notranslate"><span class="pre">BAYESIAN</span></code>.
<code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">MPPA]</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--model</span> <span class="pre">{JC,F81,EFT,HKY,JTT,BM,OU}</span></code></p></td>
<td><p>Evolutionary model for ML methods in ACR analysis.
For <strong>discrete traits</strong>: <code class="docutils literal notranslate"><span class="pre">JC</span></code>, <code class="docutils literal notranslate"><span class="pre">F81</span></code>, <code class="docutils literal notranslate"><span class="pre">EFT</span></code>, <code class="docutils literal notranslate"><span class="pre">HKY</span></code>, <code class="docutils literal notranslate"><span class="pre">JTT</span></code>
For <strong>continuous traits</strong>: <code class="docutils literal notranslate"><span class="pre">BM</span></code>, <code class="docutils literal notranslate"><span class="pre">OU</span></code>.
<code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">F81]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--threads</span> <span class="pre">THREADS</span></code></p></td>
<td><p>Number of threads to use for annotation.
<code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">4]</span></code></p></td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ls</span>
<span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax</span><span class="o">.</span><span class="n">nwk</span> <span class="n">metadata_tab</span><span class="o">.</span><span class="n">csv</span>

<span class="c1"># check metadata</span>
<span class="n">head</span> <span class="n">metadata_tab</span><span class="o">.</span><span class="n">csv</span>
<span class="nb">id</span>  <span class="n">Country</span>
<span class="mi">98</span><span class="n">CMAJ6932</span>  <span class="n">Africa</span>
<span class="mi">98</span><span class="n">CMAJ6933</span>  <span class="n">Africa</span>
<span class="mi">96</span><span class="n">CMAJ6134</span>  <span class="n">Africa</span>
<span class="mi">00</span><span class="n">SEAY5240</span>  <span class="n">WestEurope</span>
<span class="mi">97</span><span class="n">CDAF6240</span>  <span class="n">Africa</span>
<span class="mi">97</span><span class="n">CDAF6238</span>  <span class="n">Africa</span>

<span class="c1"># quick running using all default setting</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax_annotated</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">internal</span><span class="o">-</span><span class="n">parser</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">discrete</span><span class="o">-</span><span class="n">columns</span> <span class="n">Country</span>  \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># check properties</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;ROOT&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span><span class="p">,</span>
<span class="s1">&#39;Country_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa--50||Albania--31||EastEurope--10||Greece--39||WestEurope--22&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="mi">97</span><span class="n">CDAF6238</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;97CDAF6238&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.08034</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span>
<span class="p">}</span>

<span class="c1"># check output files</span>
<span class="n">head</span> <span class="n">marginal_probabilities</span><span class="o">.</span><span class="n">character_Country</span><span class="o">.</span><span class="n">model_F81</span><span class="o">.</span><span class="n">tab</span>
<span class="n">node</span>        <span class="n">Africa</span>  <span class="n">Albania</span> <span class="n">EastEurope</span>      <span class="n">Greece</span>  <span class="n">WestEurope</span>
<span class="n">ROOT</span>        <span class="mf">0.9462054466377042</span>      <span class="mf">0.0019142742715016286</span>   <span class="mf">0.011256165797407233</span>    <span class="mf">0.013434856612985015</span>    <span class="mf">0.027189256680401872</span>
<span class="n">node_1</span>      <span class="mf">0.9497450729621073</span>      <span class="mf">0.00018867741670758483</span>  <span class="mf">0.00048818236055906636</span>  <span class="mf">0.001324183303131325</span>    <span class="mf">0.04825388395749479</span>
<span class="n">node_2</span>      <span class="mf">0.9752818930521312</span>      <span class="mf">0.00048506476303705997</span>  <span class="mf">0.015213913144468159</span>    <span class="mf">0.0034043477773810613</span>   <span class="mf">0.0056147812629824085</span>
<span class="n">node_3</span>      <span class="mf">0.9473989345272481</span>      <span class="mf">0.0002801019197914036</span>   <span class="mf">0.0005949760547048478</span>   <span class="mf">0.001965821926394849</span>    <span class="mf">0.04976016557186095</span>
<span class="n">node_4</span>      <span class="mf">0.9384942099527859</span>      <span class="mf">0.0002164578877048098</span>   <span class="mf">0.00043984526187224396</span>  <span class="mf">0.00151915289715353</span>     <span class="mf">0.05933033400048369</span>
<span class="mi">00</span><span class="n">CZAY4286</span>  <span class="mf">0.0</span>     <span class="mf">0.0</span>     <span class="mf">1.0</span>     <span class="mf">0.0</span>     <span class="mf">0.0</span>
<span class="n">node_5</span>      <span class="mf">0.9999517018762923</span>      <span class="mf">9.117741186968884e-07</span>   <span class="mf">3.0195194146220156e-05</span>  <span class="mf">6.458698485629717e-06</span>   <span class="mf">1.0732456957024559e-05</span>
<span class="mi">97</span><span class="n">CDAF6238</span>  <span class="mf">1.0</span>     <span class="mf">0.0</span>     <span class="mf">0.0</span>     <span class="mf">0.0</span>     <span class="mf">0.0</span>
<span class="mi">94</span><span class="n">CYAF6237</span>  <span class="mf">0.0</span>     <span class="mf">0.0</span>     <span class="mf">0.0</span>     <span class="mf">0.0</span>     <span class="mf">1.0</span>

<span class="c1"># check output files</span>
<span class="n">head</span> <span class="n">params</span><span class="o">.</span><span class="n">character_Country</span><span class="o">.</span><span class="n">method_MPPA</span><span class="o">.</span><span class="n">model_F81</span><span class="o">.</span><span class="n">tab</span>
<span class="n">parameter</span>   <span class="n">value</span>
<span class="n">pastml_version</span>      <span class="mf">1.9.42</span>
<span class="n">character</span>   <span class="n">Country</span>
<span class="n">log_likelihood</span>      <span class="o">-</span><span class="mf">118.96060539505257</span>
<span class="n">log_likelihood_restricted_JOINT</span>     <span class="o">-</span><span class="mf">123.17363108674806</span>
<span class="n">log_likelihood_restricted_MAP</span>       <span class="o">-</span><span class="mf">123.3244296265415</span>
<span class="n">log_likelihood_restricted_MPPA</span>      <span class="o">-</span><span class="mf">120.52779174042388</span>
<span class="n">num_scenarios</span>       <span class="mi">96</span>
<span class="n">num_states_per_node_avg</span>     <span class="mf">1.023102310231023</span>
<span class="n">num_unresolved_nodes</span>        <span class="mi">6</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--acr-discrete-columns</span> <span class="pre">&lt;PROP&gt;</span></code>  allow users to calculate the ancestral character state construction via pastml package. Hence the internal node will be infered the state based on the children leaf node metadata. Users can choose the prediction method using <code class="docutils literal notranslate"><span class="pre">--prediction-method</span> <span class="pre">&lt;METHOD&gt;</span></code>. It will generate the output config file from PASTML package as</p>
<p><code class="docutils literal notranslate"><span class="pre">params.character_{prop}.method_{method}.model_{model}.tab</span></code> which contains information of likelihood from different model/method.</p>
<p><strong>MAXIMUM LIKELIHOOD (ML) METHODS</strong></p>
<p>ML approaches are based on probabilistic models of character evolution along tree branches. From a theoretical standpoint, ML methods have some optimality guaranty [Zhang and Nei, 1997, Gascuel and Steel, 2014], at least in the absence of model violation. Noted that running this ML method will generate output file as <code class="docutils literal notranslate"><span class="pre">marginal_probabilities.character_{prop}.model_{model}.tab</span></code> which contain the calculated propabilities of each character in every internal nodes. Instead <strong>MP method</strong> won’t generate it because it doesn’t compute the marginal propabilities</p>
<p>We provide three ML methods: maximum a posteriori (MAP), Joint, and marginal posterior probabilities approximation (MPPA, recommended):</p>
<ul class="simple">
<li><p><strong>MAP</strong></p></li>
</ul>
<p>(maximum a posteriori). It computes the marginal posterior probabilities of every state for each of the tree nodes, based on the information from the whole tree, i.e. tip states and branch lengths (obtained via two tree traversals: bottom-up, and then top-down). MAP then chooses a state with the highest posterior probability for each node, independently from one node to another. This could induce globally inconsistent scenarios (typically: two very close nodes with incompatible predictions).</p>
<ul class="simple">
<li><p><strong>JOINT</strong></p></li>
</ul>
<p>While MAP chooses predicted states based on all possible scenarios, Joint method [Pupko et al., 2000] reconstructs the states of the scenario with the highest likelihood.</p>
<ul class="simple">
<li><p><strong>MPPA</strong> (default)</p></li>
</ul>
<p>MAP and Joint methods choose one state per node and do not reflect the fact that with real data and large trees, billions of scenarios may have similar posterior probabilities. Based on the marginal posterior probabilities, MPPA (marginal posterior probabilities approximation) chooses for every node a subset of likely states that minimizes the prediction error measured by the Brier score. It therefore sometimes keeps multiple state predictions per node but only when they have similar and high probabilities. Note however that the states not kept by MPPA might still be significant despite being less probable – to check marginal probabilities of each state on a node consult the output marginal probabilities file (can be downloaded via the button below each compressed visualisation).</p>
<ul class="simple">
<li><p><strong>ML</strong></p></li>
</ul>
<p>All the ML methods for ML</p>
<p><strong>Character evolution models (only in ML methods)</strong></p>
<p>We provide some models of character evolution that differ in the way the equilibrium frequencies of states are calculated: <code class="docutils literal notranslate"><span class="pre">JC</span></code>, <code class="docutils literal notranslate"><span class="pre">F81</span></code> <strong>(recommended)</strong>, and <code class="docutils literal notranslate"><span class="pre">EFT</span></code> (estimate-from-tips, <em>not recommended</em>). Using <code class="docutils literal notranslate"><span class="pre">--prediction-method</span> <span class="pre">&lt;model&gt;</span></code> to set up.</p>
<ul class="simple">
<li><p><strong>JC</strong></p></li>
</ul>
<p>With JC model [Jukes and Cantor, 1969] all frequencies, and therefore rates of changes from state i to state j (i ≠ j) are equal.</p>
<ul class="simple">
<li><p><strong>F81</strong> (recommended)</p></li>
</ul>
<p>With F81 model [Felsenstein, 1981], the rate of changes from i to j (i ≠ j) is proportional to the equilibrium frequency of j. The equilibrium frequencies are optimised.</p>
<ul class="simple">
<li><p><strong>EFT</strong></p></li>
</ul>
<p>With EFT (estimate-from-tips) model, the equilibrium frequencies are calculated based on the tip state proportions, the rate of changes from i to j (i ≠ j) is proportional to the equilibrium frequency of j.</p>
<p><strong>MAXIMUM PARSIMONY (MP) METHODS</strong></p>
<p>MP methods aim to minimize the number of state changes in the tree. They are very quick but not very accurate, e.g. they do not take into account branch lengths. We provide three MP methods: <code class="docutils literal notranslate"><span class="pre">DOWNPASS</span></code>, <code class="docutils literal notranslate"><span class="pre">ACCTRAN</span></code>, and <code class="docutils literal notranslate"><span class="pre">DELTRAN</span></code>.</p>
<ul class="simple">
<li><p><strong>DOWNPASS</strong></p></li>
</ul>
<p>DOWNPASS [Maddison and Maddison, 2003] performs two tree traversals: bottom-up and top-down, at the end of which it calculates the most parsimonious states of ancestral nodes based on the information from the whole tree. However some of the nodes might be not completely resolved due to multiple parsimonious solutions.</p>
<ul class="simple">
<li><p><strong>DELTRAN</strong></p></li>
</ul>
<p>DELTRAN (delayed transformation) [Swofford and Maddison, 1987] reduces the number of node state ambiguities by making the changes as close to the tips as possible, hence prioritizing parallel mutations.</p>
<ul class="simple">
<li><p><strong>ACCTRAN</strong></p></li>
</ul>
<p>ACCTRAN (accelerated transformation) [Farris, 1970] reduces the number of node state ambiguities by forcing the state changes to be performed as close to the root as possible, and therefore prioritises the reverse mutations.</p>
<ul class="simple">
<li><p><strong>MP</strong></p></li>
</ul>
<p>all the MP methods for MP</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># using different model</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax</span><span class="o">.</span><span class="n">nwk</span> \
<span class="o">--</span><span class="n">internal</span><span class="o">-</span><span class="n">parser</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">metadata_tab</span><span class="o">.</span><span class="n">csv</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">discrete</span><span class="o">-</span><span class="n">columns</span> <span class="n">Country</span> \
<span class="o">--</span><span class="n">prediction</span><span class="o">-</span><span class="n">method</span> <span class="n">MPPA</span> \
<span class="o">--</span><span class="n">model</span> <span class="n">JC</span> \
<span class="o">--</span><span class="n">threads</span> <span class="mi">6</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;ROOT&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span><span class="p">,</span>
<span class="s1">&#39;Country_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa--50||Albania--31||EastEurope--10||Greece--39||WestEurope--22&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;97CDAF6238&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.08034</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span>
<span class="p">}</span>

<span class="c1"># using MP methods (no calculation of ancestral propababilities)</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax</span><span class="o">.</span><span class="n">nwk</span> \
<span class="o">--</span><span class="n">internal</span><span class="o">-</span><span class="n">parser</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">metadata_tab</span><span class="o">.</span><span class="n">csv</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">discrete</span><span class="o">-</span><span class="n">columns</span> <span class="n">Country</span> \
<span class="o">--</span><span class="n">prediction</span><span class="o">-</span><span class="n">method</span> <span class="n">DOWNPASS</span> \
<span class="o">--</span><span class="n">threads</span> <span class="mi">6</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>


<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;ROOT&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span><span class="p">,</span>
<span class="s1">&#39;Country_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa--50||Albania--31||EastEurope--10||Greece--39||WestEurope--22&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;97CDAF6238&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.08034</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="phylogenetic-signal-delta-statistic">
<h4><a class="toc-backref" href="#id40" role="doc-backlink">Phylogenetic signal delta statistic</a><a class="headerlink" href="#phylogenetic-signal-delta-statistic" title="Link to this heading">¶</a></h4>
<p>Running signal delta statistic required running Ancestral Character Reconstruction using MPPA or MP methods in order to have the ancestral character propabilities. Calculated delta statistic metric and p_value of given trait will be stored in root node as properties.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--delta-stats</span></code></p></td>
<td><p>Calculate delta statistic for discrete traits in ACR analysis, ONLY for MPPA or MAP prediction method. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">False]</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--ent-type</span> <span class="pre">{LSE,SE,GINI}</span></code></p></td>
<td><p>Entropy method to measure the degree of phylogenetic signal between discrete trait and phylogeny. Options: <code class="docutils literal notranslate"><span class="pre">LSE</span></code>, <code class="docutils literal notranslate"><span class="pre">SE</span></code>, <code class="docutils literal notranslate"><span class="pre">GINI</span></code>. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">SE]</span></code> for Shannon Entropy, other options are GINI for Gini impurity and LSE for Linear Shannon Entropy.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--iteration</span> <span class="pre">ITERATION</span></code></p></td>
<td><p>Number of iterations for delta statistic calculation. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">10000]</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--lambda0</span> <span class="pre">LAMBDA0</span></code></p></td>
<td><p>Rate parameter of the delta statistic calculation. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">0.1]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--se</span> <span class="pre">SE</span></code></p></td>
<td><p>Standard deviation of the delta statistic calculation. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">0.5]</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--thin</span> <span class="pre">THIN</span></code></p></td>
<td><p>Keep only each xth iterate. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">10]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--burn</span> <span class="pre">BURN</span></code></p></td>
<td><p>Burned-in iterates. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">100]</span></code></p></td>
</tr>
</tbody>
</table>
<p>Delta statistic Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax</span><span class="o">.</span><span class="n">nwk</span> \
<span class="o">--</span><span class="n">internal</span><span class="o">-</span><span class="n">parser</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">metadata_tab</span><span class="o">.</span><span class="n">csv</span> \
<span class="c1"># acr to obtain propabilities</span>
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">discrete</span><span class="o">-</span><span class="n">columns</span> <span class="n">Country</span> \
<span class="o">--</span><span class="n">prediction</span><span class="o">-</span><span class="n">method</span> <span class="n">MPPA</span> \
<span class="o">--</span><span class="n">model</span> <span class="n">F81</span> \
<span class="c1"># delta statistic</span>
<span class="o">--</span><span class="n">delta</span><span class="o">-</span><span class="n">stats</span> \
<span class="o">--</span><span class="n">ent</span><span class="o">-</span><span class="nb">type</span> <span class="n">SE</span> \
<span class="o">--</span><span class="n">iteration</span> <span class="mi">10000</span> \
<span class="o">--</span><span class="n">lambda0</span> <span class="mf">0.1</span> \
<span class="o">--</span><span class="n">se</span> <span class="mf">0.5</span> \
<span class="o">--</span><span class="n">thin</span> <span class="mi">10</span> \
<span class="o">--</span><span class="n">burn</span> <span class="mi">100</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># delta metric and p_val stored in root node</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;ROOT&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span><span class="p">,</span>
<span class="s1">&#39;Country_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa--50||Albania--31||EastEurope--10||Greece--39||WestEurope--22&#39;</span><span class="p">,</span>
<span class="s1">&#39;Country_delta&#39;</span><span class="p">:</span> <span class="s1">&#39;19.52340888828994&#39;</span><span class="p">,</span>
<span class="s1">&#39;Country_pval&#39;</span><span class="p">:</span> <span class="s1">&#39;0.0&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;97CDAF6238&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.08034</span><span class="p">,</span>
<span class="s1">&#39;Country&#39;</span><span class="p">:</span> <span class="s1">&#39;Africa&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="acr-for-contiunous-data">
<h4><a class="toc-backref" href="#id41" role="doc-backlink">ACR for contiunous data</a><a class="headerlink" href="#acr-for-contiunous-data" title="Link to this heading">¶</a></h4>
<p>TreeProfiler supports ancestral character reconstruction for continuous traits using two main approaches: <strong>Maximum Likelihood (ML)</strong> and <strong>Bayesian</strong> inference. Both methods rely on evolutionary models for continuous data, specifically the <strong>Brownian Motion (BM)</strong> and <strong>Ornstein-Uhlenbeck (OU)</strong> models.</p>
<p>TreeProfiler allows users to select the desired method and model using the following arguments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--acr-continuous-columns</span> <span class="pre">&lt;PROP&gt;</span></code>: Specify the column names for the continuous traits.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--prediction-method</span> <span class="pre">&lt;ML/BAYESIAN&gt;</span></code>: Choose between the ML or Bayesian approach.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--model</span> <span class="pre">&lt;BM/OU&gt;</span></code>: Choose the evolutionary model for continuous trait analysis.</p></li>
</ul>
<p>Here is tree with example metadata which is continuous dataset <code class="docutils literal notranslate"><span class="pre">Anolis.tre</span></code> and <code class="docutils literal notranslate"><span class="pre">svl.csv</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">svl</span><span class="o">.</span><span class="n">csv</span>
<span class="n">species</span><span class="p">,</span><span class="n">svl</span>
<span class="n">ahli</span><span class="p">,</span><span class="mf">4.039125443</span>
<span class="n">alayoni</span><span class="p">,</span><span class="mf">3.815704818</span>
<span class="n">alfaroi</span><span class="p">,</span><span class="mf">3.526654599</span>
<span class="n">aliniger</span><span class="p">,</span><span class="mf">4.036556538</span>
<span class="n">allisoni</span><span class="p">,</span><span class="mf">4.375390078</span>

<span class="c1"># now we run the acr for the continuous trait svl, here we turn off the descriptive statistic</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Anolis</span><span class="o">.</span><span class="n">tre</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">svl</span><span class="o">.</span><span class="n">csv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">continuous</span><span class="o">-</span><span class="n">columns</span> <span class="n">svl</span> \
<span class="o">--</span><span class="n">prediction</span><span class="o">-</span><span class="n">method</span> <span class="n">ML</span> \
<span class="o">--</span><span class="n">model</span> <span class="n">BM</span> \
<span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">stat</span> <span class="n">none</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># now we check the</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">Anolis_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
  <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
  <span class="s1">&#39;svl&#39;</span><span class="p">:</span> <span class="s1">&#39;4.065917563705425&#39;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">ahlicontains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
  <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;ahli&#39;</span><span class="p">,</span>
  <span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.130889</span><span class="p">,</span>
  <span class="s1">&#39;svl&#39;</span><span class="p">:</span> <span class="s1">&#39;4.039125443&#39;</span>
<span class="p">}</span>

<span class="c1"># use Bayesian method with OU model</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Anolis</span><span class="o">.</span><span class="n">tre</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">svl</span><span class="o">.</span><span class="n">csv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">continuous</span><span class="o">-</span><span class="n">columns</span> <span class="n">svl</span> \
<span class="o">--</span><span class="n">prediction</span><span class="o">-</span><span class="n">method</span> <span class="n">BAYESIAN</span> \
<span class="o">--</span><span class="n">model</span> <span class="n">OU</span> \
<span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">stat</span> <span class="n">none</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">Anolis_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
  <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
  <span class="s1">&#39;svl&#39;</span><span class="p">:</span> <span class="s1">&#39;4.443782202699844&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">ahlicontains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
  <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;ahli&#39;</span><span class="p">,</span>
  <span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.130889</span><span class="p">,</span>
  <span class="s1">&#39;svl&#39;</span><span class="p">:</span> <span class="s1">&#39;4.039125443&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="lineage-specificity-analysis">
<h4><a class="toc-backref" href="#id42" role="doc-backlink">Lineage specificity analysis</a><a class="headerlink" href="#lineage-specificity-analysis" title="Link to this heading">¶</a></h4>
<p>Using <code class="docutils literal notranslate"><span class="pre">--ls-columns</span> <span class="pre">&lt;prop_name&gt;</span></code> to start the lineage specificity analysis of boolean traits, the given trait need to be boolean value such as <code class="docutils literal notranslate"><span class="pre">True`;</span> <span class="pre">``False</span></code>; <code class="docutils literal notranslate"><span class="pre">yes</span></code>; <code class="docutils literal notranslate"><span class="pre">no</span></code>; <code class="docutils literal notranslate"><span class="pre">t</span></code>; <code class="docutils literal notranslate"><span class="pre">f</span></code>; <code class="docutils literal notranslate"><span class="pre">1</span></code>; <code class="docutils literal notranslate"><span class="pre">0</span></code>;  which fit the criteria in treeprofiler annotate. Calculated results will be stored in each internal nodes with suffix of <code class="docutils literal notranslate"><span class="pre">_prec</span></code> , <code class="docutils literal notranslate"><span class="pre">_sens</span></code> and <code class="docutils literal notranslate"><span class="pre">_f1</span></code>.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--ls-columns</span> <span class="pre">LS_COLUMNS</span> <span class="pre">[LS_COLUMNS</span> <span class="pre">...]</span></code></p></td>
<td><p>names of properties to perform lineage specificity analysis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--prec-cutoff</span> <span class="pre">PREC_CUTOFF</span></code></p></td>
<td><p>Precision cutoff for lineage specificity analysis. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">0.95]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--sens-cutoff</span> <span class="pre">SENS_CUTOFF</span></code></p></td>
<td><p>Sensitivity threshold for lineage specificity analysis. <code class="docutils literal notranslate"><span class="pre">[Default:</span> <span class="pre">0.95]</span></code></p></td>
</tr>
</tbody>
</table>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># in the example we loose the cutoff to 0.5</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo2</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">-</span><span class="n">m</span> <span class="n">demo2_ls</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">--</span><span class="n">ls</span><span class="o">-</span><span class="n">columns</span> <span class="n">profile1</span> \
<span class="o">--</span><span class="n">prec</span><span class="o">-</span><span class="n">cutoff</span> <span class="mf">0.5</span> \
<span class="o">--</span><span class="n">sens</span><span class="o">-</span><span class="n">cutoff</span> <span class="mf">0.5</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># check properties</span>
<span class="n">python</span> <span class="n">show_tree_props</span><span class="o">.</span><span class="n">py</span> <span class="n">demo2_annotated</span><span class="o">.</span><span class="n">nw</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">internal</span> <span class="n">node</span> <span class="n">Root</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">properties</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Root&#39;</span><span class="p">,</span>
<span class="s1">&#39;profile1_counter&#39;</span><span class="p">:</span> <span class="s1">&#39;False--33||True--7&#39;</span><span class="p">,</span>
<span class="s1">&#39;profile1_f1&#39;</span><span class="p">:</span> <span class="s1">&#39;0.2978723404255319&#39;</span><span class="p">,</span>
<span class="s1">&#39;profile1_prec&#39;</span><span class="p">:</span> <span class="s1">&#39;0.175&#39;</span><span class="p">,</span>
<span class="s1">&#39;profile1_sens&#39;</span><span class="p">:</span> <span class="s1">&#39;1.0&#39;</span>
<span class="p">}</span>
<span class="n">Target</span> <span class="n">tree</span> <span class="n">leaf</span> <span class="n">node</span> <span class="n">Taxa_0</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">following</span> <span class="n">propertiies</span><span class="p">:</span>
<span class="p">{</span>
<span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Taxa_3&#39;</span><span class="p">,</span>
<span class="s1">&#39;dist&#39;</span><span class="p">:</span> <span class="mf">0.315846</span><span class="p">,</span>
<span class="s1">&#39;profile1&#39;</span><span class="p">:</span> <span class="s1">&#39;False&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
</section>
<section id="output-formats-for-annotated-trees">
<h3><a class="toc-backref" href="#id43" role="doc-backlink">Output formats for annotated trees</a><a class="headerlink" href="#output-formats-for-annotated-trees" title="Link to this heading">¶</a></h3>
<p>TreeProfiler <code class="docutils literal notranslate"><span class="pre">annotate</span></code> subcommand will generate the following output file</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated.nw</em>, newick format with annotated tree</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated.ete</em>, ete format with annotated tree</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated_prop2type.txt</em>, config file where store the datatype of each annotated properties</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;input_tree&gt;</span></code> + <em>_annotated.tsv</em>,  metadata in tab-separated values format with annotated and summarized internal nodes information.</p></li>
</ol>
<p>In the following <code class="docutils literal notranslate"><span class="pre">plot</span></code> step, users can use either <code class="docutils literal notranslate"><span class="pre">.nw</span></code> or <code class="docutils literal notranslate"><span class="pre">.ete</span></code> by putting <code class="docutils literal notranslate"><span class="pre">--input-type</span> <span class="pre">{newick,</span> <span class="pre">ete}</span></code> flag to identify. The difference between <code class="docutils literal notranslate"><span class="pre">.nw</span></code> and <code class="docutils literal notranslate"><span class="pre">.ete</span></code> format is</p>
<blockquote>
<div><ul class="simple">
<li><p>newick file is more universal and be able to used in different other phylogenetic software although associated data of tree nodes will be considered as plain text, so if you use newick format, alongside with the prop2type config file which was generated before by adding <code class="docutils literal notranslate"><span class="pre">--prop2type</span> <span class="pre">&lt;prop2type_file&gt;</span></code></p></li>
<li><p>ete format is a novel format developed to solve the situation we encounter in the previous step, annotated tree can be <strong>recover easily with all the annotated data without changing the data type</strong>. Besides, the ete format optimized the tree file size after mapped with its associated data. Hence it’s very handy for programers in their own script. At this moment we can only view the ete format in treeprofiler, but we will make the ete format more universal to other phylogenetic software. <strong>Hence using ete format in `plot` subcommand is highly reccomended</strong></p></li>
</ul>
</div></blockquote>
<section id="use-standard-input-and-output-formats">
<h4><a class="toc-backref" href="#id44" role="doc-backlink">Use standard input and output formats</a><a class="headerlink" href="#use-standard-input-and-output-formats" title="Link to this heading">¶</a></h4>
<p>TreeProfiler is able to receive and produce trees in standard formats in order to integrate with command-line programs.</p>
<ul class="simple">
<li><p>Standard output: Use <code class="docutils literal notranslate"><span class="pre">--quiet</span></code> and <code class="docutils literal notranslate"><span class="pre">--stdout</span></code> at the same time to mute the log and output the annotated tree in Newick format.</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">basic_example0</span><span class="o">/</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">quiet</span> \
<span class="o">--</span><span class="n">stdout</span>

<span class="p">((</span><span class="n">Taxa_3</span><span class="p">:</span><span class="mf">0.219065</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">C</span><span class="p">],(</span><span class="n">Taxa_4</span><span class="p">:</span><span class="mf">0.188681</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">C</span><span class="p">],</span><span class="n">Taxa_2</span><span class="p">:</span><span class="mf">0.5196</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">B</span><span class="p">])</span><span class="mf">0.166914</span><span class="p">:</span><span class="mf">0.90365</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">B</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">C</span><span class="o">--</span><span class="mi">1</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">N3</span><span class="p">])</span><span class="mf">0.138062</span><span class="p">:</span><span class="mf">0.0632016</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">B</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">C</span><span class="o">--</span><span class="mi">2</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">N4</span><span class="p">],(</span><span class="n">Taxa_0</span><span class="p">:</span><span class="mf">0.190563</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">A</span><span class="p">],</span><span class="n">Taxa_1</span><span class="p">:</span><span class="mf">0.458423</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">B</span><span class="p">])</span><span class="mf">0.138062</span><span class="p">:</span><span class="mf">0.97338</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">A</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">B</span><span class="o">--</span><span class="mi">1</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">N7</span><span class="p">])[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">A</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">B</span><span class="o">--</span><span class="mi">2</span><span class="o">||</span><span class="n">C</span><span class="o">--</span><span class="mi">2</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">Root</span><span class="p">];</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Standard input, use <code class="docutils literal notranslate"><span class="pre">-</span></code> as input in <code class="docutils literal notranslate"><span class="pre">--tree</span></code> argument to take standard input into account.</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">basic_example0</span><span class="o">/</span>

<span class="n">cat</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span><span class="o">|</span><span class="n">treeprofiler</span> <span class="n">annotate</span> <span class="o">-</span><span class="n">t</span> <span class="o">-</span> <span class="o">--</span><span class="n">metadata</span> <span class="n">categorical</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">s</span> <span class="p">,</span> <span class="o">--</span><span class="n">quiet</span> <span class="o">--</span><span class="n">stdout</span>

<span class="p">((</span><span class="n">Taxa_3</span><span class="p">:</span><span class="mf">0.219065</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">C</span><span class="p">],(</span><span class="n">Taxa_4</span><span class="p">:</span><span class="mf">0.188681</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">C</span><span class="p">],</span><span class="n">Taxa_2</span><span class="p">:</span><span class="mf">0.5196</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">B</span><span class="p">])</span><span class="mf">0.166914</span><span class="p">:</span><span class="mf">0.90365</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">B</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">C</span><span class="o">--</span><span class="mi">1</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">N3</span><span class="p">])</span><span class="mf">0.138062</span><span class="p">:</span><span class="mf">0.0632016</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">B</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">C</span><span class="o">--</span><span class="mi">2</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">N4</span><span class="p">],(</span><span class="n">Taxa_0</span><span class="p">:</span><span class="mf">0.190563</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">A</span><span class="p">],</span><span class="n">Taxa_1</span><span class="p">:</span><span class="mf">0.458423</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1</span><span class="o">=</span><span class="n">B</span><span class="p">])</span><span class="mf">0.138062</span><span class="p">:</span><span class="mf">0.97338</span><span class="p">[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">A</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">B</span><span class="o">--</span><span class="mi">1</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">N7</span><span class="p">])[</span><span class="o">&amp;&amp;</span><span class="n">NHX</span><span class="p">:</span><span class="n">categorical1_counter</span><span class="o">=</span><span class="n">A</span><span class="o">--</span><span class="mi">1</span><span class="o">||</span><span class="n">B</span><span class="o">--</span><span class="mi">2</span><span class="o">||</span><span class="n">C</span><span class="o">--</span><span class="mi">2</span><span class="p">:</span><span class="n">name</span><span class="o">=</span><span class="n">Root</span><span class="p">];</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="treeprofiler-plot-visualizing-phylogenetic-profiles-and-annotation">
<h2><a class="toc-backref" href="#id45" role="doc-backlink"><code class="docutils literal notranslate"><span class="pre">treeprofiler-plot</span></code> visualizing phylogenetic profiles and annotation</a><a class="headerlink" href="#treeprofiler-plot-visualizing-phylogenetic-profiles-and-annotation" title="Link to this heading">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">plot</span></code> command can be used to visually explore any annotated tree produced by the
<code class="docutils literal notranslate"><span class="pre">annotate</span></code> program, automatically adjusting the most suitable graphical representation
for each data source. Alternatively, users can easily adjust which annotations should be shown and
choose among different predefined layouts.</p>
<section id="interactive-visualization-interface">
<h3><a class="toc-backref" href="#id46" role="doc-backlink">Interactive visualization interface</a><a class="headerlink" href="#interactive-visualization-interface" title="Link to this heading">¶</a></h3>
<p>TreeProfiler uses the new visualization framework implemented in [ETE 4.0](<a class="reference external" href="https://github.com/etetoolkit/ete/tree/ete4">https://github.com/etetoolkit/ete/tree/ete4</a>), which allows for the interactive exploration of huge phylogenies based on a context-based adaptive zooming strategy.</p>
<p>we use example from <cite>examples/basic_example1</cite> , which contain all kind of datatype</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="o">-</span><span class="mi">3</span> <span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span> <span class="n">basic_example1_metadata2</span><span class="o">.</span><span class="n">tsv</span>
<span class="o">==&gt;</span> <span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span> <span class="o">&lt;==</span>
<span class="c1">#name       sample1 sample2 sample3 sample4 sample5 random_type     bool_type       bool_type2</span>
<span class="n">Phy003I7ZJ_CHICK</span>    <span class="mf">0.05</span>    <span class="mf">0.12</span>    <span class="mf">0.86</span>    <span class="mf">0.01</span>    <span class="mf">0.69</span>    <span class="n">medium</span>  <span class="mi">1</span>       <span class="n">TRUE</span>
<span class="n">Phy0054BO3_MELGA</span>    <span class="mf">0.64</span>    <span class="mf">0.67</span>    <span class="mf">0.51</span>    <span class="mf">0.29</span>    <span class="mf">0.14</span>    <span class="n">medium</span>  <span class="mi">1</span>       <span class="n">TRUE</span>

<span class="o">==&gt;</span> <span class="n">basic_example1_metadata2</span><span class="o">.</span><span class="n">tsv</span> <span class="o">&lt;==</span>
<span class="c1">#name       abs_data        list_data       abs_data2</span>
<span class="n">Phy003I7ZJ_CHICK</span>    <span class="mi">97</span>      <span class="n">w</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">t</span>   <span class="mi">50</span>
<span class="n">Phy0054BO3_MELGA</span>    <span class="mi">16</span>      <span class="n">r</span><span class="p">,</span><span class="n">q</span><span class="p">,</span><span class="n">s</span>   <span class="mi">245</span>

<span class="c1"># annotation</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span> <span class="n">basic_example1_metadata2</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">rectangle</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">profiling</span><span class="o">-</span><span class="n">layout</span> <span class="n">list_data</span>
</pre></div>
</div>
<img alt="treeprofiler interface" class="align-center" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/control_panel-0002.png?raw=true" />
<p>Overview of the TreeProfiler visualization interface. (A) The control panel allows users to customize visualization layout and features, and to perform text-based searches. (B) An annotated example tree, from <code class="docutils literal notranslate"><span class="pre">examples/basic_example1/</span></code> after <code class="docutils literal notranslate"><span class="pre">annotate</span></code>, is launched with a command <code class="docutils literal notranslate"><span class="pre">plot</span></code>. Support values (red) and branch distance (grey) are displayed on top of branches. The properties of one of the nodes are shown on the top. The minimap (bottom right) facilitates navigation. (C) The node editor panel provides access to node-specific actions, such as creating subtrees, collapsing, pruning, rooting and more. (D) Visualized properties by order are, categorical data <code class="docutils literal notranslate"><span class="pre">random_type</span></code> in <code class="docutils literal notranslate"><span class="pre">rectangle-layout</span></code>, numerical data <code class="docutils literal notranslate"><span class="pre">sample1</span></code>, <code class="docutils literal notranslate"><span class="pre">sample2</span></code>, <code class="docutils literal notranslate"><span class="pre">sample3</span></code> in <code class="docutils literal notranslate"><span class="pre">heatmap-layout</span></code> and <code class="docutils literal notranslate"><span class="pre">sample4</span></code>, <code class="docutils literal notranslate"><span class="pre">sample5</span></code> in <code class="docutils literal notranslate"><span class="pre">barplot-layout</span></code>, categorical data <code class="docutils literal notranslate"><span class="pre">random_type</span></code> in <code class="docutils literal notranslate"><span class="pre">profiling-layout</span></code> shown as presence-absence matrix. Layouts are shown with the order as input argument order from the command line. Names of properties are shown as titles on the top of each layout. (E) Legends each layout is shown on the top right corner with the same order as the layouts.</p>
</section>
<section id="basic-options-of-visualizing-layouts">
<h3><a class="toc-backref" href="#id47" role="doc-backlink">Basic options of visualizing layouts</a><a class="headerlink" href="#basic-options-of-visualizing-layouts" title="Link to this heading">¶</a></h3>
<p>Selected properties of tree will be visualized at the aligned panel alongside with the tree, here is some basic parameters for layouts.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--column-width</span></code></p></td>
<td><p>Column width of each property in layout which shown in aligned panel. <code class="docutils literal notranslate"><span class="pre">[default:</span> <span class="pre">20]</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--padding-x</span></code></p></td>
<td><p>Customize horizontal column padding distance of each layout in aligned panel. <code class="docutils literal notranslate"><span class="pre">[default:</span> <span class="pre">1]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--padding-y</span></code></p></td>
<td><p>Customize vertical padding distance of each layout in aligned panel. <code class="docutils literal notranslate"><span class="pre">[default:</span> <span class="pre">0]</span></code></p></td>
</tr>
</tbody>
</table>
<p>Examples with basic parameters:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># change column width from default 20 px to 50px</span>
<span class="c1"># padding x from default 1 to 5</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
  <span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
  <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
  <span class="o">--</span><span class="n">rectangle</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
  <span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> \
  <span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> \
  <span class="o">--</span><span class="n">profiling</span><span class="o">-</span><span class="n">layout</span> <span class="n">list_data</span>
  <span class="o">--</span><span class="n">column</span><span class="o">-</span><span class="n">width</span> <span class="mi">50</span> \
  <span class="o">--</span><span class="n">padding</span><span class="o">-</span><span class="n">x</span> <span class="mi">5</span>
</pre></div>
</div>
<img alt="treeprofiler interface" class="align-center" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/control_panel-0003.png?raw=true" />
</section>
<section id="layouts-for-categorical-data">
<h3><a class="toc-backref" href="#id48" role="doc-backlink">Layouts for categorical data</a><a class="headerlink" href="#layouts-for-categorical-data" title="Link to this heading">¶</a></h3>
<p>Users can add the following flag to activate layouts for categorical data.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Leaf node</p></th>
<th class="head"><p>Internal node</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--label-layout</span> <span class="pre">LABEL_LAYOUT</span> <span class="pre">[LABEL_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties where values will be displayed on the aligned panel.</p></td>
<td><p>TextFace</p></td>
<td><p>Stacked Horizontal RecFace (only collapsed)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--rectangle-layout</span> <span class="pre">RECTANGLE_LAYOUT</span> <span class="pre">[RECTANGLE_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties where values will be labeled as rectangular color blocks on the aligned panel.</p></td>
<td><p>RecFace</p></td>
<td><p>Stacked Horizontal RecFace (only collapsed)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--colorbranch-layout</span> <span class="pre">COLORBRANCH_LAYOUT</span> <span class="pre">[COLORBRANCH_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties where branches will be colored based on different values.</p></td>
<td><p>Branch with color</p></td>
<td><p>Stacked Horizontal RecFace (only collapsed)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--bubble-layout</span> <span class="pre">BUBBLE_LAYOUT</span> <span class="pre">[BUBBLE_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties where nodes will be colored based on different bubble.</p></td>
<td><p>Circles with color</p></td>
<td><p>None</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--background-layout</span> <span class="pre">BACKGROUND_LAYOUT</span> <span class="pre">[BACKGROUND_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties where values will be labeled as rectangular color blocks on the aligned panel.</p></td>
<td><p>Background with color</p></td>
<td><p>Stacked Horizontal RecFace (only collapsed)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--piechart-layout</span> <span class="pre">PIECHART_LAYOUT</span> <span class="pre">[PIECHART_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties whose internal nodes need to be plotted as piechart-layout.</p></td>
<td><p>None</p></td>
<td><p>PiechartFace</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--profiling-layout</span> <span class="pre">PROFILING_LAYOUT</span> <span class="pre">[PROFILING_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties which need to be converted to a presence-absence profiling matrix of each value.</p></td>
<td><p>presence/absence matrix; Array from ete4</p></td>
<td><p>gradient of presence/total heatmap (only collapsed)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--categorical-matrix-layout</span> <span class="pre">CATEGORICAL_MATRIX_LAYOUT</span> <span class="pre">[CATEGORICAL_MATRIX_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names which need to be plotted as categorical_matrix_layout for categorical values.</p></td>
<td><p>array from ete4</p></td>
<td><p>None</p></td>
</tr>
</tbody>
</table>
<section id="lable-layout">
<h4><a class="toc-backref" href="#id49" role="doc-backlink">Lable Layout</a><a class="headerlink" href="#lable-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--label-layout</span></code> will display the values of the given properties on the aligned panel. It will be shown as TextFace on leaf nodes and Stacked Horizontal RecFace on internal nodes.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="label_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_label_layout.jpeg?raw=true" />
</section>
<section id="rectangle-layout">
<h4><a class="toc-backref" href="#id50" role="doc-backlink">Rectangle Layout</a><a class="headerlink" href="#rectangle-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--rectangle-layout</span></code> will assign a color to each variable under the property, displaying as colored rectangle in aligned panel, when internal branches collapsed,  aligned panel shows the horizontal stacked bar to demostrate the composition of each variable.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">rectangle</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="rectangle_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_rectangular_layout.jpeg?raw=true" />
</section>
<section id="colorbranch-layout">
<h4><a class="toc-backref" href="#id51" role="doc-backlink">ColorBranch Layout</a><a class="headerlink" href="#colorbranch-layout" title="Link to this heading">¶</a></h4>
<p><cite>–colorbranch-layout</cite> will assign a color to each variable under the property, displaying as colored branch of corresponding node.</p>
<p>Noted that in this case from leaf to root, if node contains the given property, it will still have the colored branch.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># every node share the property &quot;name&quot;</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">colorbranch</span><span class="o">-</span><span class="n">layout</span> <span class="n">name</span>
</pre></div>
</div>
<img alt="colorbranch_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_colorbranch_layout_name.png?raw=true" />
<p>If internal node doesn’t have the given property, once it collapsed,  aligned panel shows the horizontal stacked bar to demostrate the composition of each variable.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># show normal counter</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">colorbranch</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="colorbranch_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_colorbranch_layout_random.png?raw=true" />
</section>
<section id="bubble-layout">
<h4><a class="toc-backref" href="#id52" role="doc-backlink">Bubble Layout</a><a class="headerlink" href="#bubble-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--bubble-layout</span></code> categorizes values of selected property by displaying color bubble on the corresponing leaf node.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">bubble</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="bubble_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_bubble_layout_randomtype.png?raw=true" />
</section>
<section id="background-layout">
<h4><a class="toc-backref" href="#id53" role="doc-backlink">Background Layout</a><a class="headerlink" href="#background-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--background-layout</span></code> works just like <code class="docutils literal notranslate"><span class="pre">--colorbranch-layout</span></code>, the only difference is <code class="docutils literal notranslate"><span class="pre">--background-layout</span></code> visualize the background of the corresponding node instead of the branch color. It is useful and more visual when the tree is large.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">background</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="background_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_background_layout.png?raw=true" />
</section>
<section id="piechart-layout">
<h4><a class="toc-backref" href="#id54" role="doc-backlink">Piechart Layout</a><a class="headerlink" href="#piechart-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--piechart-layout</span></code> is unique layout designed for visualize internal nodes which contain counter of the given property from children nodes.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">piechart</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="piechart_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_piechart_layout.png?raw=true" />
</section>
<section id="profiling-layout">
<h4><a class="toc-backref" href="#id55" role="doc-backlink">Profiling Layout</a><a class="headerlink" href="#profiling-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--profiling-layout</span></code> will convert categorical trait regardless <code class="docutils literal notranslate"><span class="pre">str</span></code> or <code class="docutils literal notranslate"><span class="pre">list</span></code> into presence-absence matrix. Importantly, once it collapsed, aligned matrix will show the gradient of presence/total of corresponding trait. This layout using draw_array algorithm from ete4 therefore it is s suitable for large scale.</p>
<p>single value example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check metadata</span>
<span class="n">awk</span> <span class="s1">&#39;{print $1,$7}&#39;</span> <span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span><span class="o">|</span><span class="n">head</span>
<span class="c1">#name random_type</span>
<span class="n">Phy003I7ZJ_CHICK</span> <span class="n">medium</span>
<span class="n">Phy0054BO3_MELGA</span> <span class="n">medium</span>
<span class="n">Phy00508FR_NIPNI</span> <span class="n">low</span>
<span class="n">Phy004O1E0_APTFO</span> <span class="n">medium</span>
<span class="n">Phy004PA1B_ANAPL</span> <span class="n">medium</span>


<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">profiling</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="profiling_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_profiling_layout_single.png?raw=true" />
<p>List value example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check metadata</span>
<span class="n">awk</span> <span class="s1">&#39;{print $1,$3}&#39;</span> <span class="n">basic_example1_metadata2</span><span class="o">.</span><span class="n">tsv</span><span class="o">|</span><span class="n">head</span>
<span class="c1">#name list_data</span>
<span class="n">Phy003I7ZJ_CHICK</span> <span class="n">w</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">t</span>
<span class="n">Phy0054BO3_MELGA</span> <span class="n">r</span><span class="p">,</span><span class="n">q</span><span class="p">,</span><span class="n">s</span>
<span class="n">Phy00508FR_NIPNI</span> <span class="n">z</span><span class="p">,</span><span class="n">f</span><span class="p">,</span><span class="n">p</span>
<span class="n">Phy004O1E0_APTFO</span> <span class="n">z</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">b</span>
<span class="n">Phy004PA1B_ANAPL</span> <span class="n">z</span><span class="p">,</span><span class="n">r</span><span class="p">,</span><span class="n">p</span>
<span class="n">Phy004TLNA_APAVI</span> <span class="n">u</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">i</span>

<span class="c1"># convert each letter into presence/absence matrix</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">profiling</span><span class="o">-</span><span class="n">layout</span> <span class="n">list_data</span>
</pre></div>
</div>
<img alt="profiling_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_profiling_layout_list.png?raw=true" />
</section>
<section id="categorical-matrix-layout">
<h4><a class="toc-backref" href="#id56" role="doc-backlink">Categorical Matrix Layout</a><a class="headerlink" href="#categorical-matrix-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--categorical-matrix-layout</span></code> is similar to <code class="docutils literal notranslate"><span class="pre">--rectangle-layout</span></code> , which shows color block to represent the variable. But <code class="docutils literal notranslate"><span class="pre">--rectangle-layout</span></code> take <strong>EACH</strong> given column as one individual case as one layout, <code class="docutils literal notranslate"><span class="pre">--categorical-matrix-layout</span></code> take <strong>ALL</strong> given columns as one case to one layout.</p>
</section>
</section>
<section id="layouts-for-boolean-data">
<h3><a class="toc-backref" href="#id57" role="doc-backlink">Layouts for boolean data</a><a class="headerlink" href="#layouts-for-boolean-data" title="Link to this heading">¶</a></h3>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Leaf node</p></th>
<th class="head"><p>Internal node</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--binary-layout</span> <span class="pre">BINARY_LAYOUT</span> <span class="pre">[BINARY_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of properties which need to be plotted as binary-layout which highlights the positives. Each column has different colors. Internal node represent ratio of true/total gradient.</p></td>
<td><p>positive/negative RectFace</p></td>
<td><p>Gradient RectFace</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--binary-aggregate-layout</span> <span class="pre">BINARY_AGGREGATE_LAYOUT</span> <span class="pre">[BINARY_AGGREGATE_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of properties which need to be plotted as binary-aggregate-layout which highlights the positives. Each column has different colors. Internal node present the accumulated counts of true and represent on color</p></td>
<td><p>positive/negative RectFace</p></td>
<td><p>Accumulated RectFace</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--binary-unicolor-layout</span> <span class="pre">BINARY_UNICOLOR_LAYOUT</span> <span class="pre">[BINARY_UNICOLOR_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of properties which need to be plotted as binary-layout which highlights the positives. Each column has same colors. Internal node represent ratio of true/total gradient.</p></td>
<td><p>positive/negative RectFace</p></td>
<td><p>Gradient RectFace</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--binary-unicolor-aggregate-layout</span> <span class="pre">BINARY_UNICOLOR_AGGREGATE_LAYOUT</span> <span class="pre">[BINARY_UNICOLOR_AGGREGATE_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of properties which need to be plotted as binary-aggregate-layout which highlights the positives. Each column has same colors. Internal node present the accumulated counts of true and represent on color</p></td>
<td><p>positive/negative RectFace</p></td>
<td><p>Accumulated RectFace</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--binary-matrix-layout</span> <span class="pre">BINARY_MATRIX_LAYOUT</span> <span class="pre">[BINARY_MATRIX_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of properties which need to be plotted as binary-matrix which highlights the positives. Each column has same colors. Internal node represent ratio of true/total gradient. Suitable for large scale (&gt;100 columns)</p></td>
<td><p>Array from ete4</p></td>
<td><p>Gradient RectFace</p></td>
</tr>
</tbody>
</table>
<section id="binary-layout">
<h4><a class="toc-backref" href="#id58" role="doc-backlink">Binary Layout</a><a class="headerlink" href="#binary-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--binary-layout</span></code> highlights the positives. Each column has <strong>different</strong> colors. Internal node represent ratio of <strong>true/total gradient</strong>.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> <span class="n">bool_type2</span>
</pre></div>
</div>
<img alt="binary_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_binary_layout.png?raw=true" />
</section>
<section id="binary-aggregate-layout">
<h4><a class="toc-backref" href="#id59" role="doc-backlink">Binary Aggregate Layout</a><a class="headerlink" href="#binary-aggregate-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--binary-aggregate-layout</span></code> highlights the positives. Each column has <strong>different</strong> colors. Internal node present the <strong>accumulated counts of true</strong> and represent on color.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">aggregate</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> <span class="n">bool_type2</span>
</pre></div>
</div>
<img alt="binary_layout2 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_binary_layout2.png?raw=true" />
</section>
<section id="binary-unicolor-layout">
<h4><a class="toc-backref" href="#id60" role="doc-backlink">Binary Unicolor Layout</a><a class="headerlink" href="#binary-unicolor-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--binary-unicolor-layout</span></code> Each column has <strong>same</strong> colors. Internal node represent ratio of <strong>true/total gradient</strong>.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">unicolor</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> <span class="n">bool_type2</span>
</pre></div>
</div>
<img alt="binary_layout3 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_binary_layout3.png?raw=true" />
</section>
<section id="binary-unicolor-aggregate-layout">
<h4><a class="toc-backref" href="#id61" role="doc-backlink">Binary Unicolor Aggregate Layout</a><a class="headerlink" href="#binary-unicolor-aggregate-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--binary-unicolor-aggregate-layout</span></code> Each column has <strong>same</strong> colors. Internal node present the <strong>accumulated counts of true</strong> and represent on color.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">unicolor</span><span class="o">-</span><span class="n">aggregate</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> <span class="n">bool_type2</span>
</pre></div>
</div>
<img alt="binary_layout4 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_binary_layout4.png?raw=true" />
</section>
<section id="binary-matrix-layout">
<h4><a class="toc-backref" href="#id62" role="doc-backlink">Binary Matrix Layout</a><a class="headerlink" href="#binary-matrix-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--binary-matrix-layout</span></code> is similar to <code class="docutils literal notranslate"><span class="pre">--binary-layout</span></code> but using draw_array from ete4 therefore it ss suitable for large scale. But the limit is only work on rectangular tree</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">matrix</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> <span class="n">bool_type2</span>
</pre></div>
</div>
<img alt="binary_layout5 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_binary_layout5.png?raw=true" />
</section>
</section>
<section id="layouts-for-numerical-data">
<h3><a class="toc-backref" href="#id63" role="doc-backlink">Layouts for Numerical data</a><a class="headerlink" href="#layouts-for-numerical-data" title="Link to this heading">¶</a></h3>
<p>Users can add the following arguments to use layouts for Numerical data</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Applied on data-matrix (file)?</p></th>
<th class="head"><p>Applied on negative data?</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--colorbranch-layout</span> <span class="pre">COLORBRANCH_LAYOUT</span> <span class="pre">[COLORBRANCH_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of properties where branches will be colored based on different values.</p></td>
<td><p>no</p></td>
<td><p>yes</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--bubble-layout</span> <span class="pre">BUBBLE_LAYOUT</span> <span class="pre">[BUBBLE_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>(experimental) &lt;prop1&gt; &lt;prop2&gt; names of properties which need to be plotted as bubble-layout.</p></td>
<td><p>no</p></td>
<td><p>yes</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--heatmap-layout</span> <span class="pre">HEATMAP_LAYOUT</span> <span class="pre">[HEATMAP_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of numerical properties which need to be read as heatmap-layout.</p></td>
<td><p>no</p></td>
<td><p>yes</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--heatmap-mean-layout</span> <span class="pre">HEATMAP_MEAN_LAYOUT</span> <span class="pre">[HEATMAP_MEAN_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of numerical properties which need to be read as heatmap-layout.</p></td>
<td><p>no</p></td>
<td><p>yes</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--heatmap-zscore-layout</span> <span class="pre">HEATMAP_ZSCORE_LAYOUT</span> <span class="pre">[HEATMAP_ZSCORE_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of numerical properties which need to be read as heatmap-layout.</p></td>
<td><p>no</p></td>
<td><p>yes</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--barplot-layout</span> <span class="pre">BARPLOT_LAYOUT</span> <span class="pre">[BARPLOT_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>&lt;prop1&gt; &lt;prop2&gt; names of numerical properties which need to be read as barplot_layouts.</p></td>
<td><p>no</p></td>
<td><p>no</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--barplot-width</span></code></p></td>
<td><p>setting barplot length [default: 200]</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--barplot-scale</span> <span class="pre">&lt;prop1&gt;</span></code></p></td>
<td><p>using certain prop as the barplot scale for all of the barplot. [default: None]</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--numerical-matrix-layout</span> <span class="pre">NUMERICAL_MATRIX_LAYOUT</span> <span class="pre">[NUMERICAL_MATRIX_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p>numerical matrix that takes into account ALL values into gradient from white to red. &lt;prop1&gt; &lt;prop2&gt; names which need to be plotted as numerical_matrix_layout for numerical values.</p></td>
<td><p>yes</p></td>
<td><p>yes</p></td>
</tr>
</tbody>
</table>
<section id="id1">
<h4><a class="toc-backref" href="#id64" role="doc-backlink">ColorBranch Layout</a><a class="headerlink" href="#id1" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--colorbranch-layout</span></code> can detect if property is numerical data, then will color the branch based on color scale from minimum to maximum, including branches from the internal nodes. If internal nodes contain the corresponding property, then the relative value will be the representative; If not that the value will be the summarized value from descriptive statistic.</p>
<p>Colorbranch with property that share from leaf to root:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># all nodes have dist value</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">colorbranch</span><span class="o">-</span><span class="n">layout</span> <span class="n">dist</span>

<span class="c1"># only internal nodes have support value</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">colorbranch</span><span class="o">-</span><span class="n">layout</span> <span class="n">support</span>
</pre></div>
</div>
<img alt="colorbranch_dist example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_colorbranch_layout_dist.png?raw=true" />
<img alt="colorbranch_support example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_colorbranch_layout_support.png?raw=true" />
<p>Colorbranch with summarized property (avg will be the representative value of internal nodes):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">colorbranch</span><span class="o">-</span><span class="n">layout</span> <span class="n">abs_data2</span>
</pre></div>
</div>
<img alt="colorbranch_absdata example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_colorbranch_layout_support.png?raw=true" />
</section>
<section id="id2">
<h4><a class="toc-backref" href="#id65" role="doc-backlink">Bubble Layout</a><a class="headerlink" href="#id2" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--bubble-layout</span></code> will scale the given absolute value property as size of bubble. Noted that if value is negative, bubble will be blue, otherwise bubble is red.</p>
<p>If internal nodes contain the corresponding property, then the relative value will be the representative; If not that the value will be the summarized value from descriptive statistic.</p>
<p>Bubble with property that share from leaf to root:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> <span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> <span class="o">--</span><span class="n">bubble</span><span class="o">-</span><span class="n">layout</span> <span class="n">dist</span>
</pre></div>
</div>
<img alt="bubble_dist example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_bubble_layout_dist.png?raw=true" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> <span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> <span class="o">--</span><span class="n">bubble</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span>
</pre></div>
</div>
<img alt="bubble_sample example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_bubble_layout_sample1.png?raw=true" />
</section>
<section id="heatmap-layout">
<h4><a class="toc-backref" href="#id66" role="doc-backlink">Heatmap Layout</a><a class="headerlink" href="#heatmap-layout" title="Link to this heading">¶</a></h4>
<p>A series of heatmap layout is only the matter how to normalize the value to show the gradient of color.</p>
<p><code class="docutils literal notranslate"><span class="pre">--heatmap-layout</span></code>  values will be normalized between 0 and 1 (minimu, maximum).</p>
<p><code class="docutils literal notranslate"><span class="pre">--heatmap-mean-layout</span></code> values will be normalized between -1 and 1 by using the formula (val-mean)/(max-min)</p>
<p><code class="docutils literal notranslate"><span class="pre">--heatmap-zscore-layout</span></code>  values will be Z-score normalized by using the formula (val-mean)/std.</p>
<p><code class="docutils literal notranslate"><span class="pre">--heatmap-layout</span></code> Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span>
</pre></div>
</div>
<img alt="heatmap_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_heatmap_layout1.png?raw=true" />
<p><code class="docutils literal notranslate"><span class="pre">--heatmap-mean-layout</span></code> Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">mean</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span>
</pre></div>
</div>
<img alt="heatmap_layout2 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_heatmap_layout2.png?raw=true" />
<p><code class="docutils literal notranslate"><span class="pre">--heatmap-zscore-layout</span></code> Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">zscore</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span>
</pre></div>
</div>
<img alt="heatmap_layout3 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_heatmap_layout3.png?raw=true" />
</section>
<section id="barplot-layout">
<h4><a class="toc-backref" href="#id67" role="doc-backlink">Barplot Layout</a><a class="headerlink" href="#barplot-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--barplot-layout</span></code> will show the value as barplot, the length of barplot is the value of the property. will show barplot of numerical color. Noted that negative value won’t show in barplot</p>
<p><code class="docutils literal notranslate"><span class="pre">--barplot-width</span></code> is the parameter to set the length of barplot, default is 200px.</p>
<p><code class="docutils literal notranslate"><span class="pre">--barplot-scale</span></code> is the parameter to set the scale of barplot, if not given, the scale will be the maximum value of the property.</p>
<p><code class="docutils literal notranslate"><span class="pre">--barplot-colorby</span></code> set the color of barplot by the a categorical property from metadata.[default: None]</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span>
</pre></div>
</div>
<img alt="barplot_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_barplot_layout1.png?raw=true" />
<p>Using <code class="docutils literal notranslate"><span class="pre">--barplot-width</span></code> and <code class="docutils literal notranslate"><span class="pre">--barplot-scale</span></code> can help with different setting of barplot.</p>
<p><code class="docutils literal notranslate"><span class="pre">--barplot-scale</span></code> can use one certain prop to set the scale of all barplot in order to facilitate the cross column comparison.</p>
<p>Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># two different columns with differen value in same scale, but change the width of barplot</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">width</span> <span class="mi">100</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">layout</span> <span class="n">abs_data</span> <span class="n">abs_data2</span>

<span class="c1"># two different columns using column &quot;abs_data&quot; as scale to cross compare with &quot;abs_data2&quot;</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">layout</span> <span class="n">abs_data</span> <span class="n">abs_data2</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">scale</span> <span class="n">abs_data</span>
</pre></div>
</div>
<img alt="barplot_layout2 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_barplot_layout2.png?raw=true" />
<img alt="barplot_layout3 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_barplot_layout3.png?raw=true" />
<p>Noticed barplot in general is filled with one color for each column, using <cite>–barplot-colorby</cite> allows users to fill barplot based on other categorical data.</p>
<p>here we color barplot <code class="docutils literal notranslate"><span class="pre">abs_data</span></code> by the categorical data <code class="docutils literal notranslate"><span class="pre">random_type</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">layout</span> <span class="n">abs_data</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">colorby</span> <span class="n">random_type</span>
</pre></div>
</div>
<img alt="barplot_layout4 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_barplot_layout4.png?raw=true" />
</section>
<section id="numerical-matrix-layout">
<h4><a class="toc-backref" href="#id68" role="doc-backlink">Numerical Matrix Layout</a><a class="headerlink" href="#numerical-matrix-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--numerical-matrix-layout</span></code> is similar to <code class="docutils literal notranslate"><span class="pre">--heatmap-layout</span></code> but using draw_array from ete4 therefore it is suitable for large scale, and also visualize data matrix which was stored as a file in previous annotation step.. But the limit is only work on rectangular tree.</p>
<p>Here we example from <code class="docutils literal notranslate"><span class="pre">examples/basic_example0</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span>
<span class="n">Taxa_0</span><span class="p">,</span><span class="o">-</span><span class="mf">2.591</span><span class="p">,</span><span class="mf">1.937</span><span class="p">,</span><span class="o">-</span><span class="mf">3.898</span><span class="p">,</span><span class="mf">0.447</span><span class="p">,</span><span class="o">-</span><span class="mf">1.349</span>
<span class="n">Taxa_1</span><span class="p">,</span><span class="mf">3.366</span><span class="p">,</span><span class="o">-</span><span class="mf">1.871</span><span class="p">,</span><span class="mf">4.362</span><span class="p">,</span><span class="mf">1.585</span><span class="p">,</span><span class="o">-</span><span class="mf">2.479</span>
<span class="n">Taxa_2</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mf">0.098</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mf">3.326</span><span class="p">,</span><span class="mf">2.746</span>
<span class="n">Taxa_3</span><span class="p">,</span><span class="mf">3.671</span><span class="p">,</span><span class="o">-</span><span class="mf">0.947</span><span class="p">,</span><span class="o">-</span><span class="mf">4.509</span><span class="p">,</span><span class="o">-</span><span class="mf">3.131</span><span class="p">,</span><span class="o">-</span><span class="mf">2.194</span>
<span class="n">Taxa_4</span><span class="p">,</span><span class="mf">0.676</span><span class="p">,</span><span class="o">-</span><span class="mf">2.356</span><span class="p">,</span><span class="o">-</span><span class="mf">4.825</span><span class="p">,</span><span class="mf">0.115</span><span class="p">,</span><span class="mi">0</span>

<span class="c1">#Annotate data.array column by column</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span> \
<span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">headers</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># visualize each column using numerical matrix</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">numerical</span><span class="o">-</span><span class="n">matrix</span><span class="o">-</span><span class="n">layout</span> <span class="n">col1</span> <span class="n">col2</span> <span class="n">col3</span> <span class="n">col4</span> <span class="n">col5</span>

<span class="c1"># Or annotate data.array as a data matrix file</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">matrix</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># now data.array is a property</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">numerical</span><span class="o">-</span><span class="n">matrix</span><span class="o">-</span><span class="n">layout</span> <span class="n">data</span><span class="o">.</span><span class="n">array</span>
</pre></div>
</div>
<img alt="numerical_matrix1 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_numerical_matrix_layout1.png?raw=true" />
<img alt="numerical_matrix2 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_numerical_matrix_layout2.png?raw=true" />
</section>
</section>
<section id="layouts-for-analytical-data">
<h3><a class="toc-backref" href="#id69" role="doc-backlink">Layouts for Analytical Data</a><a class="headerlink" href="#layouts-for-analytical-data" title="Link to this heading">¶</a></h3>
<p>Users can add the following arguments to use layouts for Analytical data</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--acr-discrete-layout</span> <span class="pre">ACR_DISCRETE_LAYOUT</span> <span class="pre">[ACR_DISCRETE_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties that need to be plotted as acr-discrete-layout.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--acr-continuous-layout</span> <span class="pre">ACR_CONTINUOUS_LAYOUT</span> <span class="pre">[ACR_CONTINUOUS_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties that need to be plotted as acr-continuous-layout.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--ls-layout</span> <span class="pre">LS_LAYOUT</span> <span class="pre">[LS_LAYOUT</span> <span class="pre">...]</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">&lt;prop1&gt;</span> <span class="pre">&lt;prop2&gt;</span></code> names of properties that need to be plotted as ls-layout.</p></td>
</tr>
</tbody>
</table>
<p>We use example from <cite>examples/analytic_example</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># start ACR computing for Country trait and Lineage-Specificity analysis for is_Albania and is_Greece traits</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax</span><span class="o">.</span><span class="n">nwk</span> \
<span class="o">--</span><span class="n">internal</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">metadata_tab_ls</span><span class="o">.</span><span class="n">csv</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">discrete</span><span class="o">-</span><span class="n">columns</span> <span class="n">Country</span> \
<span class="o">--</span><span class="n">delta</span><span class="o">-</span><span class="n">stats</span> \
<span class="o">--</span><span class="n">ls</span><span class="o">-</span><span class="n">columns</span> <span class="n">is_Albania</span> <span class="n">is_Greece</span> \
<span class="o">--</span><span class="n">prec</span><span class="o">-</span><span class="n">cutoff</span> <span class="mf">0.7</span> \
<span class="o">--</span><span class="n">sens</span><span class="o">-</span><span class="n">cutoff</span> <span class="mf">0.7</span> \
<span class="o">--</span><span class="n">threads</span> <span class="mi">4</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>
</pre></div>
</div>
<section id="ancestral-character-reconstruction-and-phylogenetic-signal-layout">
<h4><a class="toc-backref" href="#id70" role="doc-backlink">Ancestral Character Reconstruction and Phylogenetic Signal Layout</a><a class="headerlink" href="#ancestral-character-reconstruction-and-phylogenetic-signal-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--acr-discrete-layout</span></code> is the layout designed for visualizing annotated tree of acr from previous annotate step or manual annotation. Noted that if phylogenetic signal test <code class="docutils literal notranslate"><span class="pre">--delta-stats</span></code> was executed, delta stat metric and p_value will be showed next to the root node.</p>
<p>The infered character of each internal node will be shown as colorbranch, including the uncertain character</p>
<p>We use example from <cite>examples/analytic_example</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize ACR annotation</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">discrete</span><span class="o">-</span><span class="n">layout</span> <span class="n">Country</span>
</pre></div>
</div>
<img alt="acr_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_acr_layout.png?raw=true" />
</section>
<section id="lineage-specificity-layout">
<h4><a class="toc-backref" href="#id71" role="doc-backlink">Lineage Specificity Layout</a><a class="headerlink" href="#lineage-specificity-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--ls-layout</span></code> is used to visualized the output tree which was executed with <code class="docutils literal notranslate"><span class="pre">--ls-columns</span></code> in previous annotate step. Noted that Lineage Specificity Calculation computes three metric, which is:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">_f1</span></code>: f1 score</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">_prec</span></code>: trait percision</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">_sens</span></code>: trati sensitivity</p></li>
</ul>
<p>Therefore the visulization will generate three layouts for each metrix shown in colorbranch with the score, plus highlighting the clade as lineage specific clade of given trait which passed the <code class="docutils literal notranslate"><span class="pre">--prec-cutoff</span></code> and <code class="docutils literal notranslate"><span class="pre">--sens-cutoff</span></code> in previous step.</p>
<p>We use example from <cite>examples/analytic_example</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Albanian</span><span class="o">.</span><span class="n">tree</span><span class="mf">.152</span><span class="n">tax_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">ls</span><span class="o">-</span><span class="n">layout</span> <span class="n">is_Albania</span> <span class="n">is_Greece</span>
</pre></div>
</div>
<img alt="ls_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_ls_layout1.png?raw=true" />
<img alt="ls_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_ls_layout2.png?raw=true" />
</section>
<section id="ancestral-character-reconstruction-for-continuous-traits">
<h4><a class="toc-backref" href="#id72" role="doc-backlink">Ancestral Character Reconstruction for continuous traits</a><a class="headerlink" href="#ancestral-character-reconstruction-for-continuous-traits" title="Link to this heading">¶</a></h4>
<p>Computing ACR for continuous traits:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Anolis</span><span class="o">.</span><span class="n">tre</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="n">svl</span><span class="o">.</span><span class="n">csv</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">continuous</span><span class="o">-</span><span class="n">columns</span> <span class="n">svl</span> \
<span class="o">--</span><span class="n">prediction</span><span class="o">-</span><span class="n">method</span> <span class="n">BAYESIAN</span> \
<span class="o">--</span><span class="n">model</span> <span class="n">OU</span> \
<span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">stat</span> <span class="n">none</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>


<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">Anolis_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">acr</span><span class="o">-</span><span class="n">continuous</span><span class="o">-</span><span class="n">layout</span> <span class="n">svl</span>
</pre></div>
</div>
<img alt="acr_continuous_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_acr_continuous_layout.png?raw=true" />
</section>
</section>
<section id="layouts-for-multiple-sequence-alignment">
<h3><a class="toc-backref" href="#id73" role="doc-backlink">Layouts for Multiple Sequence Alignment</a><a class="headerlink" href="#layouts-for-multiple-sequence-alignment" title="Link to this heading">¶</a></h3>
<p>In order to visualize multiple sequence alignment alongside with the tree, first we need to annotate alignment using <code class="docutils literal notranslate"><span class="pre">--alignment</span></code> in <code class="docutils literal notranslate"><span class="pre">annotate</span></code>. Then activate alignment layout by adding <code class="docutils literal notranslate"><span class="pre">--alignment-layout</span></code>. Consensus sequence (if calculated) will shown once internal nodes collapsed.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--alignment-layout</span></code></p></td>
<td><p>Display Multiple Sequence Alignment layout in aligned panel.</p></td>
</tr>
</tbody>
</table>
<p>In the previous Functional annotation session we already annotated the tree with multiple sequence alignment, here we will show how to visualize the annotated tree with multiple sequence alignment in <code class="docutils literal notranslate"><span class="pre">examples/pratical_example/emapper</span></code>.</p>
<p>We use example in <code class="docutils literal notranslate"><span class="pre">examples/pratical_example/emapper</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the tree with multiple sequence alignment</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">nifH_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">alignment</span><span class="o">-</span><span class="n">layout</span>
</pre></div>
</div>
<img alt="msa_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_msa_layout.png?raw=true" />
</section>
<section id="layouts-for-eggnog-mapper-pfam-smart-annotations">
<h3><a class="toc-backref" href="#id74" role="doc-backlink">Layouts for eggnog-mapper pfam/smart annotations</a><a class="headerlink" href="#layouts-for-eggnog-mapper-pfam-smart-annotations" title="Link to this heading">¶</a></h3>
<p>In order to visualize eggnog-mapper pfam/smart annotations alongside with the tree, first we need to annotate alignment using <code class="docutils literal notranslate"><span class="pre">--eggnog-mapper</span></code> in <code class="docutils literal notranslate"><span class="pre">annotate</span></code>. Then activate alignment layout by adding <code class="docutils literal notranslate"><span class="pre">--emapper-layout</span></code>.</p>
<p>We use example in <code class="docutils literal notranslate"><span class="pre">examples/pratical_example/emapper</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the tree with eggnog-mapper pfam/smart annotations</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">nifH_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">domain</span><span class="o">-</span><span class="n">layout</span>
</pre></div>
</div>
<img alt="domain_layout example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_domain_layout.png?raw=true" />
</section>
<section id="layouts-for-eggnog-mapper-functional-annotations">
<h3><a class="toc-backref" href="#id75" role="doc-backlink">Layouts for eggnog-mapper functional annotations</a><a class="headerlink" href="#layouts-for-eggnog-mapper-functional-annotations" title="Link to this heading">¶</a></h3>
<p>If metadata is output from eggnog-mapper, using <code class="docutils literal notranslate"><span class="pre">--emapper-annotations</span></code> automatically parse all information as metadata. Program will parse data of all the columns from emapper output. Once tree is annotated, using <code class="docutils literal notranslate"><span class="pre">--emapper-layout</span></code> to visualize tree with all the metadata.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--emapper-layout</span></code></p></td>
<td><p>Activate emapper_layout which displays all the annotation from EggNOG-mapper.</p></td>
</tr>
</tbody>
</table>
<p>We use example in <code class="docutils literal notranslate"><span class="pre">examples/pratical_example/emapper</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the tree with eggnog-mapper functional annotations</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">nifH_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">emapper</span><span class="o">-</span><span class="n">layout</span> \
<span class="o">--</span><span class="n">column</span><span class="o">-</span><span class="n">width</span> <span class="mi">70</span>
</pre></div>
</div>
<p>General metadata from emapper annotation</p>
<img alt="emapper_layout1 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_emapper_layout1.png?raw=true" />
<p>Functional profiling (kegg, ko, gos, etc), collapsed internal nodes shown gradient of presence/total percentage.</p>
<img alt="emapper_layout2 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_emapper_layout2.png?raw=true" />
<img alt="emapper_layout3 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_emapper_layout3.png?raw=true" />
</section>
<section id="layouts-for-taxonomy-annotation">
<h3><a class="toc-backref" href="#id76" role="doc-backlink">Layouts for Taxonomy annotation</a><a class="headerlink" href="#layouts-for-taxonomy-annotation" title="Link to this heading">¶</a></h3>
<p>If target tree was annotated with <code class="docutils literal notranslate"><span class="pre">--taxon-column</span></code> in previous annotate step successfully, now activate Taxonomic layout using the following arguments to visualize taxonomic classification. All rank levels will be generated separately and users can switch each of them on/off.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--taxonclade-layout</span></code></p></td>
<td><p>Activate taxonclade-layout which clades will be colored based on taxonomy of each node.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">--taxonrectangle-layout</span></code></p></td>
<td><p>Activate taxonrectangle-layout which taxonomy of each node will be displayed as rectangular blocks in aligned panel.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--taxoncollapse-layout</span></code></p></td>
<td><p>Activate taxoncollapse-layout which taxonomy of each node will be displayed as rectangular blocks in aligned panel. And nodes will be automatically collapsed.</p></td>
</tr>
</tbody>
</table>
<p>Only tree which was conducted with taxonomic annotation from treeprofiler in previous annotate step can use the taxonomic visualization.</p>
<section id="taxonclade-layout">
<h4><a class="toc-backref" href="#id77" role="doc-backlink">TaxonClade Layout</a><a class="headerlink" href="#taxonclade-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--taxonclade-layout</span></code> will assign different color to each taxa from each rank. Each rank will be a individual layout.</p>
<p>we use example in <code class="docutils literal notranslate"><span class="pre">examples/taxonomy_example</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the pre-annotated example archaea tree with taxonomic classification</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">archaea_annotated</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">taxonclade</span><span class="o">-</span><span class="n">layout</span>
</pre></div>
</div>
<img alt="taxa_layout1 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_taxa_layout1.png?raw=true" />
</section>
<section id="taxonrectangle-layout">
<h4><a class="toc-backref" href="#id78" role="doc-backlink">TaxonRectangle Layout</a><a class="headerlink" href="#taxonrectangle-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--taxonrectangle-layout</span></code> shows taxonomic classification as rectangular block from root to leaf.</p>
<p>we use example in <code class="docutils literal notranslate"><span class="pre">examples/taxonomy_example</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the pre-annotated example archaea tree with taxonomic classification</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">archaea_annotated</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">taxonrectangle</span><span class="o">-</span><span class="n">layout</span>
</pre></div>
</div>
<img alt="taxa_layout2 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_taxa_layout2.png?raw=true" />
</section>
<section id="taxoncollapse-layout">
<h4><a class="toc-backref" href="#id79" role="doc-backlink">TaxonCollapse Layout</a><a class="headerlink" href="#taxoncollapse-layout" title="Link to this heading">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">--taxoncollapse-layout</span></code> is similar to <code class="docutils literal notranslate"><span class="pre">--taxonrectangle-layout</span></code>, but it only show commen ancestor on one column in aligned panel which is shown as corresponding taxa from the rank. Users can switch which layout of rank to be visualized in control panel, once layout of rank is activated, tree will be collapsed based on the according rank. It is more accurate than <code class="docutils literal notranslate"><span class="pre">--taxonrectangle-layout</span></code> because it also shown the representative taxa in corresponding rank.</p>
<p>we use example in <code class="docutils literal notranslate"><span class="pre">examples/taxonomy_example</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the pre-annotated example archaea tree with taxonomic classification</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">archaea_annotated</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">taxoncollapse</span><span class="o">-</span><span class="n">layout</span>
</pre></div>
</div>
<p>Collapse at <em>phylumn</em> level:</p>
<img alt="taxa_layout3 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_taxa_layout3.png?raw=true" />
<p>Collapse at <em>order</em> level:</p>
<img alt="taxa_layout4 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_taxa_layout4.png?raw=true" />
</section>
</section>
<section id="customize-color-in-layout-with-color-config">
<h3><a class="toc-backref" href="#id80" role="doc-backlink">Customize color in layout with color config</a><a class="headerlink" href="#customize-color-in-layout-with-color-config" title="Link to this heading">¶</a></h3>
<p>treeprofiler provides the option to cusomize the color on layouts:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">--color-config</span> <span class="pre">COLOR_CONFIG</span></code></p></td>
<td><p>Path to the file to find the color for each variable. [default: None]</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">CONFIG_SEP,</span> <span class="pre">--config-sep</span> <span class="pre">CONFIG_SEP</span></code></p></td>
<td><p>Column separator of color table. [default: t]</p></td>
</tr>
</tbody>
</table>
<p>A basic template of <code class="docutils literal notranslate"><span class="pre">color.config</span></code> that you can find in <code class="docutils literal notranslate"><span class="pre">examples/custom_color/color.config.template</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>
<span class="c1"># tab-limited or comman-limited</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply. Using hex code or word</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
</pre></div>
</div>
<p>here is the detail of the config:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Description</p></th>
<th class="head"><p>Details</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>PROP</p></td>
<td><p>Property of the annotated tree node</p></td>
</tr>
<tr class="row-odd"><td><p>VALUE</p></td>
<td><p>Value of the property to be colored</p></td>
</tr>
<tr class="row-even"><td><p>COLOR</p></td>
<td><p>Color to apply</p></td>
</tr>
<tr class="row-odd"><td><p>CONDITION</p></td>
<td><p>Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</p></td>
</tr>
</tbody>
</table>
<section id="basic-setting">
<h4><a class="toc-backref" href="#id81" role="doc-backlink">Basic setting</a><a class="headerlink" href="#basic-setting" title="Link to this heading">¶</a></h4>
<p>Here we will use the example from <code class="docutils literal notranslate"><span class="pre">examples/custom_color</span></code> , where the tree <code class="docutils literal notranslate"><span class="pre">basic_example1.nw</span></code> is annotated with <code class="docutils literal notranslate"><span class="pre">basic_example1_metadata1.tsv</span></code> and <code class="docutils literal notranslate"><span class="pre">basic_example1_metadata2.tsv</span></code>. Therefore the output tree is <code class="docutils literal notranslate"><span class="pre">basic_example1_annotated.ete</span></code> and <code class="docutils literal notranslate"><span class="pre">basic_example1_annotated.nw</span></code>.</p>
<p>Check metadata</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="o">-</span><span class="mi">5</span> <span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span> <span class="n">basic_example1_metadata2</span><span class="o">.</span><span class="n">tsv</span>
<span class="o">==&gt;</span> <span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span> <span class="o">&lt;==</span>
<span class="c1">#name       sample1 sample2 sample3 sample4 sample5 random_type     bool_type       bool_type2</span>
<span class="n">Phy003I7ZJ_CHICK</span>    <span class="mf">0.05</span>    <span class="mf">0.12</span>    <span class="mf">0.86</span>    <span class="mf">0.01</span>    <span class="mf">0.69</span>    <span class="n">medium</span>  <span class="mi">1</span>       <span class="n">TRUE</span>
<span class="n">Phy0054BO3_MELGA</span>    <span class="mf">0.64</span>    <span class="mf">0.67</span>    <span class="mf">0.51</span>    <span class="mf">0.29</span>    <span class="mf">0.14</span>    <span class="n">medium</span>  <span class="mi">1</span>       <span class="n">TRUE</span>
<span class="n">Phy00508FR_NIPNI</span>    <span class="mf">0.89</span>    <span class="mf">0.38</span>    <span class="mf">0.97</span>    <span class="mf">0.49</span>    <span class="mf">0.26</span>    <span class="n">low</span>     <span class="mi">1</span>       <span class="n">FALSE</span>
<span class="n">Phy004O1E0_APTFO</span>    <span class="mf">0.1</span>     <span class="mf">0.09</span>    <span class="mf">0.38</span>    <span class="mf">0.31</span>    <span class="mf">0.41</span>    <span class="n">medium</span>  <span class="mi">0</span>       <span class="n">TRUE</span>


<span class="o">==&gt;</span> <span class="n">basic_example1_metadata2</span><span class="o">.</span><span class="n">tsv</span> <span class="o">&lt;==</span>
<span class="c1">#name       abs_data        list_data       abs_data2</span>
<span class="n">Phy003I7ZJ_CHICK</span>    <span class="mi">97</span>      <span class="n">w</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">t</span>   <span class="mi">50</span>
<span class="n">Phy0054BO3_MELGA</span>    <span class="mi">16</span>      <span class="n">r</span><span class="p">,</span><span class="n">q</span><span class="p">,</span><span class="n">s</span>   <span class="mi">245</span>
<span class="n">Phy00508FR_NIPNI</span>    <span class="mi">87</span>      <span class="n">z</span><span class="p">,</span><span class="n">f</span><span class="p">,</span><span class="n">p</span>   <span class="mi">122</span>
<span class="n">Phy004O1E0_APTFO</span>    <span class="mi">6</span>       <span class="n">z</span><span class="p">,</span><span class="n">t</span><span class="p">,</span><span class="n">b</span>   <span class="mi">138</span>
</pre></div>
</div>
<section id="customize-color-for-categorical-data">
<h5><a class="toc-backref" href="#id82" role="doc-backlink">Customize color for categorical data</a><a class="headerlink" href="#customize-color-for-categorical-data" title="Link to this heading">¶</a></h5>
<p>Column <code class="docutils literal notranslate"><span class="pre">random_type</span></code> from the metadata is annotated as categorical data property called <code class="docutils literal notranslate"><span class="pre">random_type</span></code> in result tree <code class="docutils literal notranslate"><span class="pre">basic_example1_annotated.nw</span></code>. By default, all variables in col6 are: <code class="docutils literal notranslate"><span class="pre">high</span></code>, <code class="docutils literal notranslate"><span class="pre">medium</span></code> and <code class="docutils literal notranslate"><span class="pre">low</span></code>, so let’s customize the in <code class="docutils literal notranslate"><span class="pre">color.categorical.config</span></code> file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>
<span class="c1"># tab-limited or comman-limited</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">random_type</span><span class="p">,</span><span class="n">high</span><span class="p">,</span><span class="n">red</span><span class="p">,</span>
<span class="n">random_type</span><span class="p">,</span><span class="n">medium</span><span class="p">,</span><span class="n">blue</span><span class="p">,</span>
<span class="n">random_type</span><span class="p">,</span><span class="n">low</span><span class="p">,</span><span class="c1">#008000,</span>
</pre></div>
</div>
<p>here we can apply this color config to all the layouts for categorical data, for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">rectangle</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">background</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">colorbranch</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">piechart</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">categorical</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<p>Now all the layout will be colored based on the color config file.</p>
<img alt="plot_color_config1 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config1.png?raw=true" />
<p>Noted that if the value is not in the color config file, the default color black will be used.</p>
<p>Missing <code class="docutils literal notranslate"><span class="pre">low</span></code> value in the color config file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>
<span class="c1"># tab-limited or comman-limited</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">random_type</span><span class="p">,</span><span class="n">high</span><span class="p">,</span><span class="n">red</span><span class="p">,</span>
<span class="n">random_type</span><span class="p">,</span><span class="n">medium</span><span class="p">,</span><span class="n">blue</span><span class="p">,</span>
</pre></div>
</div>
<p>rerun the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">rectangle</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">background</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">colorbranch</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">piechart</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">categorical</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config2 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config2.png?raw=true" />
</section>
<section id="customize-color-for-boolean-data">
<h5><a class="toc-backref" href="#id83" role="doc-backlink">Customize color for boolean data</a><a class="headerlink" href="#customize-color-for-boolean-data" title="Link to this heading">¶</a></h5>
<p>Column <code class="docutils literal notranslate"><span class="pre">bool_type</span></code> and <code class="docutils literal notranslate"><span class="pre">bool_type2</span></code> are boolean data, therefore users can only decide the color representing True value. We stored the following config in <code class="docutils literal notranslate"><span class="pre">color.boolean.config</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>
<span class="c1"># tab-limited or comman-limited</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">bool_type</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">blue</span><span class="p">,</span>
<span class="n">bool_type2</span><span class="p">,</span><span class="n">TRUE</span><span class="p">,</span><span class="n">yellow</span><span class="p">,</span>
</pre></div>
</div>
<p>Now we appiled them to the layouts for boolean data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">binary</span><span class="o">-</span><span class="n">layout</span> <span class="n">bool_type</span> <span class="n">bool_type2</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">boolean</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config3 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config3.png?raw=true" />
<p>Noted that for boolean data, the color_config file should only contain the color for True value.</p>
</section>
<section id="customize-color-for-numerical-data">
<h5><a class="toc-backref" href="#id84" role="doc-backlink">Customize color for numerical data</a><a class="headerlink" href="#customize-color-for-numerical-data" title="Link to this heading">¶</a></h5>
<p>We applied the color config to layouts for numerical data in <cite>color.numerical.config</cite> .</p>
<p>We can still customize certain color for specific value.</p>
<p>For example, we want to make make the heatmap layout with yellow color when value is <code class="docutils literal notranslate"><span class="pre">0</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID, COLOR_NAN)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">sample1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">yellow</span><span class="p">,</span>
<span class="n">sample2</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">yellow</span><span class="p">,</span>
<span class="n">sample3</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">yellow</span><span class="p">,</span>
<span class="n">sample4</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">yellow</span><span class="p">,</span>
<span class="n">sample5</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">yellow</span><span class="p">,</span>
</pre></div>
</div>
<p>Then we run it on heatmap layout:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">numerical</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config4 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config4.png?raw=true" />
<p>similar on numerical matrix layout:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">numerical</span><span class="o">-</span><span class="n">matrix</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">numerical</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config5 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config5.png?raw=true" />
</section>
</section>
<section id="advanced-setting">
<h4><a class="toc-backref" href="#id85" role="doc-backlink">Advanced Setting</a><a class="headerlink" href="#advanced-setting" title="Link to this heading">¶</a></h4>
<p>In the Basic setting, we only set the color for specific value, but in the Advanced setting, we can set the color for the range of value.</p>
<section id="setting-color-range-for-numerical-data">
<h5><a class="toc-backref" href="#id86" role="doc-backlink">Setting color range for numerical data</a><a class="headerlink" href="#setting-color-range-for-numerical-data" title="Link to this heading">¶</a></h5>
<p>We can set the color range using <code class="docutils literal notranslate"><span class="pre">COLOR_MIN</span></code>, <code class="docutils literal notranslate"><span class="pre">COLOR_MAX</span></code>, <code class="docutils literal notranslate"><span class="pre">COLOR_MID</span></code> or <code class="docutils literal notranslate"><span class="pre">COLOR_NAN</span></code> under the column <code class="docutils literal notranslate"><span class="pre">CONDITION</span></code>.</p>
<p>Here we set the color for the range of value in <cite>color.numerical2.config</cite>, noted that now the <code class="docutils literal notranslate"><span class="pre">VALUE</span></code> column is empty therefore the maximum value will be assigned to <code class="docutils literal notranslate"><span class="pre">COLOR_MAX</span></code>, the minimum value will be assigned to <code class="docutils literal notranslate"><span class="pre">COLOR_MIN</span></code>, and the middle value will be assigned to <code class="docutils literal notranslate"><span class="pre">COLOR_MID</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">sample1</span><span class="p">,,</span><span class="n">red</span><span class="p">,</span><span class="n">COLOR_MAX</span>
<span class="n">sample1</span><span class="p">,,</span><span class="n">green</span><span class="p">,</span><span class="n">COLOR_MID</span>
<span class="n">sample1</span><span class="p">,,</span><span class="n">blue</span><span class="p">,</span><span class="n">COLOR_MIN</span>
</pre></div>
</div>
<p>Then we run it on heatmap layout:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">numerical2</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<p>As you see, the color is assigned based on the range of value of property <code class="docutils literal notranslate"><span class="pre">sample1</span></code>:</p>
<img alt="plot_color_config6 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config6.png?raw=true" />
</section>
<section id="combining-settings-of-value-and-color-range-for-numerical-data">
<h5><a class="toc-backref" href="#id87" role="doc-backlink">Combining settings of value and color range for numerical data</a><a class="headerlink" href="#combining-settings-of-value-and-color-range-for-numerical-data" title="Link to this heading">¶</a></h5>
<p>In the previous example, we set the color range for the values, the value range is the range of minimum and maximum value by default, but we can also cusomize the value together with the color conditions</p>
<p>Here we set the color for the range of value in <cite>color.numerical3.config</cite>, noted that now the <code class="docutils literal notranslate"><span class="pre">VALUE</span></code> column is not empty as we set the range from -5 to 5, therefore the value will be assigned to the corresponding color:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">sample1</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="n">red</span><span class="p">,</span><span class="n">COLOR_MAX</span>
<span class="n">sample1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">green</span><span class="p">,</span><span class="n">COLOR_MID</span>
<span class="n">sample1</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="n">blue</span><span class="p">,</span><span class="n">COLOR_MIN</span>
</pre></div>
</div>
<p>Then we run it on <code class="docutils literal notranslate"><span class="pre">heatmap-layout</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">numerical3</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config7 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config7.png?raw=true" />
<p>If we only want to set the range of value without customize the color, is also possible:</p>
<p>Here we want to set up the range from -5 to 5, but we only set them as <code class="docutils literal notranslate"><span class="pre">CONDITION</span></code> without setting specific color in <code class="docutils literal notranslate"><span class="pre">COLOR</span></code> column:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">sample1</span><span class="p">,</span><span class="mi">5</span><span class="p">,,</span><span class="n">COLOR_MAX</span>
<span class="n">sample1</span><span class="p">,</span><span class="mi">0</span><span class="p">,,</span><span class="n">COLOR_MID</span>
<span class="n">sample1</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,,</span><span class="n">COLOR_MIN</span>
</pre></div>
</div>
<p>Rerun:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">numerical3</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config8 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config8.png?raw=true" />
</section>
<section id="wildcard-settings-of-color-range-for-numerical-matrix">
<h5><a class="toc-backref" href="#id88" role="doc-backlink">Wildcard settings of color range for numerical matrix</a><a class="headerlink" href="#wildcard-settings-of-color-range-for-numerical-matrix" title="Link to this heading">¶</a></h5>
<p>In the session of Layouts for numerical data, we describe the difference between <code class="docutils literal notranslate"><span class="pre">--heatmap-layout</span></code> and <code class="docutils literal notranslate"><span class="pre">--numerical-matrix-layout</span></code>. In short, <code class="docutils literal notranslate"><span class="pre">--heatmap-layout</span></code> treat each property individually therefore the range each column of heatmap is different, but <code class="docutils literal notranslate"><span class="pre">--numerical-matrix-layout</span></code> treat the whole matrix as a single property, therefore the range of value is the same for all columns.</p>
<p>Therefore, when we set the color config for the numerical matrix layout, we can use the wildcard <code class="docutils literal notranslate"><span class="pre">*</span></code> to represent all the columns in the matrix.</p>
<p>Here we set the color for the range of value in <cite>color.datamatrix.config</cite>, noted that now the <code class="docutils literal notranslate"><span class="pre">PROP</span></code> column is <code class="docutils literal notranslate"><span class="pre">*</span></code> which means all the columns in the matrix, therefore the value will be assigned to the corresponding color:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Color Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: Special case for numerical values (optional, e.g., COLOR_MIN, COLOR_MAX, COLOR_MID)</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="o">*</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="n">red</span><span class="p">,</span><span class="n">COLOR_MAX</span>
<span class="o">*</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">white</span><span class="p">,</span><span class="n">COLOR_MID</span>
<span class="o">*</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="n">blue</span><span class="p">,</span><span class="n">COLOR_MIN</span>
</pre></div>
</div>
<p>now we run it on <code class="docutils literal notranslate"><span class="pre">numerical-matrix-layout</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">numerical</span><span class="o">-</span><span class="n">matrix</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">datamatrix</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config9 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config9.png?raw=true" />
<p>It applied in also for data matrix properties:</p>
<p>Here we annotate and plot the <code class="docutils literal notranslate"><span class="pre">data_2col.array``(a</span> <span class="pre">data</span> <span class="pre">matrix</span> <span class="pre">file)</span> <span class="pre">to</span> <span class="pre">demo</span> <span class="pre">tree</span> <span class="pre">``demo1.tree</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1</span><span class="o">.</span><span class="n">tree</span> \
<span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="n">matrix</span> <span class="n">data_2col</span><span class="o">.</span><span class="n">array</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">demo1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">numerical</span><span class="o">-</span><span class="n">matrix</span><span class="o">-</span><span class="n">layout</span> <span class="n">data_2col</span><span class="o">.</span><span class="n">array</span> \
<span class="o">--</span><span class="n">color</span><span class="o">-</span><span class="n">config</span> <span class="n">color</span><span class="o">.</span><span class="n">datamatrix</span><span class="o">.</span><span class="n">config</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="plot_color_config10 example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_color_config10.png?raw=true" />
</section>
<section id="customize-color-settings-with-complex-conditions">
<h5><a class="toc-backref" href="#id89" role="doc-backlink">Customize color settings with complex conditions</a><a class="headerlink" href="#customize-color-settings-with-complex-conditions" title="Link to this heading">¶</a></h5>
<p>TBD</p>
</section>
</section>
</section>
</section>
<section id="conditional-query-in-annotated-tree-can-be-used-in-both-annotate-and-plot">
<h2><a class="toc-backref" href="#id90" role="doc-backlink">Conditional query in annotated tree (can be used in both annotate and plot)</a><a class="headerlink" href="#conditional-query-in-annotated-tree-can-be-used-in-both-annotate-and-plot" title="Link to this heading">¶</a></h2>
<p>TreeProfiler allows users to perform conditional process based on different circumstances:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Command Option</p></th>
<th class="head"><p>Subcommand Availability</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>–pruned-by</p></td>
<td><p>annotate, plot</p></td>
<td><p>Prunes the annotated tree based on specific conditions, removing branches or clades that do not meet the criteria.</p></td>
</tr>
<tr class="row-odd"><td><p>–rank-limit</p></td>
<td><p>annotate, plot</p></td>
<td><p>Enables pruning of a taxonomically annotated tree based on the rank of classification.</p></td>
</tr>
<tr class="row-even"><td><p>–collapsed-by</p></td>
<td><p>plot</p></td>
<td><p>Allows users to collapse tree branches based on custom conditions, mainly focusing on internal nodes.</p></td>
</tr>
<tr class="row-odd"><td><p>–highlighted-by</p></td>
<td><p>plot</p></td>
<td><p>Enables users to highlight tree nodes that meet specific conditions.</p></td>
</tr>
<tr class="row-even"><td><p>–internal-plot-measure {sum,avg,max,min,std,none}</p></td>
<td><p>plot</p></td>
<td><p>statistic measures to be shown in numerical layout for internal nodes. [default: avg]</p></td>
</tr>
</tbody>
</table>
<section id="query-syntax">
<h3><a class="toc-backref" href="#id91" role="doc-backlink">Query Syntax</a><a class="headerlink" href="#query-syntax" title="Link to this heading">¶</a></h3>
<p>In this session we will use examples in <code class="docutils literal notranslate"><span class="pre">examples/automatic_query</span></code></p>
<section id="basic-query-syntax">
<h4><a class="toc-backref" href="#id92" role="doc-backlink">Basic Query Syntax</a><a class="headerlink" href="#basic-query-syntax" title="Link to this heading">¶</a></h4>
<p>All the conditional query shared the same syntax, a standard query consists the following</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">--</span><span class="n">pruned</span><span class="o">-</span><span class="n">by</span><span class="o">|</span><span class="n">collapsed</span><span class="o">-</span><span class="n">by</span><span class="o">|</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;&lt;left_value&gt; &lt;operator&gt; &lt;right_value&gt;&quot;</span>
</pre></div>
</div>
<p>Here is the detail of the query syntax:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>Left Value</p></th>
<th class="head"><p>Operator</p></th>
<th class="head"><p>Right Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Description</p></td>
<td><p>This is the property of either a leaf node or an internal node within the tree. It could be any metadata feature linked to the node, such as its taxonomic classification, gene expression levels, or other biological markers. For example, <code class="docutils literal notranslate"><span class="pre">name</span></code>, <code class="docutils literal notranslate"><span class="pre">sci_name</span></code>, <code class="docutils literal notranslate"><span class="pre">support</span></code>.</p></td>
<td><p>This part of the query specifies the relationship between the left and right values. Operators include: <code class="docutils literal notranslate"><span class="pre">=</span></code>, <code class="docutils literal notranslate"><span class="pre">!=</span></code>, <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>, <code class="docutils literal notranslate"><span class="pre">&gt;=</span></code>, <code class="docutils literal notranslate"><span class="pre">&lt;</span></code>, <code class="docutils literal notranslate"><span class="pre">&lt;=</span></code>, and <code class="docutils literal notranslate"><span class="pre">contains</span></code>.</p></td>
<td><p>This is the custom criterion against which the left value is compared. Depending on the nature of the left value, this could be a numerical figure, a string, or even a list of values.</p></td>
</tr>
</tbody>
</table>
<section id="conditional-pruning">
<h5><a class="toc-backref" href="#id93" role="doc-backlink">Conditional pruning</a><a class="headerlink" href="#conditional-pruning" title="Link to this heading">¶</a></h5>
<p>Conditional pruning, prune leaf node whose name contain “FALPE”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## annotate tree</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="o">./</span><span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">--</span><span class="n">counter</span><span class="o">-</span><span class="n">stat</span> <span class="n">relative</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># Conditional pruning, prune leaf node whose name contain &quot;FALPE&quot;</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">pruned</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;name contains FALPE&quot;</span>
</pre></div>
</div>
<p>Left panel is tree before prune, right panel is result after prune:</p>
<img alt="pruned" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/prune.png?raw=true" />
</section>
<section id="conditional-highlight">
<h5><a class="toc-backref" href="#id94" role="doc-backlink">Conditional highlight</a><a class="headerlink" href="#conditional-highlight" title="Link to this heading">¶</a></h5>
<p>select tree node whose name contains <cite>FALPE</cite> character:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## annotate tree</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">metadata</span> <span class="o">./</span><span class="n">basic_example1_metadata1</span><span class="o">.</span><span class="n">tsv</span> \
<span class="o">--</span><span class="n">counter</span><span class="o">-</span><span class="n">stat</span> <span class="n">relative</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>

<span class="c1"># select tree node whose name contains `FALPE` character</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">newick</span> \
<span class="o">--</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;name contains FALPE&quot;</span>
</pre></div>
</div>
<img alt="highlighted" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/highlighted.jpeg?raw=true" />
<p>The same annotated tree, but select tree node whose sample1 feature &gt; 0.50, here we using ete format which can resume the datatype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;sample1 &gt; 0.50&quot;</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span>
</pre></div>
</div>
<p>if use tree in newick format, we need to attach the prop2type file which can resume the datatype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">newick</span> \
<span class="o">--</span><span class="n">prop2type</span> <span class="n">basic_example1_prop2type</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;sample1 &gt; 0.50&quot;</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span>
</pre></div>
</div>
<img alt="highlighted_numeric" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/highlighted_numeric.png?raw=true" />
</section>
</section>
<section id="query-in-internal-nodes">
<h4><a class="toc-backref" href="#id95" role="doc-backlink">Query in internal nodes</a><a class="headerlink" href="#query-in-internal-nodes" title="Link to this heading">¶</a></h4>
<p>Query in internal nodes’ properties is also available, in this case, <code class="docutils literal notranslate"><span class="pre">left_value</span></code> of query will be the internal node property, remember to add the proper suffixes such as <code class="docutils literal notranslate"><span class="pre">_avg</span></code>, <code class="docutils literal notranslate"><span class="pre">_max</span></code>,etc, for the numerical data or <cite>_counter</cite> for categorical and boolean data.</p>
<p>Syntax for internal node numerical data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select tree internal node where sample1_avg feature &lt; 0.50</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">collapsed</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;sample1_avg &lt; 0.50&quot;</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span>
</pre></div>
</div>
<img alt="collapsed_numeric" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/collapsed_numeric.png?raw=true" />
<p>Syntax for internal node categorical data in counter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># collapse tree internal nodes, where `high` relative counter &gt; 0.35 in random_type_counter property</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">rectangle</span><span class="o">-</span><span class="n">layout</span> <span class="n">random_type</span> \
<span class="o">--</span><span class="n">collapsed</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;random_type_counter:high &gt; 0.35&quot;</span> \
<span class="o">--</span><span class="n">column</span><span class="o">-</span><span class="n">width</span> <span class="mi">70</span>
</pre></div>
</div>
<img alt="collapsed_counter" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/collapsed_counter.png?raw=true" />
</section>
<section id="and-and-or-conditions">
<h4><a class="toc-backref" href="#id96" role="doc-backlink">AND and OR conditions</a><a class="headerlink" href="#and-and-or-conditions" title="Link to this heading">¶</a></h4>
<p>The syntax for the AND condition and OR condition in TreeProfiler is:</p>
<p>AND condition will be under one argument, syntax seperated by <code class="docutils literal notranslate"><span class="pre">,</span></code>, such as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select tree  node where sample1 feature &gt; 0.50 AND sample2 &lt; 0.2</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;sample1&gt;0.50,sample2&lt;0.2&quot;</span>
</pre></div>
</div>
<img alt="highlighted_and" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/highlighted_and.png?raw=true" />
<p>OR condition will be used more than one arguments:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select tree node where sample1 feature &gt; 0.50 OR sample2 &lt; 0.2</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span> <span class="n">sample2</span> <span class="n">sample3</span> <span class="n">sample4</span> <span class="n">sample5</span> \
<span class="o">--</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;sample1&gt;0.50&quot;</span> \
<span class="o">--</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="s2">&quot;sample2&lt;0.2&quot;</span>
</pre></div>
</div>
<img alt="highlighted_or" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/highlighted_or.png?raw=true" />
</section>
<section id="conditional-query-with-config-file">
<h4><a class="toc-backref" href="#id97" role="doc-backlink">Conditional query with config file</a><a class="headerlink" href="#conditional-query-with-config-file" title="Link to this heading">¶</a></h4>
<p>In conditinary query, it also accept to use config to customize the highlighted and collapsed visualization. Example config file in <cite>color.config.template</cite>. It has the same structure as color config but the CONDITION column here is for operators as we mentioned in previous session.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cat color.config.template`
# TreeProfiler Conditional query Configuration
# This file defines custom colors for properties of annotated tree nodes.
# Columns: PROP, VALUE, COLOR, CONDITION

# PROP: Property of the annotated tree node
# VALUE: Value of the property to be colored
# COLOR: Color to apply
# CONDITION: operator

PROP,VALUE,COLOR,CONDITION
</pre></div>
</div>
<section id="highlight-with-custom-config">
<h5><a class="toc-backref" href="#id98" role="doc-backlink">Highlight with custom config</a><a class="headerlink" href="#highlight-with-custom-config" title="Link to this heading">¶</a></h5>
<p>we use example of <code class="docutils literal notranslate"><span class="pre">basic_example1.nw</span></code> and <code class="docutils literal notranslate"><span class="pre">basic_example1_metadata1.tsv</span></code>, we hope to highlight nodes whose <cite>random_type</cite> is <cite>low</cite> AND <cite>sample1</cite> greater than <cite>0.50</cite>, and we want to highlight those nodes in red.</p>
<p>First we customize the config <code class="docutils literal notranslate"><span class="pre">color.config.query</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Conditional query Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: operator</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="n">random_type</span><span class="p">,</span><span class="n">low</span><span class="p">,</span><span class="n">red</span><span class="p">,</span><span class="o">=</span>
<span class="n">sample1</span><span class="p">,</span><span class="mf">0.50</span><span class="p">,</span><span class="n">red</span><span class="p">,</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>we attach the config using the same <code class="docutils literal notranslate"><span class="pre">--highlighted-by</span></code>, dont forget about the <code class="docutils literal notranslate"><span class="pre">-s</span></code> to indicate the seperator</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">highlighted</span><span class="o">-</span><span class="n">by</span> <span class="n">color</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">query</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
<img alt="highlighted_query" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_query_config1.png?raw=true" />
</section>
<section id="collapse-with-custom-config">
<h5><a class="toc-backref" href="#id99" role="doc-backlink">Collapse with custom config</a><a class="headerlink" href="#collapse-with-custom-config" title="Link to this heading">¶</a></h5>
<p>Similar to <code class="docutils literal notranslate"><span class="pre">--highlighted-by</span></code>, in this case we try to collapse the internal nodes whose average of leaf properties <cite>sample1</cite> greater than <cite>0.50</cite>, with color of green.</p>
<p>The config is in <code class="docutils literal notranslate"><span class="pre">color.config.collapsed</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TreeProfiler Conditional query Configuration</span>
<span class="c1"># This file defines custom colors for properties of annotated tree nodes.</span>
<span class="c1"># Columns: PROP, VALUE, COLOR, CONDITION</span>

<span class="c1"># PROP: Property of the annotated tree node</span>
<span class="c1"># VALUE: Value of the property to be colored</span>
<span class="c1"># COLOR: Color to apply</span>
<span class="c1"># CONDITION: operator</span>

<span class="n">PROP</span><span class="p">,</span><span class="n">VALUE</span><span class="p">,</span><span class="n">COLOR</span><span class="p">,</span><span class="n">CONDITION</span>
<span class="c1">#random_type_counter:high,0.35,red,&gt;</span>
<span class="n">sample1_avg</span><span class="p">,</span><span class="mf">0.50</span><span class="p">,</span><span class="n">green</span><span class="p">,</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>example run</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">-</span><span class="n">t</span> <span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">collapsed</span><span class="o">-</span><span class="n">by</span> <span class="n">color</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">collapsed</span> \
<span class="o">-</span><span class="n">s</span> <span class="p">,</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="conditional-limit-based-on-taxonomic-level">
<h3><a class="toc-backref" href="#id100" role="doc-backlink">Conditional limit based on taxonomic level</a><a class="headerlink" href="#conditional-limit-based-on-taxonomic-level" title="Link to this heading">¶</a></h3>
<p>Prune taxonomic annotated tree based on following taxonomic rank level,
<code class="docutils literal notranslate"><span class="pre">kingdom</span></code>, <code class="docutils literal notranslate"><span class="pre">phylum</span></code>, <code class="docutils literal notranslate"><span class="pre">class</span></code>, <code class="docutils literal notranslate"><span class="pre">order</span></code>, <code class="docutils literal notranslate"><span class="pre">family</span></code>, <code class="docutils literal notranslate"><span class="pre">genus</span></code>, <code class="docutils literal notranslate"><span class="pre">species</span></code>, <code class="docutils literal notranslate"><span class="pre">subspecies</span></code></p>
<p>case in GTDB in <code class="docutils literal notranslate"><span class="pre">examples/taxonomy_example/</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># before pruning</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">archaea_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">taxonclade</span><span class="o">-</span><span class="n">layout</span>
</pre></div>
</div>
<p>Before rank limit</p>
<img alt="prunt" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_taxaprune1.png?raw=true" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># before pruning</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">archaea_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">taxonclade</span><span class="o">-</span><span class="n">layout</span>
</pre></div>
</div>
<p>After rank_limit</p>
<img alt="prunt" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_taxaprune2.png?raw=true" />
<p>As you see, class branches of target gtdb tree are all pruned and only left the internal branches which rank as class.</p>
</section>
<section id="choose-the-metric-for-internal-nodes-in-numerical-layout">
<h3><a class="toc-backref" href="#id101" role="doc-backlink">Choose the metric for internal nodes in numerical layout</a><a class="headerlink" href="#choose-the-metric-for-internal-nodes-in-numerical-layout" title="Link to this heading">¶</a></h3>
<p>Internal nodes of numerical data are process descriptive statistic analysis by default, hence when users collapse any branch, barplot_layout or heatmap_layout will demonstrate representative value, <code class="docutils literal notranslate"><span class="pre">avg</span></code> by default. Representative value can be changed by using <code class="docutils literal notranslate"><span class="pre">--internal-plot-measure</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select max instead of avg as internal node plotting representative</span>
<span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">examples</span><span class="o">/</span><span class="n">basic_example1</span><span class="o">/</span><span class="n">basic_example1_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">sample1</span><span class="p">,</span><span class="n">sample2</span><span class="p">,</span><span class="n">sample3</span><span class="p">,</span><span class="n">sample4</span><span class="p">,</span><span class="n">sample5</span> \
<span class="o">--</span><span class="n">internal</span><span class="o">-</span><span class="n">plot</span><span class="o">-</span><span class="n">measure</span> <span class="nb">max</span>
</pre></div>
</div>
<p>Before collapsed</p>
<img alt="heatmap_uncollapsed" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_heatmap_uncollapsed.jpeg?raw=true" />
<p>After collapsed
avg as internal plot measure</p>
<img alt="heatmap_collapsed" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_heatmap_collapsed.jpeg?raw=true" />
<p>max as internal plot measure</p>
<img alt="heatmap_max_collapsed" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/plot_heatmap_collapsed_max.jpeg?raw=true" />
</section>
</section>
<section id="demo1-explore-gtdb-taxonomic-tree-with-metadata-and-habitat-information-of-progenome3">
<h2><a class="toc-backref" href="#id102" role="doc-backlink">Demo1 Explore GTDB taxonomic tree with metadata and habitat information of progenome3</a><a class="headerlink" href="#demo1-explore-gtdb-taxonomic-tree-with-metadata-and-habitat-information-of-progenome3" title="Link to this heading">¶</a></h2>
<p>To illustrate the easiness and flexibility of TreeProfiler, we use it to annotate and visualize the version 202 of the GTDB prokaryotic phylogeny, which represents a species tree with 60,000 representative bacterial and archaeal species in <a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release202/">here</a>. GTDB provides the tree in plain newick format and massive datatable with various associated to such species. Apart from the metadata provided by the GTDB, here we also include annotations of genomes and species clusters to habitats from proGenomes3 (<a class="reference external" href="https://progenomes.embl.de/">Fullam et al. 2023</a>).</p>
<p>Example can be found in directory <code class="docutils literal notranslate"><span class="pre">./examples/pratical/gtdb_r202/</span></code>. We already prepared the gtdb v202 taxonomic tree <code class="docutils literal notranslate"><span class="pre">gtdbv202.nw</span></code> by merging Bacteria and Archaea trees, detailed steps are included in <code class="docutils literal notranslate"><span class="pre">merge_gtdbtree.py</span></code>. Based on the difference of computational capacity, complete steps and pipeline can be found in <code class="docutils literal notranslate"><span class="pre">gtdbv202full_demo.sh</span></code> and <code class="docutils literal notranslate"><span class="pre">gtdbv202lite_demo.sh</span></code>.</p>
<p>1) A glance of habitat information of progenome3</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">examples</span><span class="o">/</span><span class="n">pratical</span><span class="o">/</span><span class="n">gtdb_r202</span><span class="o">/</span>

<span class="n">zcat</span> <span class="n">progenome3</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span><span class="o">|</span><span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="mi">4</span>
<span class="n">progenome3</span><span class="o">.</span><span class="n">tsv0000664000175000017500000343216414447266674012351</span> <span class="mi">0</span><span class="n">ustar</span>  <span class="n">dengdenggtdb_genome_representative</span>  <span class="n">aquatic_habitat</span> <span class="n">host_associated</span> <span class="n">soil_habitat</span>
<span class="n">RS_GCF_004210275</span><span class="mf">.1</span>  <span class="n">f</span>       <span class="n">t</span>       <span class="n">f</span>
<span class="n">GB_GCA_014116815</span><span class="mf">.1</span>
<span class="n">RS_GCF_000730245</span><span class="mf">.1</span>  <span class="n">f</span>       <span class="n">t</span>       <span class="n">f</span>
</pre></div>
</div>
<p>2) Download metadata of archaea and bacteria from gtdb</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">data</span><span class="o">.</span><span class="n">gtdb</span><span class="o">.</span><span class="n">ecogenomic</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">releases</span><span class="o">/</span><span class="n">release202</span><span class="o">/</span><span class="mf">202.0</span><span class="o">/</span><span class="n">ar122_metadata_r202</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
<span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">data</span><span class="o">.</span><span class="n">gtdb</span><span class="o">.</span><span class="n">ecogenomic</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">releases</span><span class="o">/</span><span class="n">release202</span><span class="o">/</span><span class="mf">202.0</span><span class="o">/</span><span class="n">bac120_metadata_r202</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p>GTDB metadata annotation</p></li>
</ol>
<p>Considering the size of GTDB metadata and phylogeney, here we provide two pipelines for user to choose base on their computational resources.</p>
<ul class="simple">
<li><p>GTDB partial annotation (lightweight), which we will extract only a few of columns from metadata for annotation</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract genome_size, protein_count, gc_percentage, ncbi_assembly_level, ncbi_genome_category columns from GTDB metadata</span>
<span class="n">tar</span> <span class="o">-</span><span class="n">xf</span> <span class="n">ar122_metadata_r202</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">O</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f1</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">89</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">46</span><span class="p">,</span><span class="mi">56</span> <span class="o">&gt;</span> <span class="n">ar122_metadata_r202_lite</span><span class="o">.</span><span class="n">tsv</span>
<span class="n">tar</span> <span class="o">-</span><span class="n">xf</span> <span class="n">bac120_metadata_r202</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">O</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f1</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">89</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">46</span><span class="p">,</span><span class="mi">56</span> <span class="o">&gt;</span> <span class="n">bac120_metadata_r202_lite</span><span class="o">.</span><span class="n">tsv</span>

<span class="c1"># start annotation</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">gtdbv202</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">newick</span> \
<span class="o">--</span><span class="n">metadata</span> \
<span class="n">ar122_metadata_r202_lite</span><span class="o">.</span><span class="n">tsv</span> <span class="n">bac120_metadata_r202_lite</span><span class="o">.</span><span class="n">tsv</span> <span class="n">progenome3</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">gtdb</span><span class="o">-</span><span class="n">version</span> <span class="mi">202</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">GTDB</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>
</pre></div>
</div>
<ul class="simple">
<li><p>GTDB full annotation, which requires <strong>&gt;6G disk space and &gt;15G RAM memory</strong>.</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Annotate metadatas to taxonomic tree(this step may take a few minutes)</span>
<span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">gtdbv202</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">newick</span> \
<span class="o">--</span><span class="n">metadata</span> \
<span class="n">ar122_metadata_r202</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="n">bac120_metadata_r202</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="n">progenome3</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">gtdb</span><span class="o">-</span><span class="n">version</span> <span class="mi">202</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">GTDB</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>
</pre></div>
</div>
<p>4) Visualizing annotated GTDB tree with GTDB metadata, which are
- <code class="docutils literal notranslate"><span class="pre">genome_size</span></code>
- <code class="docutils literal notranslate"><span class="pre">protein_count</span></code>
- <code class="docutils literal notranslate"><span class="pre">gc_percentage</span></code>
- <code class="docutils literal notranslate"><span class="pre">ncbi_assembly_level</span></code>
- <code class="docutils literal notranslate"><span class="pre">ncbi_genome_category</span></code></p>
<p>and progenome3 habitat information
- <code class="docutils literal notranslate"><span class="pre">aquatic_habitat</span></code>
- <code class="docutils literal notranslate"><span class="pre">host_associated</span></code>
- <code class="docutils literal notranslate"><span class="pre">soil_habitat</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">gtdbv202_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">type</span> <span class="n">ete</span> \
<span class="o">--</span><span class="n">barplot</span><span class="o">-</span><span class="n">layout</span> <span class="n">genome_size</span> <span class="n">protein_count</span> \
<span class="o">--</span><span class="n">heatmap</span><span class="o">-</span><span class="n">layout</span> <span class="n">gc_percentage</span> \
<span class="o">---</span><span class="n">binary</span><span class="o">-</span><span class="n">layout</span> <span class="n">aquatic_habitat</span> <span class="n">host_associated</span> <span class="n">soil_habitat</span> \
<span class="o">--</span><span class="n">rectangle</span><span class="o">-</span><span class="n">layout</span> <span class="n">ncbi_assembly_level</span> <span class="n">ncbi_genome_category</span> \
<span class="o">--</span><span class="n">taxonclade</span><span class="o">-</span><span class="n">layout</span> \
<span class="o">--</span><span class="n">column</span><span class="o">-</span><span class="n">width</span> <span class="mi">70</span>
</pre></div>
</div>
<img alt="GTDB v202 General" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/gtdb_v4.png?raw=true" />
<p>Here we show the GTDB v202 taxonomy tree (bacteria+archaea, 47894 leaves) in rectangular tree layout, with selected annotated properties which are displayed by order in aligned panel. Numerical data <code class="docutils literal notranslate"><span class="pre">genome_size</span></code> and <code class="docutils literal notranslate"><span class="pre">protein_count</span></code> are visualized as barplot, <code class="docutils literal notranslate"><span class="pre">gc_percentage</span></code> is shown as heatmap. Habitat information of progenome3, <code class="docutils literal notranslate"><span class="pre">aquatic_habitat</span></code>, <code class="docutils literal notranslate"><span class="pre">host_associated</span></code> and <code class="docutils literal notranslate"><span class="pre">soil_habitat</span></code> are shown as binary layout. Two categorical data <code class="docutils literal notranslate"><span class="pre">ncbi_assembly_level</span></code> and <code class="docutils literal notranslate"><span class="pre">ncbi_genome_category</span></code> are visualized as rectangular layout. In order to improve memory efficiency, tree has default collapse level (10) hence multiple leaf nodes are collapsed as default. If nodes are collapsed, aligned layouts represented corresponding values of each property of annotated internal nodes. In this level, <code class="docutils literal notranslate"><span class="pre">taxonclade-layout</span></code> of the highest classification <code class="docutils literal notranslate"><span class="pre">kingdom</span></code> is activated, which demonstrates <code class="docutils literal notranslate"><span class="pre">bacteria</span></code> in salmon, <code class="docutils literal notranslate"><span class="pre">archaea</span></code> in blue.</p>
<img alt="GTDB v202 Closeup" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/gtdb_v5.png?raw=true" />
<p>Once zoom into smaller view in tree, collapse level reduces automatically (or manually) to 1, then leaf nodes are dynamically displayed and rendered. Therefore associated layouts are shown as representing values of annotated leaves. <code class="docutils literal notranslate"><span class="pre">taxonclade-layout</span></code> colored leaf nodes in <code class="docutils literal notranslate"><span class="pre">species</span></code> rank level.</p>
<img alt="Progenome3 Example" src="https://raw.githubusercontent.com/dengzq1234/treeprofiler_gallery/main/progenome_example.jpeg" />
<p>Annotated tree and layouts can be shown as circular tree layout.</p>
</section>
<section id="demo2-explore-large-nifh-gene-tree-with-functional-and-taxonomic-information">
<h2><a class="toc-backref" href="#id103" role="doc-backlink">Demo2 Explore large NifH gene tree with functional and taxonomic information</a><a class="headerlink" href="#demo2-explore-large-nifh-gene-tree-with-functional-and-taxonomic-information" title="Link to this heading">¶</a></h2>
<p>Here we analyzed the nitrogenase iron protein NifH gene family across bacteria from EggNOG6 with EggNOG-mapper, a tool for functional annotation based on precomputed orthology assignments. TreeProfiler provides options which allows users to directly map EggNOG-mapper outputs including functional annotations and pfam/smart domain predictions. Hence are then able to map these functional annotations to their respective phylogenetic gene trees and them with the evolutionary history, tracing from leaf to root level.</p>
<p>Map emapper annotation, pfam annotation and taxonomic annotation to target tree.</p>
<p>Here we use example in <code class="docutils literal notranslate"><span class="pre">examples/pratical_example/emapper/</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">annotate</span> \
<span class="o">--</span><span class="n">tree</span>  <span class="n">nifH</span><span class="o">.</span><span class="n">nw</span> \
<span class="o">--</span><span class="n">emapper</span><span class="o">-</span><span class="n">annotation</span> <span class="n">nifH</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">emapper</span><span class="o">.</span><span class="n">annotations</span> \
<span class="o">--</span><span class="n">emapper</span><span class="o">-</span><span class="n">pfam</span> <span class="n">nifH</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">emapper</span><span class="o">.</span><span class="n">pfam</span> \
<span class="o">--</span><span class="n">alignment</span> <span class="n">nifH</span><span class="o">.</span><span class="n">faa</span><span class="o">.</span><span class="n">aln</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">column</span> <span class="n">name</span> \
<span class="o">--</span><span class="n">taxon</span><span class="o">-</span><span class="n">delimiter</span> <span class="o">.</span> \
<span class="o">--</span><span class="n">taxa</span><span class="o">-</span><span class="n">field</span> <span class="mi">0</span> \
<span class="o">--</span><span class="n">taxadb</span> <span class="n">NCBI</span> \
<span class="o">-</span><span class="n">o</span> <span class="o">./</span>
</pre></div>
</div>
<p>Visualize annotations of emapper, pfam domain and ncbi taxonomy</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">treeprofiler</span> <span class="n">plot</span> \
<span class="o">--</span><span class="n">tree</span> <span class="n">nifH_annotated</span><span class="o">.</span><span class="n">ete</span> \
<span class="o">--</span><span class="n">emapper</span><span class="o">-</span><span class="n">layout</span> \
<span class="o">--</span><span class="n">domain</span><span class="o">-</span><span class="n">layout</span> \
<span class="o">--</span><span class="n">taxonclade</span><span class="o">-</span><span class="n">layout</span> \
<span class="o">--</span><span class="n">column</span><span class="o">-</span><span class="n">width</span> <span class="mi">70</span>
</pre></div>
</div>
<p>Visualization of categorical data <code class="docutils literal notranslate"><span class="pre">seed_orthologs</span></code>, <code class="docutils literal notranslate"><span class="pre">max_annot_lvl</span></code>, <code class="docutils literal notranslate"><span class="pre">COG_category</span></code>, <code class="docutils literal notranslate"><span class="pre">Description</span></code>, <code class="docutils literal notranslate"><span class="pre">Preferred_name</span></code>, and numerical data <code class="docutils literal notranslate"><span class="pre">score</span></code></p>
<img alt="emapper example" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/emapper_nifh_v1.png?raw=true" />
<p>Visualization of <code class="docutils literal notranslate"><span class="pre">KEGG_Pathway</span></code> profiling</p>
<img alt="emapper example2" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/emapper_nifh_v2.png?raw=true" />
<p>Visualization of <code class="docutils literal notranslate"><span class="pre">KEGG_ko</span></code> profiling</p>
<img alt="emapper example3" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/emapper_nifh_v3.png?raw=true" />
<p>Visualization of <code class="docutils literal notranslate"><span class="pre">domain</span></code> annotation</p>
<img alt="emapper example4" src="https://github.com/dengzq1234/treeprofiler_gallery/blob/main/emapper_nifh_v4.png?raw=true" />
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">treeprofiler</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">TreeProfiler Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="#table-of-contents">Table of Contents</a></li>
<li class="toctree-l1"><a class="reference internal" href="#installation">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#dependencies">Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quick-install-via-pip">Quick install via pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quick-start-with-examples-dataset">Quick Start with examples dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#manual-installation">Manual installation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#install-ete-v4">Install ETE v4</a></li>
<li class="toctree-l3"><a class="reference internal" href="#install-treeprofiler">Install TreeProfiler</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#input-files">Input files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#basic-usage">Basic Usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#using-treeprofiler">Using TreeProfiler</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#reading-input-tree">Reading input tree</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#tree-format">Tree format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tree-parser">Tree parser</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#treeprofiler-annotate-computing-phylogenetic-profiles-and-annotation"><code class="docutils literal notranslate"><span class="pre">treeprofiler-annotate</span></code> computing phylogenetic profiles and annotation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#annotate-metadata-into-tree">Annotate metadata into tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="#taxonomic-annotation">Taxonomic annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#functional-annotation">Functional annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#analytical-methods-in-treeprofiler">Analytical methods in treeProfiler</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-formats-for-annotated-trees">Output formats for annotated trees</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#treeprofiler-plot-visualizing-phylogenetic-profiles-and-annotation"><code class="docutils literal notranslate"><span class="pre">treeprofiler-plot</span></code> visualizing phylogenetic profiles and annotation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#interactive-visualization-interface">Interactive visualization interface</a></li>
<li class="toctree-l3"><a class="reference internal" href="#basic-options-of-visualizing-layouts">Basic options of visualizing layouts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-categorical-data">Layouts for categorical data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-boolean-data">Layouts for boolean data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-numerical-data">Layouts for Numerical data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-analytical-data">Layouts for Analytical Data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-multiple-sequence-alignment">Layouts for Multiple Sequence Alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-eggnog-mapper-pfam-smart-annotations">Layouts for eggnog-mapper pfam/smart annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-eggnog-mapper-functional-annotations">Layouts for eggnog-mapper functional annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#layouts-for-taxonomy-annotation">Layouts for Taxonomy annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#customize-color-in-layout-with-color-config">Customize color in layout with color config</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#conditional-query-in-annotated-tree-can-be-used-in-both-annotate-and-plot">Conditional query in annotated tree (can be used in both annotate and plot)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#query-syntax">Query Syntax</a></li>
<li class="toctree-l3"><a class="reference internal" href="#conditional-limit-based-on-taxonomic-level">Conditional limit based on taxonomic level</a></li>
<li class="toctree-l3"><a class="reference internal" href="#choose-the-metric-for-internal-nodes-in-numerical-layout">Choose the metric for internal nodes in numerical layout</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#demo1-explore-gtdb-taxonomic-tree-with-metadata-and-habitat-information-of-progenome3">Demo1 Explore GTDB taxonomic tree with metadata and habitat information of progenome3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#demo2-explore-large-nifh-gene-tree-with-functional-and-taxonomic-information">Demo2 Explore large NifH gene tree with functional and taxonomic information</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to treeprofiler’s documentation!</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Ziqi Deng & Jaime Huerta-Cepas.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/documentations.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>